



```{r message=FALSE, warning=FALSE}
library(data.table)
library(stringr)
library(rhmmer)
library(Biostrings)
library(IRanges)
library(GenomicRanges)
library(ggplot2)
library(harrypotter)
library(gameofthrones)
library(dplyr)
library(BSgenome)
library(DescTools)
library(tidyr)
library(Gviz)
library(ggpubr)
library(cowplot)
setwd("F:/1_script/rad")

all_rnaspades<-fread("all_rnaspades_trans.txt")
all_trinity<-fread("all_trinity_trans.txt")
setnames(all_rnaspades,c("transcript","length"))
setnames(all_trinity,c("transcript","length"))
#rRNA filtering 
blastn_rrna_rnaspades<-fread("blastn_rnaspades_together_to_rrna_both_subunits.outfmt6")
blastn_rrna_trinity<-fread("blastn_trinity_together_to_rrna_both_subunits.outfmt6")
setnames(blastn_rrna_rnaspades,c("qseqid","sseqid","pident","length","mismatch","gapopen","qstart","qend","qlen","sstart","send","slen","evalue","bitscore"))
setnames(blastn_rrna_trinity,c("qseqid","sseqid","pident","length","mismatch","gapopen","qstart","qend","qlen","sstart","send","slen","evalue","bitscore"))


rrna_rnaspades<-unique(blastn_rrna_rnaspades[evalue<=1e-20]$qseqid)
rrna_trinity<-unique(blastn_rrna_trinity[evalue<=1e-20]$qseqid)

rnaspades_without_rrna<-all_rnaspades[transcript%in%rrna_rnaspades==F]
trinity_without_rrna<-all_trinity[transcript%in%rrna_trinity==F]

write.table(rnaspades_without_rrna,file="rnaspades_without_rrna.txt",row.names = F,col.names = F,quote = F,sep = "\t")
write.table(trinity_without_rrna,file="trinity_without_rrna.txt",row.names = F,col.names = F,quote = F,sep = "\t")

#Length filtering - 200 

all_rnaspades_no_rrna_more_than_200<-rnaspades_without_rrna[length>200]
all_trinity_no_rrna_more_than_200<-trinity_without_rrna[length>200]


write.table(all_rnaspades_no_rrna_more_than_200$transcript,file="rnaspades_without_rrna_more_than_200.txt",row.names = F,col.names = F,quote = F,sep = "\t")
write.table(all_trinity_no_rrna_more_than_200$transcript,file="trinity_without_rrna_more_than_200.txt",row.names = F,col.names = F,quote = F,sep = "\t")

#ORF filtering
orf_table_transdecoder_rnaspades<-fread("SPAdesd1d10_transdecoder.cds_table.csv")
orf_table_transdecoder_trinity<-fread("Trinityd1d10_transdecoder.cds_table.csv")


setnames(orf_table_transdecoder_rnaspades,c("protein","feature","feature_status","feature_length","strand","score"))
setnames(orf_table_transdecoder_trinity,c("protein","feature","feature_status","feature_length","strand","score"))

orf_table_transdecoder_rnaspades[,transcript:=str_extract(protein,".*(?=\\.p.+)")]
orf_table_transdecoder_trinity[,transcript:=str_extract(protein,".*(?=\\.p.+)")]

orf_table_transdecoder_rnaspades_maxlen<-orf_table_transdecoder_rnaspades[,.(max_length=max(feature_length)),by=transcript]
orf_table_transdecoder_trinity_maxlen<-orf_table_transdecoder_trinity[,.(max_length=max(feature_length)),by=transcript]


less_than_150_orfs_rnaspades<-unique(orf_table_transdecoder_rnaspades_maxlen[max_length<=75]$transcript)
less_than_150_orfs_rtrinity<-unique(orf_table_transdecoder_trinity_maxlen[max_length<=75]$transcript)

no_orfs_rnaspades<-setdiff(all_rnaspades_no_rrna_more_than_200$transcript,orf_table_transdecoder_rnaspades$transcript)
no_orfs_trinity<-setdiff(all_trinity_no_rrna_more_than_200$transcript,orf_table_transdecoder_trinity$transcript)

less_than_150_or_no_orfs_rnaspades<-c(no_orfs_rnaspades,less_than_150_orfs_rnaspades)
less_than_150_or_no_orfs_trinity<-c(no_orfs_trinity,less_than_150_orfs_rtrinity)

write.table(less_than_150_or_no_orfs_rnaspades,file="less_than_150_or_no_orfs_rnaspades.txt",row.names = F,col.names = F,quote = F,sep = "\t")
write.table(less_than_150_or_no_orfs_trinity,file="less_than_150_or_no_orfs_trinity.txt",row.names = F,col.names = F,quote = F,sep = "\t")

#BLASTX filtering
diamond_rnaspades<-fread("diamond_rnaspades_1_10_nr.txt")
diamond_trinity<-fread("diamond_trinity_1_10_nr.txt")
#setnames(diamond_rnaspades,c("qseqid","qlen","sseqid","slen","staxid","stitle","qcovhsp","length","pident","mismatch","qstart","qend","sstart","send","evalue","bitscore"))
#setnames(diamond_trinity,c("qseqid","qlen","sseqid","slen","staxid","stitle","qcovhsp","length","pident","mismatch","qstart","qend","sstart","send","evalue","bitscore"))
setnames(diamond_rnaspades,c("qseqid","sseqid","qcovhsp","length","pident","mismatch","qstart","qend","sstart","send","evalue","bitscore"))
setnames(diamond_trinity,c("qseqid","sseqid","qcovhsp","length","pident","mismatch","qstart","qend","sstart","send","evalue","bitscore"))


diamond_rnaspades_besthit<-diamond_rnaspades[,.(eval_min=min(evalue)),by=qseqid]
diamond_trinity_besthit<-diamond_trinity[,.(eval_min=min(evalue)),by=qseqid]


rnaspades_less_than_150_ORF_without_blast<-setdiff(less_than_150_or_no_orfs_rnaspades,diamond_rnaspades_besthit[eval_min<=1e-10]$qseqid)
trinity_less_than_150_ORF_without_blast<-setdiff(less_than_150_or_no_orfs_trinity,diamond_trinity_besthit[eval_min<=1e-10]$qseqid)


write.table(rnaspades_less_than_150_ORF_without_blast,file="rnaspades_less_than_150_ORF_without_blast.txt",row.names = F,col.names = F,quote = F,sep = "\t")
write.table(trinity_less_than_150_ORF_without_blast,file="trinity_less_than_150_ORF_without_blast.txt",row.names = F,col.names = F,quote = F,sep = "\t")


#HMMER filtering
hmmer_rnaspades<-as.data.table(read_domtblout("PFAM_Esu_rnaspades_pep_out.txt"))
hmmer_trinity<-as.data.table(read_domtblout("PFAM_Esu_trinity_pep_out.txt"))


hmmer_rnaspades[,transcript:=str_extract(query_name,".*(?=\\.p.+)")]
hmmer_trinity[,transcript:=str_extract(query_name,".*(?=\\.p.+)")]


hmmer_rnaspades_besthit<-hmmer_rnaspades[,.(eval_min=min(domain_ievalue)),by=transcript]
hmmer_trinity_besthit<-hmmer_trinity[,.(eval_min=min(domain_ievalue)),by=transcript]


rnaspades_no_hmmer<-setdiff(rnaspades_less_than_150_ORF_without_blast,hmmer_rnaspades_besthit[eval_min<=1e-5]$transcript)
trinity_no_hmmer<-setdiff(trinity_less_than_150_ORF_without_blast,hmmer_trinity_besthit[eval_min<=1e-5]$transcript)


write.table(rnaspades_no_hmmer,file="rnaspades_no_hmmer.txt",row.names = F,col.names = F,quote = F,sep = "\t")
write.table(trinity_no_hmmer,file="trinity_no_hmmer.txt",row.names = F,col.names = F,quote = F,sep = "\t")

#Feelnc

feelncrna_rnaspades_sequences<-readDNAStringSet("rnaspades_no_hmmer.fa.lncRNA.fa",format = "fasta")
feelncrna_trinity_sequences<-readDNAStringSet("trinity_no_hmmer.fa.lncRNA.fa",format = "fasta")

feelncrna_rnaspades<-names(feelncrna_rnaspades_sequences)
feelncrna_trinity<-names(feelncrna_trinity_sequences)
feelncrna_trinity<-str_extract(feelncrna_trinity,"TRINITY.*i\\d+")

#mapping

rnaspades_paf<-fread("cut_rnaspades_together_lncrna_to_nohic_Esu.paf")
trinity_paf<-fread("cut_trinity_together_lncrna_to_nohic_Esu.paf")
setnames(rnaspades_paf,c("query_name","query_length","query_start","query_end","strand_ignore","target_name","target_length","target_start","target_end","matches","alignment_block","mapping_quality"))
setnames(trinity_paf,c("query_name","query_length","query_start","query_end","strand_ignore","target_name","target_length","target_start","target_end","matches","alignment_block","mapping_quality"))


rnaspades_paf[,numb_same_trans:=nrow(.SD),by=query_name]
rnaspades_paf[numb_same_trans>1,query_name:=paste(query_name,numb_same_trans,sep="_")]
trinity_paf[,numb_same_trans:=nrow(.SD),by=query_name]
trinity_paf[numb_same_trans>1,query_name:=paste(query_name,numb_same_trans,sep="_")]
rnaspades_paf$numb_same_trans<-NULL
trinity_paf$numb_same_trans<-NULL

rnaspades_paf[,perc_matches:=100*(matches/query_length)]
trinity_paf[,perc_matches:=100*(matches/query_length)]


rnaspades_paf_perc<-rnaspades_paf[perc_matches>=95]
trinity_paf_perc<-trinity_paf[perc_matches>=95]

#bacterial scaffolds filtering 

taxonomy_table<-fread("MEGAN_my_taxonomy280220.csv")

taxonomy_table_nonbact<-unique(taxonomy_table[my_taxonomy=="Metazoa" | my_taxonomy=="Other"]$seqid )

rnaspades_paf_nobact<-rnaspades_paf_perc[target_name %in% taxonomy_table_nonbact]
trinity_paf_nobact<-trinity_paf_perc[target_name %in% taxonomy_table_nonbact]

write.table(taxonomy_table_nonbact,file="metazoan_scaffolds.txt",row.names = F,col.names = F,quote = F,sep = "\t")

#one exon filtering
rnaspades_lncrna_exons<-fread("rnaspades_esu_minimap_exons.txt")
trinity_lncrna_exons<-fread("trinity_esu_minimap_exons.txt")

rnaspades_lncrna_more_than_1_exon<-rnaspades_lncrna_exons[,.N,by=transcript][N!=1]$transcript
trinity_lncrna_more_than_1_exon<-trinity_lncrna_exons[,.N,by=transcript][N!=1]$transcript

rnaspades_lncrna_paf_more_than_one_exon<-rnaspades_paf_nobact[query_name%in%rnaspades_lncrna_more_than_1_exon]
trinity_lncrna_paf_more_than_one_exon<-trinity_paf_nobact[query_name%in%trinity_lncrna_more_than_1_exon]




protein_coding_annotation<-fread("augustus.hints.gtf",fill=T)[is.na(V4)==F]
setnames(protein_coding_annotation,c("seqname","source","feature","start","end","score","strand_ignore","frame","attribute"))

protein_coding_annotation<-protein_coding_annotation[seqname%in%taxonomy_table_nonbact]

protein_exons_dt<-protein_coding_annotation[feature=="CDS"]

protein_exons<-makeGRangesFromDataFrame(df=protein_exons_dt,keep.extra.columns = T,ignore.strand = T)



rnaspades_filtered_exons<-rnaspades_lncrna_exons[transcript%in%rnaspades_lncrna_paf_more_than_one_exon$query_name]
trinity_filtered_exons<-trinity_lncrna_exons[transcript%in%trinity_lncrna_paf_more_than_one_exon$query_name]


colnames(rnaspades_filtered_exons)[1]<-"query_name"
colnames(trinity_filtered_exons)[1]<-"query_name"
rnaspades_filtered_exons<-merge(rnaspades_filtered_exons,rnaspades_lncrna_paf_more_than_one_exon[,.SD,.SDcols=c("query_name","target_start","target_end")])
trinity_filtered_exons<-merge(trinity_filtered_exons,trinity_lncrna_paf_more_than_one_exon[,.SD,.SDcols=c("query_name","target_start","target_end")])
rnaspades_filtered_exons[,is_first:=ifelse(.SD==min(.SD),T,F),.SDcols="start",by=query_name]
rnaspades_filtered_exons[,change_ex_first:=ifelse(is_first==T & start!=target_start,T,F)]
rnaspades_filtered_exons[change_ex_first==T,start:=target_start]
rnaspades_filtered_exons[,is_last:=ifelse(.SD==max(.SD),T,F),.SDcols="end",by=query_name]
rnaspades_filtered_exons[,change_ex_last:=ifelse(is_last==T & end!=target_end,T,F)]
rnaspades_filtered_exons[change_ex_last==T,end:=target_end]
rnaspades_filtered_exons<-rnaspades_filtered_exons[,.SD,.SDcols=c("query_name","chromosome","start","end")]
colnames(rnaspades_filtered_exons)[1]<-"transcript"




trinity_filtered_exons[,is_first:=ifelse(.SD==min(.SD),T,F),.SDcols="start",by=query_name]
trinity_filtered_exons[,change_ex_first:=ifelse(is_first==T & start!=target_start,T,F)]
trinity_filtered_exons[change_ex_first==T,start:=target_start]
trinity_filtered_exons[,is_last:=ifelse(.SD==max(.SD),T,F),.SDcols="end",by=query_name]
trinity_filtered_exons[,change_ex_last:=ifelse(is_last==T & end!=target_end,T,F)]
trinity_filtered_exons[change_ex_last==T,end:=target_end]
trinity_filtered_exons<-trinity_filtered_exons[,.SD,.SDcols=c("query_name","chromosome","start","end")]
colnames(trinity_filtered_exons)[1]<-"transcript"












rnaspades_filtered_exons_granges<-makeGRangesFromDataFrame(rnaspades_filtered_exons,seqnames.field = "chromosome",ignore.strand = T,keep.extra.columns = T)
trinity_filtered_exons_granges<-makeGRangesFromDataFrame(trinity_filtered_exons,seqnames.field = "chromosome",ignore.strand = T,keep.extra.columns = T)

overlaps_lncrnaexon_genomexon_rnaspades<-findOverlaps(rnaspades_filtered_exons_granges,protein_exons)

rnaspades_transcripts_coding_overlap<-unique(rnaspades_filtered_exons[queryHits(overlaps_lncrnaexon_genomexon_rnaspades)]$transcript)

overlaps_lncrnaexon_genomexon_trinity<-findOverlaps(trinity_filtered_exons_granges,protein_exons)

trinity_transcripts_coding_overlap<-unique(trinity_filtered_exons[queryHits(overlaps_lncrnaexon_genomexon_trinity)]$transcript)

rnaspades_paf_filtered<-rnaspades_lncrna_paf_more_than_one_exon[query_name%in%rnaspades_transcripts_coding_overlap==F]
trinity_paf_filtered<-trinity_lncrna_paf_more_than_one_exon[query_name%in%trinity_transcripts_coding_overlap==F]

rnaspades_filtered_list<-rnaspades_paf_filtered$query_name
trinity_filtered_list<-trinity_paf_filtered$query_name


rnaspades_filtered_exons<-rnaspades_filtered_exons[transcript%in%rnaspades_filtered_list]
trinity_filtered_exons<-trinity_filtered_exons[transcript%in%trinity_filtered_list]


write.table(rnaspades_filtered_exons,file="rnaspades_filtered_exons.txt",row.names = F,quote = F,sep = "\t")
write.table(trinity_filtered_exons,file="trinity_filtered_exons.txt",row.names = F,quote = F,sep = "\t")





lengths_table<-data.table(filtering=rep(c("All transcripts","rRNA filtering","Length filtering","ORF filtering","DIAMOND filtering","HMMER filtering","FEELnc filtering","Percent identity filtering","Bacterial lncRNA filtering ","Exon number filtering","Protein-coding exons overlap filtering"),2),number_of_lncRNA_candidates=c(nrow(all_rnaspades),nrow(rnaspades_without_rrna),nrow(all_rnaspades_no_rrna_more_than_200),length(less_than_150_or_no_orfs_rnaspades),length(rnaspades_less_than_150_ORF_without_blast),length(rnaspades_no_hmmer),length(feelncrna_rnaspades),nrow(rnaspades_paf_perc),nrow(rnaspades_paf_nobact),nrow(rnaspades_lncrna_paf_more_than_one_exon),length(rnaspades_filtered_list),nrow(all_trinity),nrow(trinity_without_rrna),nrow(all_trinity_no_rrna_more_than_200),length(less_than_150_or_no_orfs_trinity),length(trinity_less_than_150_ORF_without_blast),length(trinity_no_hmmer),length(feelncrna_trinity),nrow(trinity_paf_perc),nrow(trinity_paf_nobact),nrow(trinity_lncrna_paf_more_than_one_exon),length(trinity_filtered_list)
),Transcriptome=c(rep("rnaSPAdes",11),rep("Trinity",11)))

lengths_table[,filtering:=factor(filtering,levels=rev(c("All transcripts","rRNA filtering","Length filtering","ORF filtering","DIAMOND filtering","HMMER filtering","FEELnc filtering","Percent identity filtering","Bacterial lncRNA filtering ","Exon number filtering","Protein-coding exons overlap filtering")))]


lengths_plot<-ggplot(data=lengths_table,aes(x=filtering,y=number_of_lncRNA_candidates,fill=Transcriptome)) +
  geom_bar(position="dodge",stat = "identity", width=0.65,col="black") +
  theme_bw() +
  facet_grid(row=vars(Transcriptome)) +
  theme(legend.position = "top") +
  theme(axis.title.y = element_blank()) +
  coord_flip() +
  ylab("Number of transcripts") +
  scale_x_discrete(labels=c("Protein-coding exons overlap filtering","Exon number filtering","Bacterial lncRNA filtering","Percent identity filtering","FEELnc filtering","HMMER filtering","DIAMOND filtering","ORF filtering","Length filtering","rRNA filtering","All transcripts"))
lengths_plot
ggsave(lengths_plot,filename = "lengths_plot.png",width = 7,height = 5)



#rnaspades and trinity union

rnaspades_paf_filtered[,gene_name:=str_extract(query_name,"NODE_\\d*")]
trinity_paf_filtered[,gene_name:=str_extract(query_name,".*(?=_i\\d*)")]
rnaspades_paf_filtered_granges<-makeGRangesFromDataFrame(rnaspades_paf_filtered,seqnames.field = "target_name",start.field = "target_start",end.field = "target_end",ignore.strand=T,keep.extra.columns = T)
trinity_paf_filtered_granges<-makeGRangesFromDataFrame(trinity_paf_filtered,seqnames.field = "target_name",start.field = "target_start",end.field = "target_end",ignore.strand=T,keep.extra.columns = T)


#finding genes
rnaspades_paf_filtered_granges_split<-split(rnaspades_paf_filtered_granges,rnaspades_paf_filtered_granges$gene_name)
trinity_paf_filtered_granges_split<-split(trinity_paf_filtered_granges,trinity_paf_filtered_granges$gene_name)

rnaspades_paf_filtered_granges_reduced<-reduce(rnaspades_paf_filtered_granges_split,ignore.strand=T)
trinity_paf_filtered_granges_reduced<-reduce(trinity_paf_filtered_granges_split,ignore.strand=T)

rnaspades_paf_filtered_granges_reduced<-unlist(rnaspades_paf_filtered_granges_reduced,use.names = T)
trinity_paf_filtered_granges_reduced<-unlist(trinity_paf_filtered_granges_reduced,use.names = T)
rnaspades_paf_filtered_granges_reduced$gene<-names(rnaspades_paf_filtered_granges_reduced)
rnaspades_paf_filtered_granges_reduced<-unique(rnaspades_paf_filtered_granges_reduced)
trinity_paf_filtered_granges_reduced$gene<-names(trinity_paf_filtered_granges_reduced)
trinity_paf_filtered_granges_reduced<-unique(trinity_paf_filtered_granges_reduced)
#removing the nested overlaps of genes which don't overlap between the transcriptomes



nested_overlaps_rnaspades<-findOverlaps(rnaspades_paf_filtered_granges_reduced,rnaspades_paf_filtered_granges_reduced,type = "within")
nested_overlap_table_rnaspades<-cbind(as.data.table(rnaspades_paf_filtered_granges_reduced[queryHits(nested_overlaps_rnaspades)]),as.data.table(rnaspades_paf_filtered_granges_reduced[subjectHits(nested_overlaps_rnaspades)]))

setnames(nested_overlap_table_rnaspades,c("seqnames1","start1","end1","width1","strand1","gene1","seqnames2","start2","end2","width2","strand2","gene2"))

inter_width_vector<-c()
for (i in 1:nrow(nested_overlap_table_rnaspades)) {
  inter_width_vector<-c(inter_width_vector,1+Overlap(c(nested_overlap_table_rnaspades$start1[i],nested_overlap_table_rnaspades$end1[i]),c(nested_overlap_table_rnaspades$start2[i],nested_overlap_table_rnaspades$end2[i])))
}
nested_overlap_table_rnaspades[,intersect_widths:=inter_width_vector]
nested_overlap_table_rnaspades[width1<=width2,max_width:=width2]
nested_overlap_table_rnaspades[width1>width2,max_width:=width1]
nested_overlap_table_rnaspades[,length_frac:=100*(intersect_widths/max_width)]
nested_overlap_table_rnaspades<-nested_overlap_table_rnaspades[length_frac!=100]
nested_overlap_table_rnaspades_genes<-unique(nested_overlap_table_rnaspades[,.SD,.SDcols=c("seqnames1","start1","end1","width1","strand1","gene1")])
setnames(nested_overlap_table_rnaspades_genes,c("seqnames","start","end","width","strand","gene"))


nested_overlaps_trinity<-findOverlaps(trinity_paf_filtered_granges_reduced,trinity_paf_filtered_granges_reduced,type = "within")
nested_overlap_table_trinity<-cbind(as.data.table(trinity_paf_filtered_granges_reduced[queryHits(nested_overlaps_trinity)]),as.data.table(trinity_paf_filtered_granges_reduced[subjectHits(nested_overlaps_trinity)]))

setnames(nested_overlap_table_trinity,c("seqnames1","start1","end1","width1","strand1","gene1","seqnames2","start2","end2","width2","strand2","gene2"))

inter_width_vector<-c()
for (i in 1:nrow(nested_overlap_table_trinity)) {
  inter_width_vector<-c(inter_width_vector,1+Overlap(c(nested_overlap_table_trinity$start1[i],nested_overlap_table_trinity$end1[i]),c(nested_overlap_table_trinity$start2[i],nested_overlap_table_trinity$end2[i])))
}
nested_overlap_table_trinity[,intersect_widths:=inter_width_vector]
nested_overlap_table_trinity[width1<=width2,max_width:=width2]
nested_overlap_table_trinity[width1>width2,max_width:=width1]
nested_overlap_table_trinity[,length_frac:=100*(intersect_widths/max_width)]
nested_overlap_table_trinity<-nested_overlap_table_trinity[length_frac!=100]
nested_overlap_table_trinity_genes<-unique(nested_overlap_table_trinity[,.SD,.SDcols=c("seqnames1","start1","end1","width1","strand1","gene1")])
setnames(nested_overlap_table_trinity_genes,c("seqnames","start","end","width","strand","gene"))


#finding between-transcriptome gene overlaps

rnaspades_paf_filtered_granges_reduced_dt<-as.data.table(rnaspades_paf_filtered_granges_reduced)
rnaspades_paf_filtered_granges_reduced_dt<-anti_join(rnaspades_paf_filtered_granges_reduced_dt,nested_overlap_table_rnaspades_genes)
rnaspades_paf_filtered_granges_reduced<-makeGRangesFromDataFrame(rnaspades_paf_filtered_granges_reduced_dt,keep.extra.columns = T)
trinity_paf_filtered_granges_reduced_dt<-as.data.table(trinity_paf_filtered_granges_reduced)
trinity_paf_filtered_granges_reduced_dt<-anti_join(trinity_paf_filtered_granges_reduced_dt,nested_overlap_table_trinity_genes)
trinity_paf_filtered_granges_reduced<-makeGRangesFromDataFrame(trinity_paf_filtered_granges_reduced_dt,keep.extra.columns = T)

reduced_overlaps<-findOverlaps(rnaspades_paf_filtered_granges_reduced,trinity_paf_filtered_granges_reduced,ignore.strand=T)
overlap_table<-cbind(as.data.table(rnaspades_paf_filtered_granges_reduced[queryHits(reduced_overlaps)]),as.data.table(trinity_paf_filtered_granges_reduced[subjectHits(reduced_overlaps)]))

setnames(overlap_table,c("seqnames1","start1","end1","width1","strand1","gene1","seqnames2","start2","end2","width2","strand2","gene2"))


inter_width_vector<-c()
for (i in 1:nrow(overlap_table)) {
  inter_width_vector<-c(inter_width_vector,Overlap(c(overlap_table$start1[i],overlap_table$end1[i]),c(overlap_table$start2[i],overlap_table$end2[i])))
}
overlap_table[,intersect_widths:=inter_width_vector]
overlap_table[width1<=width2,max_width:=width2]
overlap_table[width1>width2,max_width:=width1]


overlap_table[,length_frac:=100*(intersect_widths/max_width)]

#intersect plot

perc_plot<-ggplot(data=overlap_table,aes(x=length_frac))+
  geom_histogram(fill="indianred1",col="black",binwidth = 5) +
  theme_bw() +
  xlab("Intersect percentage") +
  ylab("Number of rnaSPades and Trinity gene pairs") +
  theme(axis.title.y = element_text(size=8))

perc_plot

ggsave(perc_plot,file="perc_plot.jpg",width = 5,height = 4)

overlap_table_filtered_more_than_20_length_frac<-overlap_table[length_frac>=20]
overlap_table_filtered_less_than_20_length_frac<-overlap_table[length_frac<20]

overlap_table_filtered_more_than_20_length_frac[,':='(first_longer=ifelse(width1==max_width,T,F),second_longer=ifelse(width2==max_width,T,F))]

reduced_morethan20_longer_rnaspades<-overlap_table_filtered_more_than_20_length_frac[first_longer==T,.SD,.SDcols=c("seqnames1","start1","end1","width1","gene1")]
reduced_morethan20_longer_trinity<-overlap_table_filtered_more_than_20_length_frac[second_longer==T & first_longer!=T,.SD,.SDcols=c("seqnames2","start2","end2","width2","gene2")]
setnames(reduced_morethan20_longer_rnaspades,c("seqnames","start","end","width","gene"))
setnames(reduced_morethan20_longer_trinity,c("seqnames","start","end","width","gene"))
reduced_morethan20_longer_together<-unique(rbind(reduced_morethan20_longer_rnaspades,reduced_morethan20_longer_trinity))

reduced_morethan20_shorter_rnaspades<-overlap_table_filtered_more_than_20_length_frac[first_longer==F,.SD,.SDcols=c("seqnames1","start1","end1","width1","gene1")]
reduced_morethan20_shorter_trinity<-overlap_table_filtered_more_than_20_length_frac[second_longer==F,.SD,.SDcols=c("seqnames2","start2","end2","width2","gene2")]
setnames(reduced_morethan20_shorter_rnaspades,c("seqnames","start","end","width","gene"))
setnames(reduced_morethan20_shorter_trinity,c("seqnames","start","end","width","gene"))
reduced_morethan20_shorter_together<-unique(rbind(reduced_morethan20_shorter_rnaspades,reduced_morethan20_shorter_trinity))


overlap_table_filtered_less_than_20_length_frac[,':='(first_longer=ifelse(width1==max_width,T,F),second_longer=ifelse(width2==max_width,T,F))]

reduced_lessthan20_longer_rnaspades<-unique(overlap_table_filtered_less_than_20_length_frac[first_longer==T,.SD,.SDcols=c("seqnames1","start1","end1","width1","gene1")])
reduced_lessthan20_longer_trinity<-unique(overlap_table_filtered_less_than_20_length_frac[second_longer==T,.SD,.SDcols=c("seqnames2","start2","end2","width2","gene2")])
setnames(reduced_lessthan20_longer_rnaspades,c("seqnames","start","end","width","gene"))
setnames(reduced_lessthan20_longer_trinity,c("seqnames","start","end","width","gene"))
reduced_lessthan20_longer_together<-unique(rbind(reduced_lessthan20_longer_rnaspades,reduced_lessthan20_longer_trinity))

reduced_lessthan20_shorter_rnaspades<-unique(overlap_table_filtered_less_than_20_length_frac[first_longer==F,.SD,.SDcols=c("seqnames1","start1","end1","width1","gene1")])
reduced_lessthan20_shorter_trinity<-unique(overlap_table_filtered_less_than_20_length_frac[second_longer==F,.SD,.SDcols=c("seqnames2","start2","end2","width2","gene2")])
setnames(reduced_lessthan20_shorter_rnaspades,c("seqnames","start","end","width","gene"))
setnames(reduced_lessthan20_shorter_trinity,c("seqnames","start","end","width","gene"))


reduced_lessthan20_shorter_rnaspades_gr<-makeGRangesFromDataFrame(reduced_lessthan20_shorter_rnaspades,keep.extra.columns = T)
short_within_overlap_rnaspades<-findOverlaps(reduced_lessthan20_shorter_rnaspades_gr,trinity_paf_filtered_granges_reduced,ignore.strand=T,type="within")
short_within_overlap_rnaspades_table<-cbind(as.data.table(reduced_lessthan20_shorter_rnaspades_gr[queryHits(short_within_overlap_rnaspades)]),as.data.table(trinity_paf_filtered_granges_reduced[subjectHits(short_within_overlap_rnaspades)]))
setnames(short_within_overlap_rnaspades_table,c("seqnames","start","end","width","strand","gene","seqnames2","start2","end2","width2","strand2","gene2"))
reduced_lessthan20_shorter_rnaspades<-anti_join(reduced_lessthan20_shorter_rnaspades,short_within_overlap_rnaspades_table[,.SD,.SDcols=c("seqnames","start","end","width","gene")])
rnaspades_paf_filtered_granges_reduced_dt<-as.data.table(rnaspades_paf_filtered_granges_reduced)
rnaspades_paf_filtered_granges_reduced_dt<-anti_join(rnaspades_paf_filtered_granges_reduced_dt,short_within_overlap_rnaspades_table[,.SD,.SDcols=c("seqnames","start","end","width","strand","gene")])
rnaspades_paf_filtered_granges_reduced<-makeGRangesFromDataFrame(rnaspades_paf_filtered_granges_reduced_dt,keep.extra.columns = T)








diff_rnaspades_less_longer<-anti_join(as.data.table(rnaspades_paf_filtered_granges_reduced)[,.SD,.SDcols=c("seqnames","start","end","width","gene")],reduced_lessthan20_longer_rnaspades)

diff_rnaspades_less_and_more_longer<-anti_join(diff_rnaspades_less_longer,reduced_morethan20_longer_rnaspades)
diff_rnaspades_less_and_more_longer_less_shorter<-anti_join(diff_rnaspades_less_and_more_longer,reduced_lessthan20_shorter_rnaspades)
diff_rnaspades_less_and_more_longer_less_and_more_shorter<-anti_join(diff_rnaspades_less_and_more_longer_less_shorter,reduced_morethan20_shorter_rnaspades)

	
reduced_lessthan20_shorter_trinity_gr<-makeGRangesFromDataFrame(reduced_lessthan20_shorter_trinity,keep.extra.columns = T)
short_within_overlap_trinity<-findOverlaps(reduced_lessthan20_shorter_trinity_gr,rnaspades_paf_filtered_granges_reduced,ignore.strand=T,type="within")
short_within_overlap_trinity_table<-cbind(as.data.table(reduced_lessthan20_shorter_trinity_gr[queryHits(short_within_overlap_trinity)]),as.data.table(rnaspades_paf_filtered_granges_reduced[subjectHits(short_within_overlap_trinity)]))
setnames(short_within_overlap_trinity_table,c("seqnames","start","end","width","strand","gene","seqnames2","start2","end2","width2","strand2","gene2"))
reduced_lessthan20_shorter_trinity<-anti_join(reduced_lessthan20_shorter_trinity,short_within_overlap_trinity_table[,.SD,.SDcols=c("seqnames","start","end","width","gene")])
trinity_paf_filtered_granges_reduced_dt<-as.data.table(trinity_paf_filtered_granges_reduced)
trinity_paf_filtered_granges_reduced_dt<-anti_join(trinity_paf_filtered_granges_reduced_dt,short_within_overlap_trinity_table[,.SD,.SDcols=c("seqnames","start","end","width","strand","gene")])
trinity_paf_filtered_granges_reduced<-makeGRangesFromDataFrame(trinity_paf_filtered_granges_reduced_dt,keep.extra.columns = T)
reduced_lessthan20_shorter_together<-unique(rbind(reduced_lessthan20_shorter_rnaspades,reduced_lessthan20_shorter_trinity))




diff_trinity_less_longer<-anti_join(as.data.table(trinity_paf_filtered_granges_reduced)[,.SD,.SDcols=c("seqnames","start","end","width","gene")],reduced_lessthan20_longer_trinity)
diff_trinity_less_and_more_longer<-anti_join(diff_trinity_less_longer,reduced_morethan20_longer_trinity)
diff_trinity_less_and_more_longer_less_shorter<-anti_join(diff_trinity_less_and_more_longer,reduced_lessthan20_shorter_trinity)
diff_trinity_less_and_more_longer_less_and_more_shorter<-anti_join(diff_trinity_less_and_more_longer_less_shorter,reduced_morethan20_shorter_trinity)


#lncRNAs which I will use

lncrnas_final_reduced_rnaspades<-unique(makeGRangesFromDataFrame(rbind(reduced_morethan20_longer_rnaspades,reduced_lessthan20_longer_rnaspades,reduced_lessthan20_shorter_rnaspades,diff_rnaspades_less_and_more_longer_less_and_more_shorter),keep.extra.columns = T))
lncrnas_final_reduced_rnaspades$gene_id<-1:length(lncrnas_final_reduced_rnaspades)
lncrnas_final_reduced_trinity<-unique(makeGRangesFromDataFrame(rbind(reduced_morethan20_longer_trinity,reduced_lessthan20_longer_trinity,reduced_lessthan20_shorter_trinity,diff_trinity_less_and_more_longer_less_and_more_shorter),keep.extra.columns = T))
#removing the 100% overlaps from one of the transcriptomes (Trinity)
total_overlap<-overlap_table_filtered_more_than_20_length_frac[first_longer==second_longer]
setnames(total_overlap,c("seqnames1","start1","end1","width1","strand1","gene1","seqnames","start","end","width","strand","gene","intersect_widths","max_width","length_frac","first_longer","second_longer"))
lncrnas_final_reduced_trinity_dt<-as.data.table(lncrnas_final_reduced_trinity)
lncrnas_final_reduced_trinity_dt_anti<-anti_join(lncrnas_final_reduced_trinity_dt,total_overlap[,.SD,.SDcols=c("seqnames","start","end","width","strand","gene")])
lncrnas_final_reduced_trinity<-makeGRangesFromDataFrame(lncrnas_final_reduced_trinity_dt_anti,keep.extra.columns = T)
	
trinity_paf_filtered_granges<-trinity_paf_filtered_granges[trinity_paf_filtered_granges$gene_name%in%nested_overlap_table_trinity_genes$gene==F]


lncrnas_final_reduced_trinity$gene_id<-(length(lncrnas_final_reduced_rnaspades)+1):(length(lncrnas_final_reduced_trinity)+length(lncrnas_final_reduced_rnaspades))
lncrnas_final_reduced_together<-unique(c(lncrnas_final_reduced_rnaspades,lncrnas_final_reduced_trinity))


#numbers
all_together_reduced_num<-length(rnaspades_paf_filtered_granges_reduced)+length(trinity_paf_filtered_granges_reduced)

rnaspades_reduced_length<-length(rnaspades_paf_filtered_granges_reduced)

trinity_reduced_length<-length(trinity_paf_filtered_granges_reduced)

same_overlaps_reduced_num<-nrow(total_overlap)

overlap_rnaspades_morethan20_longer_reduced_num<-nrow(reduced_morethan20_longer_rnaspades)

overlap_trinity_morethan20_longer_reduced_num<-nrow(reduced_morethan20_longer_trinity)

overlap_rnaspades_lessthan20_longerreduced_num<-nrow(reduced_lessthan20_longer_rnaspades)

overlap_trinity_lessthan20_longer_reduced_num<-nrow(reduced_lessthan20_longer_trinity)

overlap_together_morethan20_longer_reduced_num<-nrow(reduced_morethan20_longer_together)

overlap_together_lessthan20_longer_reduced_num<-nrow(reduced_lessthan20_longer_together)

overlap_rnaspades_morethan20_shorter_reduced_num<-nrow(reduced_morethan20_shorter_rnaspades)

overlap_trinity_morethan20_shorter_reduced_num<-nrow(reduced_morethan20_shorter_trinity)

overlap_rnaspades_lessthan20_shorter_reduced_num<-nrow(reduced_lessthan20_shorter_rnaspades)

overlap_trinity_lessthan20_shorter_reduced_num<-nrow(reduced_lessthan20_shorter_trinity)

overlap_together_morethan20_shorter_reduced_num<-nrow(reduced_morethan20_shorter_together)

overlap_together_lessthan20_shorter_reduced_num<-nrow(reduced_lessthan20_shorter_together)

rnaspades_unique<-nrow(diff_rnaspades_less_and_more_longer_less_and_more_shorter)

trinity_unique<-nrow(diff_trinity_less_and_more_longer_less_and_more_shorter)

#final number of lncRNA genes
#rnaspades
lncrnas_final_reduced_num_rnaspades<-length(lncrnas_final_reduced_rnaspades)
#trinity
lncrnas_final_reduced_num_trinity<-length(lncrnas_final_reduced_trinity)
#together
lncrnas_final_reduced_num_together<-length(lncrnas_final_reduced_together)


write.table(as.data.table(lncrnas_final_reduced_together),file="lncrnas_final_reduced_together.txt",col.names = T,row.names = F,quote = F,sep = "\t")



#finding the transcripts

overlaps_isoform_reduced_rnaspades<-findOverlaps(rnaspades_paf_filtered_granges,lncrnas_final_reduced_rnaspades,ignore.strand=T)
overlaps_isoform_reduced_trinity<-findOverlaps(trinity_paf_filtered_granges,lncrnas_final_reduced_trinity,ignore.strand=T)

isoform_overlap_table_rnaspades<-unique(cbind(as.data.table(rnaspades_paf_filtered_granges[queryHits(overlaps_isoform_reduced_rnaspades)]),as.data.table(lncrnas_final_reduced_rnaspades[subjectHits(overlaps_isoform_reduced_rnaspades)])))
setnames(isoform_overlap_table_rnaspades,c("seqnames1","start1","end1","width1","strand1","transcript_name","transcript_length","transcript_start","transcript_end","transcript_strand","contig_length","matches","alignment_block","mapping_quality","perc_matches","gene_tr","seqnames2","start2","end2","width2","strand2","gene","gene_id"))

isoform_overlap_table_trinity<-unique(cbind(as.data.table(trinity_paf_filtered_granges[queryHits(overlaps_isoform_reduced_trinity)]),as.data.table(lncrnas_final_reduced_trinity[subjectHits(overlaps_isoform_reduced_trinity)])))
setnames(isoform_overlap_table_trinity,c("seqnames1","start1","end1","width1","strand1","transcript_name","transcript_length","transcript_start","transcript_end","transcript_strand","contig_length","matches","alignment_block","mapping_quality","perc_matches","gene_tr","seqnames2","start2","end2","width2","strand2","gene","gene_id"))







isoform_overlap_table_rnaspades<-isoform_overlap_table_rnaspades[gene_tr==gene]
isoform_overlap_table_trinity<-isoform_overlap_table_trinity[gene_tr==gene]


final_lncrna_list_rnaspades<-unique(isoform_overlap_table_rnaspades$transcript_name)
final_lncrna_list_trinity<-unique(isoform_overlap_table_trinity$transcript_name)
final_lncrna_list_rnaspades_length<-length(final_lncrna_list_rnaspades)
final_lncrna_list_trinity_length<-length(final_lncrna_list_trinity)

final_lncrna_list_together_length<-length(final_lncrna_list_rnaspades) + length(final_lncrna_list_trinity)

#numbers table

lncrna_numbers_table<-data.table(category=c("All  genes","RNASpades genes","Trinity genes","100% overlapping fenes", "Longer RNASpades genes with more than 20% overlap","Longer Trinity genes with more than 20% overlap","Longer RNASpades genes with less than 20% overlap","Longer Trinity genes with less than 20% overlap","All genes with more than 20% overlap","All longer genes with less than 20% overlap","Shorter RNASpades genes with more than 20% overlap","Shorter Trinity genes with more than 20% overlap","Shorter RNASpades genes with less than 20% overlap","Shorter Trinity genes with less than 20% overlap","All short genes with more than 20% overlap","All short genes with less than 20% overlap","RNASpades genes without overlaps","Trinity genes without overlap","Final RNASpades genes","Final Trinity genes","Final genes","Final RNASpades transcripts","Final Trinity transcripts","Final transcripts"),number=c(all_together_reduced_num,rnaspades_reduced_length,trinity_reduced_length,same_overlaps_reduced_num,overlap_rnaspades_morethan20_longer_reduced_num,overlap_trinity_morethan20_longer_reduced_num,overlap_rnaspades_lessthan20_longerreduced_num,overlap_trinity_lessthan20_longer_reduced_num,overlap_together_morethan20_longer_reduced_num,overlap_together_lessthan20_longer_reduced_num,overlap_rnaspades_morethan20_shorter_reduced_num,overlap_trinity_morethan20_shorter_reduced_num,overlap_rnaspades_lessthan20_shorter_reduced_num,overlap_trinity_lessthan20_shorter_reduced_num,overlap_together_morethan20_shorter_reduced_num,overlap_together_lessthan20_shorter_reduced_num,rnaspades_unique,trinity_unique,lncrnas_final_reduced_num_rnaspades,lncrnas_final_reduced_num_trinity,lncrnas_final_reduced_num_together,final_lncrna_list_rnaspades_length,final_lncrna_list_trinity_length,final_lncrna_list_together_length))

lncrna_numbers_table

write.table(final_lncrna_list_rnaspades,file="final_lncrna_list_rnaspades.txt",row.names = F,col.names = F,quote = F,sep = "\t")
write.table(final_lncrna_list_trinity,file="final_lncrna_list_trinity.txt",row.names = F,col.names = F,quote = F,sep = "\t")




write.table(lncrna_numbers_table,file="lncrna_numbers_table.txt",row.names = F,quote = F)


lncrna_numbers_table_plot<-lncrna_numbers_table[category%in%c("RNASpades genes","Trinity genes","Longer RNASpades genes with more than 20% overlap","Shorter RNASpades genes with more than 20% overlap","Longer Trinity genes with more than 20% overlap","Shorter Trinity genes with more than 20% overlap","Longer RNASpades genes with less than 20% overlap","Shorter RNASpades genes with less than 20% overlap","Longer Trinity genes with less than 20% overlap","Shorter Trinity genes with less than 20% overlap","RNASpades genes without overlaps","Trinity genes without overlap","Final RNASpades genes","Final Trinity genes")]

lncrna_numbers_table_plot[,category:=c("All genes","All genes","Longer genes (>= 20% overlap)","Longer genes (>= 20% overlap)","Shorter genes (>= 20% overlap)","Shorter genes (>= 20% overlap)","Longer genes (< 20% overlap)","Longer genes (< 20% overlap)","Shorter genes (< 20% overlap)","Shorter genes (< 20% overlap)","Unique genes","Unique genes","Final genes","Final genes")]
lncrna_numbers_table_plot[,category:=factor(category,levels = rev(c("All genes","Unique genes","Longer genes (>= 20% overlap)","Shorter genes (>= 20% overlap)","Longer genes (< 20% overlap)","Shorter genes (< 20% overlap)","Final genes")))]
lncrna_numbers_table_plot[,Transcriptome:=rep(c("rnaSPades","Trinity"),7)]

plot_consensus<-ggplot(data=lncrna_numbers_table_plot,aes(x=category,y=number,fill=Transcriptome)) +
  geom_bar(stat="identity",position = "stack",color="black",width = 0.5) +
  theme_bw() +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(size=8,angle=45,hjust=0.9)) +
  ylab("Number of genes") +
  theme(axis.title.y = element_text(size=8)) +
  theme(axis.title.x = element_blank()) 

ggsave(plot_consensus,filename = "plot_consensus.png",height = 4,width = 6)


figure_one<-ggdraw() +
  draw_plot(lengths_plot, x = 0, y = 0.15, width = .5, height = .7) +
  draw_plot(perc_plot, x = 0.54, y = 0.47, width = .35, height = .3) +
  draw_plot(plot_consensus, x = 0.54, y = 0.11, width = 0.35, height = 0.35) +
  draw_plot_label(label = c("A", "B", "C"), size = 15,
                  x = c(0, 0.52, 0.52), y = c(0.8, 0.8, 0.5))

figure_one

ggsave(figure_one,filename = "figure_one.jpg",height = 10,width = 13)
#mapping analysis


rnaspades_paf_filtered_union<-rnaspades_paf_filtered[query_name%in%final_lncrna_list_rnaspades]
trinity_paf_filtered_union<-trinity_paf_filtered[query_name%in%final_lncrna_list_trinity]
rnaspades_paf_filtered_iso<-rnaspades_paf_filtered[query_name%in%final_lncrna_list_rnaspades]
trinity_paf_filtered_iso<-trinity_paf_filtered[query_name%in%final_lncrna_list_trinity]


rnaspades_paf_filtered_more_iso<-rnaspades_paf_filtered[query_name%in%final_lncrna_list_rnaspades]
trinity_paf_filtered_more_iso<-trinity_paf_filtered[query_name%in%final_lncrna_list_trinity]

isoform_overlap_table_rnaspades<-isoform_overlap_table_rnaspades[gene_tr==gene]
isoform_overlap_table_trinity<-isoform_overlap_table_trinity[gene_tr==gene]
isoform_map_table_rnaspades<-isoform_overlap_table_rnaspades[,.SD,.SDcols=c("transcript_name","gene_id")]
isoform_map_table_trinity<-isoform_overlap_table_trinity[,.SD,.SDcols=c("transcript_name","gene_id")]
isoform_map_table_together<-rbind(isoform_map_table_rnaspades,isoform_map_table_trinity)
isoforms_number_rnaspades<-isoform_overlap_table_rnaspades[,.(number_of_isoforms=.N),by=c("seqnames2","start2","end2","gene")]
isoforms_number_trinity<-isoform_overlap_table_trinity[,.(number_of_isoforms=.N),by=c("seqnames2","start2","end2","gene")]
#lncRNA isoforms
isoforms_number_together_lncrna<-rbind(isoforms_number_rnaspades,isoforms_number_trinity)
isoforms_number_together_lncrna<-isoforms_number_together_lncrna[,.SD,.SDcols="number_of_isoforms"]
isoforms_number_together_lncrna_num<-isoforms_number_together_lncrna[,.N,by=number_of_isoforms]
isoforms_number_together_lncrna_num[,category:="lncRNA"]
isoforms_number_together_lncrna_num[,proportion:=100*N/sum(N)]
isoforms_number_together_lncrna_num_copy<-copy(isoforms_number_together_lncrna_num)
isoforms_number_together_lncrna_num<-isoforms_number_together_lncrna_num[number_of_isoforms<10]
isoforms_number_together_lncrna_num<-rbind(isoforms_number_together_lncrna_num,data.table(number_of_isoforms=">9",N=sum(isoforms_number_together_lncrna_num_copy[number_of_isoforms>9]$N),category="lncRNA",proportion=sum(isoforms_number_together_lncrna_num_copy[number_of_isoforms>9]$proportion)))
#mRNA isoforms
prot_transcript_for_isoforms<-protein_coding_annotation[feature=="transcript"]
prot_transcript_for_isoforms[,gene:=str_remove(attribute,"\\.t\\d+")]
prot_transcript_for_isoforms_table<-prot_transcript_for_isoforms[,.(number_of_isoforms=.N),by=gene]
prot_transcript_for_isoforms_table<-prot_transcript_for_isoforms_table[,.SD,.SDcols="number_of_isoforms"]
prot_transcript_for_isoforms_table_num<-prot_transcript_for_isoforms_table[,.N,by=number_of_isoforms]
prot_transcript_for_isoforms_table_num[,category:="protein_coding_genes"]
prot_transcript_for_isoforms_table_num[,proportion:=100*N/sum(N)]
prot_transcript_for_isoforms_table_num_copy<-copy(prot_transcript_for_isoforms_table_num)
prot_transcript_for_isoforms_table_num<-prot_transcript_for_isoforms_table_num[number_of_isoforms<10]
prot_transcript_for_isoforms_table_num<-rbind(prot_transcript_for_isoforms_table_num,data.table(number_of_isoforms=">9",N=sum(prot_transcript_for_isoforms_table_num_copy[number_of_isoforms>9]$N),category="protein_coding_genes",proportion=sum(prot_transcript_for_isoforms_table_num_copy[number_of_isoforms>9]$proportion)))
isoforms_table_together_num<-rbind(isoforms_number_together_lncrna_num,prot_transcript_for_isoforms_table_num)
isoforms_plot_proportions<-ggplot(data=isoforms_table_together_num,aes(x=number_of_isoforms,y=proportion,fill=category)) +
  geom_bar(col="black",position="dodge",stat="identity",width=0.5,alpha=0.9) +
  theme_bw() +
  xlab("Number of isoforms per gene") +
  ylab("Percentage of genes") +
  scale_x_discrete(limits=c(1:9,">9")) +
  theme(legend.position = "none") +
  scale_fill_manual(values=c("indianred2","steelblue"), name = "Category", labels=c("lncRNA-coding genes","Protein-coding genes")) +
  theme(legend.title = element_text(size=9,face="bold")) 
isoforms_plot_proportions
ggsave(isoforms_plot_proportions,file="isoforms_plot_proportions.png",width = 6,height = 4.2)
#exons and introns
##isoforms with the biggest number of exons per gene
rnaspades_filtered_exons_copy<-copy(rnaspades_filtered_exons)
trinity_filtered_exons_copy<-copy(trinity_filtered_exons)
rnaspades_filtered_exons_numb_per_trans<-rnaspades_filtered_exons_copy[,.N,by=transcript]
trinity_filtered_exons_numb_per_trans<-trinity_filtered_exons_copy[,.N,by=transcript]
setnames(rnaspades_filtered_exons_numb_per_trans,c("transcript_name","number_of_exons"))
setnames(trinity_filtered_exons_numb_per_trans,c("transcript_name","number_of_exons"))
isoform_overlap_table_rnaspades_merged<-merge(isoform_overlap_table_rnaspades,rnaspades_filtered_exons_numb_per_trans,by="transcript_name")
isoform_overlap_table_trinity_merged<-merge(isoform_overlap_table_trinity,trinity_filtered_exons_numb_per_trans,by="transcript_name")  
isoform_overlap_table_rnaspades_merged[,is_max_exons_iso:=ifelse(.SD==max(.SD),T,F),.SDcols="number_of_exons",by=c("seqnames2","start2","end2","gene")]
isoform_overlap_table_rnaspades_merged[is_max_exons_iso==T,max_len:=ifelse(.SD==max(.SD),T,F),.SDcols="transcript_length",by=c("seqnames2","start2","end2","gene")]
isoform_overlap_table_trinity_merged[,is_max_exons_iso:=ifelse(.SD==max(.SD),T,F),.SDcols="number_of_exons",by=c("seqnames2","start2","end2","gene")]
isoform_overlap_table_trinity_merged[is_max_exons_iso==T,max_len:=ifelse(.SD==max(.SD),T,F),.SDcols="transcript_length",by=c("seqnames2","start2","end2","gene")]
isoform_overlap_table_rnaspades_merged<-unique(isoform_overlap_table_rnaspades_merged,by=c("seqnames2","start2","end2","gene","number_of_exons","transcript_length"))
isoform_overlap_table_trinity_merged<-unique(isoform_overlap_table_trinity_merged,by=c("seqnames2","start2","end2","gene","number_of_exons","transcript_length"))
rnaspades_longest_iso<-unique(isoform_overlap_table_rnaspades_merged[max_len==T]$transcript_name)
trinity_longest_iso<-unique(isoform_overlap_table_trinity_merged[max_len==T]$transcript_name)
write.table(rnaspades_longest_iso,file="rnaspades_longest_iso",row.names = F,col.names = F,quote = F,sep = "\t")
write.table(trinity_longest_iso,file="trinity_longest_iso",row.names = F,col.names = F,quote = F,sep = "\t")
rnaspades_paf_filtered_union_copy<-copy(rnaspades_paf_filtered_union)
trinity_paf_filtered_union_copy<-copy(trinity_paf_filtered_union)
rnaspades_paf_filtered_union<-rnaspades_paf_filtered_union[rnaspades_paf_filtered_union$query_name%in%rnaspades_longest_iso]
trinity_paf_filtered_union<-trinity_paf_filtered_union[trinity_paf_filtered_union$query_name%in%trinity_longest_iso]
filtered_together_paf_union<-rbind(rnaspades_paf_filtered_union_copy,trinity_paf_filtered_union_copy)
rnaspades_paf_filtered_union_granges<-makeGRangesFromDataFrame(rnaspades_paf_filtered_union,seqnames.field = "target_name",start.field = "target_start",end.field = "target_end",ignore.strand=T,keep.extra.columns = T)
trinity_paf_filtered_union_granges<-makeGRangesFromDataFrame(trinity_paf_filtered_union,seqnames.field = "target_name",start.field = "target_start",end.field = "target_end",ignore.strand=T,keep.extra.columns = T)
filtered_together_paf_union_granges<-c(rnaspades_paf_filtered_union_granges,trinity_paf_filtered_union_granges)
rnaspades_paf_filtered_union_exons<-rnaspades_filtered_exons[transcript%in%rnaspades_longest_iso]
trinity_paf_filtered_union_exons<-trinity_filtered_exons[transcript%in%trinity_longest_iso]
filtered_together_paf_union_exons<-rbind(rnaspades_paf_filtered_union_exons,trinity_paf_filtered_union_exons)
#only metazoan annotation
filtered_together_paf_union<-rbind(rnaspades_paf_filtered_union,trinity_paf_filtered_union)
protein_coding_annotation_clean2<-copy(protein_coding_annotation)[,1:9]
protein_coding_annotation[,width:=end-start+1]
protein_coding_annotation[feature!="gene",attribute:=str_extract(attribute,'".*t\\d+"')]
protein_coding_annotation[feature!="gene",attribute:=str_remove_all(attribute,'"')]
protein_coding_annotation[feature!="gene",gene:=str_extract(attribute,".*(?=\\.t\\d*)")]
protein_coding_annotation[feature=="gene",gene:=attribute]
##proteins - unique exons
protein_coding_annotation[feature=="CDS" | feature=="transcript",exon_width_sum:=sum(.SD[feature=="CDS"]$width),by=attribute]
protein_coding_annotation[feature=="CDS" | feature=="transcript",number_of_exons:=nrow(.SD)-1,by=attribute]
protein_coding_annotation[feature=="transcript",most_exons_transcript:=ifelse(.SD==max(.SD),T,F),.SDcols="number_of_exons",by=gene]
protein_coding_annotation[most_exons_transcript==T,max_length:=ifelse(.SD==max(.SD),T,F),.SDcols="exon_width_sum",by=gene]
protein_transcripts<-protein_coding_annotation[most_exons_transcript==T]
write.table(protein_transcripts$attribute,file="aug_long_iso.txt",row.names = F,col.names = F,quote = F,sep = "\t")
protein_exons<-protein_coding_annotation[feature=="CDS" & attribute %in% protein_transcripts$attribute]
protein_coding_annotation_clean<-copy(protein_coding_annotation)
#synteny
protein_coding_annotation_clean_ex<-protein_coding_annotation_clean[feature=="CDS"&attribute%in%protein_transcripts$attribute]
protein_coding_annotation_clean_rest<-protein_coding_annotation_clean[feature!="CDS"]
protein_coding_annotation_clean<-rbind(protein_coding_annotation_clean_rest,protein_coding_annotation_clean_ex)[order(attribute)] 
protein_coding_annotation_save<-protein_coding_annotation_clean[feature=="transcript" | feature=="CDS"]
protein_coding_annotation_save[feature=="transcript",feature:="mRNA"]
protein_coding_annotation_save[feature=="CDS",max_length:=T]
protein_coding_annotation_save<-protein_coding_annotation_save[max_length==T]
protein_coding_annotation_save[,seqname:=str_remove_all(seqname,"_")]
protein_coding_annotation_save[,seqname:=paste("Esu",seqname,sep="")]
lncrna_exons_together<-rbind(rnaspades_filtered_exons,trinity_filtered_exons)
lncrna_exons_together<-lncrna_exons_together[transcript%in%filtered_together_paf_union$query_name]
lncrna_gtf_exons<-data.table(seqname=lncrna_exons_together$chromosome,source="AUGUSTUS",feature="exon",start=lncrna_exons_together$start,end=lncrna_exons_together$end,score=0.5,strand_ignore="+",frame=1,attribute=lncrna_exons_together$transcript)
lncrna_gtf_cds<-data.table(seqname=lncrna_exons_together$chromosome,source="AUGUSTUS",feature="CDS",start=lncrna_exons_together$start,end=lncrna_exons_together$end,score=0.5,strand_ignore="+",frame=1,attribute=lncrna_exons_together$transcript)
lncrna_gtf_mrna<-data.table(seqname=filtered_together_paf_union$target_name,source="AUGUSTUS",feature="mRNA",start=filtered_together_paf_union$target_start,end=filtered_together_paf_union$target_end,score=0.5,strand_ignore="+",frame=1,attribute=filtered_together_paf_union$query_name)
lncrna_gtf<-rbind(lncrna_gtf_mrna,lncrna_gtf_cds)[order(attribute)]
lncrna_gtf[,seqname:=str_remove_all(seqname,"_")]
lncrna_gtf[,seqname:=paste("Esu",seqname,sep="")]
final_gtf<-rbind(protein_coding_annotation_save[,1:9],lncrna_gtf)[order(seqname)]
final_gtf[,strand_ignore:="+"]
write.table(final_gtf,file = "final_gtf_Esu.txt",quote = F, row.names = F, col.names = F,sep="\t")
#number of exons per transcript
number_of_exons_table_lncrna<-filtered_together_paf_union_exons[,.(number_of_exons=.N),by=transcript]
number_of_exons_table_lncrna_num<-number_of_exons_table_lncrna[,.N,by=number_of_exons]
for(i in 1:9) {
  if(i > max(number_of_exons_table_lncrna_num$number_of_exons)) {
    number_of_exons_table_lncrna_num<-rbind(number_of_exons_table_lncrna_num,data.table(number_of_exons=i,N=0))
  }
}
number_of_exons_table_lncrna_num[,category:="lncRNA"]
number_of_exons_table_lncrna_num[,proportion:=100*N/sum(N)]
number_of_exons_table_lncrna_num<-rbind(number_of_exons_table_lncrna_num,data.table(number_of_exons=paste(">",9,sep=""),N=sum(number_of_exons_table_lncrna_num[number_of_exons>9]$N),category="lncRNA",proportion=sum(number_of_exons_table_lncrna_num[number_of_exons>9]$proportion)))
protein_exons_table<-protein_exons[,.(number_of_exons=.N),by=attribute][order(-number_of_exons)]
protein_exons_table_num<-protein_exons_table[,.N,by=number_of_exons]
protein_exons_table_num[,category:="protein_coding_gene"]
protein_exons_table_num<-protein_exons_table_num[number_of_exons>1]
protein_exons_table_num[,proportion:=100*N/sum(N)]
protein_exons_table_num_copy<-copy(protein_exons_table_num)
protein_exons_table_num<-protein_exons_table_num[number_of_exons<=9]
protein_exons_table_num<-rbind(protein_exons_table_num,data.table(number_of_exons=paste(">",9,sep=""),N=sum(protein_exons_table_num_copy[number_of_exons>9]$N),category="protein_coding_gene",proportion=sum(protein_exons_table_num_copy[number_of_exons>9]$proportion)))
exon_number_table_together_num<-rbind(number_of_exons_table_lncrna_num,protein_exons_table_num)

exons_plot_proportion<-ggplot(data=exon_number_table_together_num,aes(x=number_of_exons,y=proportion,fill=category)) +
  geom_bar(col="black",position="dodge",stat="identity",width=0.5,alpha=0.9) +
  theme_bw() +
  xlab("Number of exons per gene") +
  ylab("Percentage of genes") +
  theme(legend.position = "none") +
  scale_x_discrete(limits=c(2:9,">9"),drop=F) +
  scale_fill_manual(values=c("indianred2","steelblue"), name = "Category", labels=c("lncRNA-coding genes","Protein-coding genes")) +
  theme(legend.title = element_text(size=9,face="bold")) 
exons_plot_proportion
ggsave(exons_plot_proportion,file="exons_plot_proportion.png",width=6, height = 4.2)



#lncRNA and protein exon width comparison
protein_exons<-makeGRangesFromDataFrame(protein_exons,keep.extra.columns = T)
filtered_together_paf_union_exons[,width:=end-start+1]
exon_widths_table<-data.table(width=c(filtered_together_paf_union_exons$width,width(protein_exons)),category=c(rep("lncRNA",nrow(filtered_together_paf_union_exons)),rep("protein_coding_genes",length(protein_exons))),feature=rep("Exons",nrow(filtered_together_paf_union_exons)+length(protein_exons)))
wilcox.test(exon_widths_table[category=="lncRNA"]$width,exon_widths_table[category=="protein_coding_genes"]$width)
#lncRNA introns
rnaspades_filtered_exons_granges<-makeGRangesFromDataFrame(rnaspades_paf_filtered_union_exons,seqnames.field = "chromosome",keep.extra.columns = T)
trinity_filtered_exons_granges<-makeGRangesFromDataFrame(trinity_paf_filtered_union_exons,seqnames.field = "chromosome",keep.extra.columns = T)
together_filtered_exons_granges<-c(rnaspades_filtered_exons_granges,trinity_filtered_exons_granges)
rnaspades_filtered_exons_granges_split<-split(rnaspades_filtered_exons_granges,rnaspades_filtered_exons_granges$transcript)
trinity_filtered_exons_granges_split<-split(trinity_filtered_exons_granges,trinity_filtered_exons_granges$transcript)
rnaspades_paf_filtered_union_granges_split<-split(rnaspades_paf_filtered_union_granges,rnaspades_paf_filtered_union_granges$query_name)
trinity_paf_filtered_union_granges_split<-split(trinity_paf_filtered_union_granges,trinity_paf_filtered_union_granges$query_name)
rnaspades_filtered_introns<-unlist(GenomicRanges::setdiff(rnaspades_paf_filtered_union_granges_split,rnaspades_filtered_exons_granges_split))
trinity_filtered_introns<-unlist(GenomicRanges::setdiff(trinity_paf_filtered_union_granges_split,trinity_filtered_exons_granges_split))
lncrna_filtered_introns<-c(rnaspades_filtered_introns,trinity_filtered_introns)
lncrna_filtered_introns$transcript<-names(lncrna_filtered_introns)
#protein introns
protein_exons_split<-split(protein_exons,protein_exons$attribute)
protein_transcripts_granges<-makeGRangesFromDataFrame(protein_transcripts,ignore.strand = T,keep.extra.columns = T)
protein_transcripts_granges_split<-split(protein_transcripts_granges,protein_transcripts_granges$attribute)
protein_transcripts_granges_split<-protein_transcripts_granges_split[names(protein_transcripts_granges_split)%in%names(protein_exons_split)]
protein_introns<-unlist(GenomicRanges::setdiff(protein_transcripts_granges_split,protein_exons_split))
protein_introns$transcript<-names(protein_introns)
#lncRNA and protein intron comparison
intron_width_table<-data.table(width=c(width(lncrna_filtered_introns),width(protein_introns)),category=c(rep("lncRNA",length(lncrna_filtered_introns)),rep("protein_coding_genes",length(protein_introns))),feature=rep("Introns",length(lncrna_filtered_introns)+length(protein_introns)))
wilcox.test(intron_width_table[category=="lncRNA"]$width,intron_width_table[category=="protein_coding_genes"]$width)
#transcript lengths (spliced!)
esu_genome<-readDNAStringSet("ESU_pilon_pathPol.fasta")
#only metazoan scaffolds
esu_genome<-esu_genome[names(esu_genome)%in%taxonomy_table_nonbact]
lncrna_intron_sequences<-getSeq(esu_genome,lncrna_filtered_introns)
lncrna_exons_together_gr<-makeGRangesFromDataFrame(df = lncrna_exons_together,seqnames.field = "chromosome")
lncrna_exon_sequences<-getSeq(esu_genome,lncrna_exons_together_gr)
protein_exons_sequences<-getSeq(esu_genome,protein_exons)
names(protein_exons_sequences)<-protein_exons$attribute
exon_table<-as.data.table(protein_exons_sequences)
exon_table[,name:=names(protein_exons_sequences)]
exon_table[,transcripts:=lapply(.SD, paste0, collapse=""), by = name]
exon_table_unique<-unique(exon_table,by="transcripts")
protein_transcript_sequences<-DNAStringSet(exon_table_unique$transcripts)
names(protein_transcript_sequences)<-exon_table_unique$name
protein_transcript_lengths<-sapply(protein_transcript_sequences,length)
#protein_transcript_lengths<-protein_transcript_lengths[protein_transcript_lengths>=100]
lncrna_transcript_lengths<-filtered_together_paf_union_granges$query_length
transcript_length_table<-data.table(width=c(lncrna_transcript_lengths,protein_transcript_lengths),category=c(rep("lncRNA",length(lncrna_transcript_lengths)),rep("protein_coding_genes",length(protein_transcript_lengths))),feature="Transcripts")
wilcox.test(transcript_length_table[category=="lncRNA"]$width,transcript_length_table[category=="protein_coding_genes"]$width)
#gene length
protein_coding_genes<-protein_coding_annotation[feature=="gene"]
write.table(protein_coding_genes,file="protein_coding_genes.gtf",quote = F,col.names = F,row.names = F,sep = "\t")
protein_coding_genes_granges<-makeGRangesFromDataFrame(protein_coding_genes,ignore.strand = T,keep.extra.columns = T)
lncrna_gene_lengths<-c(width(rnaspades_paf_filtered_granges_reduced),width(trinity_paf_filtered_granges_reduced))
protein_coding_gene_lengths<-width(protein_coding_genes_granges)
gene_length_table<-data.table(width=c(lncrna_gene_lengths,protein_coding_gene_lengths),category=c(rep("lncRNA",length(lncrna_gene_lengths)),rep("protein_coding_genes",length(protein_coding_gene_lengths))),feature="Genes")
wilcox.test(gene_length_table[category=="lncRNA"]$width,gene_length_table[category=="protein_coding_genes"]$width)  
#Element lengths
element_width_table<-rbind(exon_widths_table,intron_width_table,transcript_length_table,gene_length_table)
element_width_table[,feature:=factor(feature,levels = c("Exons","Introns","Transcripts","Genes"))]

features_comparison<-list(c("lncRNA","protein_coding_genes"))
element_length_boxplot<-ggplot(data=element_width_table,aes(x=category,y=width,fill=category)) +
  geom_boxplot(alpha=0.9) +
  facet_grid(.~feature) +
  scale_y_log10() +
  theme_bw() +
  scale_fill_manual(values=c("indianred2","steelblue"),name="Category",label=c("LncRNA-coding genes","Protein-coding genes")) +
  theme(legend.position = "bottom") +
  theme(legend.title = element_text(size=9,face="bold")) +
  theme(axis.title.x=element_blank()) +
  theme(axis.text.x =element_blank()) +
  ylab("Length / bp (log10 scaled)")  +
  stat_compare_means(comparisons = features_comparison,label = "p.signif")
element_length_boxplot 
ggsave(element_length_boxplot,file="element_length_boxplot.png",width=6,height = 4) 


figure_two<-ggdraw() +
  draw_plot(isoforms_plot_proportions, x = 0.109, y = 0.5, width = .36, height = .38) +
  draw_plot(exons_plot_proportion, x = 0.5, y = 0.5, width = .36, height = .38) +
  draw_plot(element_length_boxplot, x = 0.163, y = 0.02, width = 0.65, height = 0.45) +
  draw_plot_label(label = c("A", "B", "C"), size = 15,
                  x = c(0.1, 0.52, 0.18), y = c(0.9, 0.9, 0.5))

figure_two

ggsave(figure_two,filename = "figure_two.jpg",height = 10,width = 13)


#splice sites
lncrna_introns_first_two_bases<-subseq(lncrna_intron_sequences,start=1,end=2)
lncrna_introns_last_two_bases<-reverse(subseq(reverse(lncrna_intron_sequences),start=1,end=2))
splice_sites_table_lncrna<-data.table(first_two=as.character(lncrna_introns_first_two_bases),last_two=as.character(lncrna_introns_last_two_bases))
splice_sites_table_total_lncrna<-splice_sites_table_lncrna[,.N,by=c("first_two","last_two")]
splice_sites_table_total_lncrna<-splice_sites_table_total_lncrna[first_two!="NN" & last_two!="NN"]
all_dinucleotides<-colnames(dinucleotideFrequency(esu_genome))
all_dinucleotide_combinations<-as.data.table(expand.grid(all_dinucleotides,all_dinucleotides))
setnames(all_dinucleotide_combinations,c("first_two","last_two"))
missing_dinucleotides_lncrna<-anti_join(all_dinucleotide_combinations,splice_sites_table_total_lncrna[,.SD,.SDcols=1:2])
zeros_lncrna<-data.table(N=rep(0,nrow(missing_dinucleotides_lncrna)))
missing_dinucleotides_frequencies_lncrna<-cbind(missing_dinucleotides_lncrna,zeros_lncrna)
splice_sites_table_total_lncrna<-rbind(splice_sites_table_total_lncrna,missing_dinucleotides_frequencies_lncrna)
splice_sites_table_total_lncrna[,category:="lncRNA"]
splice_sites_table_total_lncrna[,proportion:=N/sum(N)]
protein_intron_sequences<-getSeq(esu_genome,protein_introns)
protein_introns_first_two_bases<-subseq(protein_intron_sequences,start=1,end=2)
protein_introns_last_two_bases<-reverse(subseq(reverse(protein_intron_sequences),start=1,end=2))
splice_sites_table_protein<-data.table(first_two=as.character(protein_introns_first_two_bases),last_two=as.character(protein_introns_last_two_bases))
splice_sites_table_total_protein<-splice_sites_table_protein[,.N,by=c("first_two","last_two")]
splice_sites_table_total_protein<-splice_sites_table_total_protein[first_two!="NN" & last_two!="NN"]
missing_dinucleotides_protein<-anti_join(all_dinucleotide_combinations,splice_sites_table_total_protein[,.SD,.SDcols=1:2])
zeros_protein<-data.table(N=rep(0,nrow(missing_dinucleotides_protein)))
missing_dinucleotides_frequencies_protein<-cbind(missing_dinucleotides_protein,zeros_protein)
splice_sites_table_total_protein<-rbind(splice_sites_table_total_protein,missing_dinucleotides_frequencies_protein)
splice_sites_table_total_protein[,category:="protein_coding_genes"]
splice_sites_table_total_protein[,proportion:=N/sum(N)]
splice_sites_table_total_together<-rbind(splice_sites_table_total_lncrna,splice_sites_table_total_protein)
splice_sites_table_total_together[,category:=ifelse(category=="lncRNA","LncRNA genes","Protein-coding genes")]
heatmap_splice_sites<-ggplot(data=splice_sites_table_total_together,aes(x=first_two,y=last_two,fill=proportion)) +
  geom_tile(col="black")  +
  facet_wrap(.~category) +
  scale_fill_gradient(low="white",high="indianred2",trans="sqrt",na.value = "white",breaks=c(0,0.1,0.2,0.3,0.5))  +
  xlab("5' splice site / 3* RC splice site") +
  ylab("3' splice site / 5' RC splice site") +
  labs(fill = "Intron ratio (sqrt scaled)") +
  theme(legend.position="top") +
  theme(legend.key.size = unit(0.5, "cm"),
        legend.key.width = unit(0.6,"cm")) +
  theme(legend.title = element_text( size = 8)) +
  theme(axis.title.x = element_text(size=8)) +
  theme(axis.title.y = element_text(size=8)) +
  theme(text = element_text(size=6)) +
  theme(legend.title = element_text(size=9,face="bold")) 
heatmap_splice_sites
ggsave(heatmap_splice_sites,file="heatmap_splice_sites.png")
#mapping locations (genes)
#intronic
protein_coding_annotation_clean2[feature!="gene",attribute:=str_extract(attribute,'".*t\\d+"')]
protein_coding_annotation_clean2[feature!="gene",attribute:=str_remove_all(attribute,'"')]
protein_coding_annotation_exons_clean<-protein_coding_annotation_clean2[feature=="CDS"]
protein_coding_annotation_transcripts_clean<-protein_coding_annotation[feature=="transcript"]
protein_exons_clean<-makeGRangesFromDataFrame(protein_coding_annotation_exons_clean,keep.extra.columns = T)
protein_exons_clean_split<-split(protein_exons_clean,protein_exons_clean$attribute)
protein_transcripts_clean_granges<-makeGRangesFromDataFrame(protein_coding_annotation_transcripts_clean,ignore.strand = T,keep.extra.columns = T)
protein_transcripts_clean_granges_split<-split(protein_transcripts_clean_granges,protein_transcripts_clean_granges$attribute)
protein_exons_clean_split<-protein_exons_clean_split[names(protein_exons_clean_split) %in% names(protein_transcripts_clean_granges_split)]
protein_introns_clean<-unlist(GenomicRanges::setdiff(protein_transcripts_clean_granges_split,protein_exons_clean_split))
protein_introns_clean$transcript<-names(protein_introns_clean)
protein_introns<-protein_introns_clean
protein_exons<-protein_exons_clean
protein_coding_genes_granges<-makeGRangesFromDataFrame(protein_coding_annotation_clean2[feature=="gene"],keep.extra.columns = T)
overlaps_lncRNA_prot_within<-findOverlaps(filtered_together_paf_union_granges,protein_coding_genes_granges,ignore.strand=T,type = "within")
overlaps_lncRNA_prot_within_table<-cbind(as.data.table(filtered_together_paf_union_granges[queryHits(overlaps_lncRNA_prot_within)]),as.data.table(protein_coding_genes_granges[subjectHits(overlaps_lncRNA_prot_within)]))
setnames(overlaps_lncRNA_prot_within_table,c("seqnames1","start1","end1","width1","strand1","query_name","query_length","query_start","query_end","strand_ignore22","target_length","matches","alignment_block","mapping_quality","perc_matches","gene","seqnames2","start2","end2","width2","strand2","source","feature","score","strand_ignore2","frame","attribute"))
intronic_lncrna<-unique(overlaps_lncRNA_prot_within_table$query_name)
#overlapping
introns_table<-as.data.table(lncrna_filtered_introns)
introns_table[,lncrna_transcript:=names(lncrna_filtered_introns)]
overlaps_prot_lncrna_within<-findOverlaps(protein_coding_genes_granges,filtered_together_paf_union_granges,ignore.strand=T,type="within")
overlaps_prot_lncrna_within_table<-cbind(as.data.table(protein_coding_genes_granges[queryHits(overlaps_prot_lncrna_within)]),as.data.table(filtered_together_paf_union_granges[subjectHits(overlaps_prot_lncrna_within)]))
setnames(overlaps_prot_lncrna_within_table,c("seqnames1","start1","end1","width1","strand1","source","feature","score","strand_ignore1","frame","attribute","seqnames2","start2","end2","width2","strand2","query_name","query_length","query_start","query_end","strand_ignore22","target_length","matches","alignment_block","mapping_quality","perc_matches","gene"))
overlapping_lncrna<-unique(overlaps_prot_lncrna_within_table$query_name)
intronic_lncrna<-intronic_lncrna[intronic_lncrna%in%intersect(overlapping_lncrna,intronic_lncrna)==F]
#intergenic
lncrna_prot_transcripts_overlaps<-findOverlaps(filtered_together_paf_union_granges,protein_coding_genes_granges,ignore.strand=T)
lncrna_prot_transcripts_overlaps_table<-cbind(as.data.table(filtered_together_paf_union_granges[queryHits(lncrna_prot_transcripts_overlaps)]),as.data.table(protein_coding_genes_granges[subjectHits(lncrna_prot_transcripts_overlaps)]))
setnames(lncrna_prot_transcripts_overlaps_table,c("seqnames1","start1","end1","width1","strand1","query_name","query_length","query_start","query_end","strand_ignore","target_length","matches","aligmnent_block","mapping_quality","perc_matches","gene","seqnames2","start2","end2","width2","strand2","source","feature","score","strand_ignore2","frame","attribute"))
intergenic_lncrna<-unique(filtered_together_paf_union[query_name %in% lncrna_prot_transcripts_overlaps_table$query_name==F]$query_name)
 
#intron exon overlap
int_ex_overlap_candidates_intronic<-setdiff(filtered_together_paf_union_granges$query_name,intronic_lncrna)
int_ex_overlap_candidates_intronic_overlapping<-setdiff(int_ex_overlap_candidates_intronic,overlapping_lncrna)
int_ex_overlap_candidates_intronic_overlapping_intergenic<-setdiff(int_ex_overlap_candidates_intronic_overlapping,intergenic_lncrna)
int_ex_overlap_candidates_exons<-makeGRangesFromDataFrame(filtered_together_paf_union_exons[filtered_together_paf_union_exons$transcript %in% int_ex_overlap_candidates_intronic_overlapping_intergenic],seqnames.field = "chromosome",ignore.strand=T,keep.extra.columns = T)
overlaps_candidates_exons_prot_introns<-findOverlaps(int_ex_overlap_candidates_exons,protein_introns,ignore.strand=T)
overlaps_candidates_exons_prot_introns_table<-cbind(as.data.table(int_ex_overlap_candidates_exons[queryHits(overlaps_candidates_exons_prot_introns)]),as.data.table(protein_introns[subjectHits(overlaps_candidates_exons_prot_introns)]))
setnames(overlaps_candidates_exons_prot_introns_table,c("seqnames1","start1","end1","width1","strand1","transcript","seqnames2","start2","end2","width2","strand2","transcript_int"))
overlapping_lncrna<-c(overlapping_lncrna,overlaps_candidates_exons_prot_introns_table$transcript)
#transcript number per class
lncRNA_classification_table<-unique(data.table(transcript=c(intronic_lncrna,overlapping_lncrna,intergenic_lncrna),class=c(rep("intronic_lncRNA",length(intronic_lncrna)),rep("overlapping_lncRNA",length(overlapping_lncrna)),rep("intergenic_lncRNA",length(intergenic_lncrna)))))
number_of_lncrna_classes<-lncRNA_classification_table[,.(number_of_lncRNA=.N),by=class]
  
number_of_lncrna_classes[,class:=factor(class,levels = c("intronic_lncRNA","intergenic_lncRNA","overlapping_lncRNA"))]
number_of_lncrna_classes_plot<-ggplot(data=number_of_lncrna_classes,aes(x=class,y=number_of_lncRNA,fill=class)) +
          geom_bar(color="black",width=0.5,stat="identity") +
  theme_bw() +
  theme(legend.position = "none") + 
    scale_fill_hp(discrete = TRUE, option = "NewtScamander")  +
  theme(axis.title.x=element_blank()) +
  ylab("Number of genes") +
   theme(axis.text.x = element_text(angle=45,hjust=0.9)) +
  scale_x_discrete(labels=c("Intronic lncRNA","Intergenic lncRNA","Overlapping lncRNA"))

  number_of_lncrna_classes_plot
  
  ggsave(number_of_lncrna_classes_plot,filename = "number_of_lncrna_classes_plot.jpg",height = 4,width = 6)
  
    
  
all_rnaspades_sequences<-readDNAStringSet("transcripts.fasta")
all_trinity_sequences<-readDNAStringSet("Trinity.fasta")
names(all_trinity_sequences)<-str_extract(names(all_trinity_sequences),"TRINITY_.*_i\\d*")
lncrna_rnaspades_sequences<-all_rnaspades_sequences[names(all_rnaspades_sequences)%in%rnaspades_longest_iso]
lncrna_trinity_sequences<-all_trinity_sequences[names(all_trinity_sequences)%in%trinity_longest_iso]
lncrna_together_sequences<-c(lncrna_rnaspades_sequences,lncrna_trinity_sequences)
#transcript lengths by class
lncRNA_classification_table_name<-copy(lncRNA_classification_table)
setnames(lncRNA_classification_table_name,c("query_name","class"))
lncrna_paf_with_classification<-merge(filtered_together_paf_union,lncRNA_classification_table_name,by="query_name")
class_transcripts_table<-data.table(element_length=c(lncrna_paf_with_classification[class=="intergenic_lncRNA"]$query_length,lncrna_paf_with_classification[class=="intronic_lncRNA"]$query_length,lncrna_paf_with_classification[class=="overlapping_lncRNA"]$query_length),class=c(rep("intergenic_lncRNA",length(lncrna_paf_with_classification[class=="intergenic_lncRNA"]$query_length)),rep("intronic_lncRNA",length(lncrna_paf_with_classification[class=="intronic_lncRNA"]$query_length)),rep("overlapping_lncRNA",length(lncrna_paf_with_classification[class=="overlapping_lncRNA"]$query_length))))
class_transcripts_table[,class:=factor(class,levels=c("intergenic_lncRNA","intronic_lncRNA","overlapping_lncRNA"))]
class_transcripts_table[,feature:="Transcripts"]
#class exons
filtered_together_paf_union_exons_class<-merge(filtered_together_paf_union_exons,lncRNA_classification_table,by="transcript")
class_exons_table<-filtered_together_paf_union_exons_class[,.SD,.SDcols=c("width","class")]
class_exons_table[,feature:="Exons"]
setnames(class_exons_table,c("element_length","class","feature"))
#class introns
filtered_together_paf_union_introns_class<-merge(introns_table,lncRNA_classification_table,by="transcript")
class_introns_table<-filtered_together_paf_union_introns_class[,.SD,.SDcols=c("width","class")]
class_introns_table[,feature:="Introns"]
setnames(class_introns_table,c("element_length","class","feature"))
#class genes
setnames(lncRNA_classification_table,c("transcript_name","class"))
overlaps_table_rnaspades_with_class<-merge(isoform_overlap_table_rnaspades,lncRNA_classification_table,by="transcript_name")
overlaps_table_trinity_with_class<-merge(isoform_overlap_table_trinity,lncRNA_classification_table,by="transcript_name")
class_genes_rnaspades<-unique(overlaps_table_rnaspades_with_class[,.SD,.SDcols=c("seqnames1","start1","end1","width1","class")])
class_genes_trinity<-unique(overlaps_table_trinity_with_class[,.SD,.SDcols=c("seqnames1","start1","end1","width1","class")])
class_genes_table<-rbind(class_genes_rnaspades,class_genes_trinity)
class_genes_table<-class_genes_table[,.SD,.SDcols=c("width1","class")]
setnames(class_genes_table,c("element_length","class"))
class_genes_table[,feature:="Genes"]
class_together_table<-rbind(class_transcripts_table,class_exons_table,class_introns_table,class_genes_table)
class_together_table<-class_together_table[,feature:=factor(feature,c("Exons","Introns","Transcripts","Genes"))]
class_together_table[,class:=factor(class,levels = c("intronic_lncRNA","intergenic_lncRNA","overlapping_lncRNA"))]

class_comparison<-list(c("intronic_lncRNA","intergenic_lncRNA"),c("intronic_lncRNA","overlapping_lncRNA"),c("intergenic_lncRNA","overlapping_lncRNA"))
class_lengths_plot<-ggplot(data=class_together_table,aes(x=class,y=element_length,fill=class)) +
  geom_boxplot(alpha=0.8) +
  scale_y_log10() +
  theme_bw() +
  facet_grid(.~feature) +
  theme(legend.position = "bottom") +
  theme(legend.title = element_text(size=9,face="bold")) +
  theme(axis.title.x=element_blank()) +
  theme(axis.text.x =element_blank()) +
  theme(axis.title.y   = element_text(size=9)) +
  scale_fill_hp(discrete = TRUE, option = "NewtScamander", name = "Class",labels=c("Intronic lncRNA","Intergenic lncRNA","Overlapping lncRNA")) +
  ylab("Length / bp (log10 scaled)")  +
  stat_compare_means(comparisons = class_comparison,label = "p.signif")
  



class_lengths_plot
ggsave(class_lengths_plot,file="class_lengths_plot.png",width=6,height = 4)



#genesum of bases by class
class_genes_table
class_genes_table_sum_width<-class_genes_table[,.(bases_sum=sum(element_length)),by=class]

class_genes_table_sum_width[,bases_sum_kb:=bases_sum/1000]
class_genes_table_sum_width[,class:=factor(class,levels = c("intronic_lncRNA","intergenic_lncRNA","overlapping_lncRNA"))]
class_genes_table_sum_width_plot<-ggplot(data=class_genes_table_sum_width,aes(x=class,y=bases_sum_kb,fill=class)) +
          geom_bar(color="black",width=0.5,stat="identity") +
  theme_bw() +
  theme(legend.position = "none") + 
    scale_fill_hp(discrete = TRUE, option = "NewtScamander")  +
  theme(axis.title.x=element_blank()) +
  ylab("Total length / kb") +
   theme(axis.text.x = element_text(angle=45,hjust=0.9)) +
  scale_x_discrete(labels=c("Intronic lncRNA","Intergenic lncRNA","Overlapping lncRNA"))

  class_genes_table_sum_width_plot
  
  ggsave(class_genes_table_sum_width_plot,filename = "class_genes_table_sum_width_plot.jpg",height = 4,width = 6)
  
  
#intron sum

class_introns_table_sum_width<-class_introns_table[,.(intron_sum=sum(element_length)),by=class]

#exon sum


class_exons_table_sum_widthh<-class_exons_table[,.(exon_sum=sum(element_length)),by=class]

class_table_lengths<-merge(class_genes_table_sum_width,class_exons_table_sum_widthh,by="class")
class_table_lengths<-merge(class_table_lengths,class_introns_table_sum_width,by="class")

class_table_lengths[,intron_percentage:=100*intron_sum/bases_sum]

kruskal.test(x = c(class_exons_table[class=="intergenic_lncRNA"]$element_length,class_exons_table[class=="intronic_lncRNA"]$element_length,class_exons_table[class=="overlapping_lncRNA"]$element_length),g=c(rep(1,length(class_exons_table[class=="intergenic_lncRNA"]$element_length)),rep(2,length(class_exons_table[class=="intronic_lncRNA"]$element_length)),rep(3,length(class_exons_table[class=="overlapping_lncRNA"]$element_length))))

kruskal.test(x = c(class_introns_table[class=="intergenic_lncRNA"]$element_length,class_introns_table[class=="intronic_lncRNA"]$element_length,class_introns_table[class=="overlapping_lncRNA"]$element_length),g=c(rep(1,length(class_introns_table[class=="intergenic_lncRNA"]$element_length)),rep(2,length(class_introns_table[class=="intronic_lncRNA"]$element_length)),rep(3,length(class_introns_table[class=="overlapping_lncRNA"]$element_length))))

kruskal.test(x = c(class_genes_table[class=="intergenic_lncRNA"]$element_length,class_genes_table[class=="intronic_lncRNA"]$element_length,class_genes_table[class=="overlapping_lncRNA"]$element_length),g=c(rep(1,length(class_genes_table[class=="intergenic_lncRNA"]$element_length)),rep(2,length(class_genes_table[class=="intronic_lncRNA"]$element_length)),rep(3,length(class_genes_table[class=="overlapping_lncRNA"]$element_length))))

kruskal.test(x = c(class_transcripts_table[class=="intronic_lncRNA"]$element_length,class_transcripts_table[class=="intergenic_lncRNA"]$element_length,class_transcripts_table[class=="overlapping_lncRNA"]$element_length),g=c(rep(1,length(class_transcripts_table[class=="intronic_lncRNA"]$element_length)),rep(2,length(class_transcripts_table[class=="intergenic_lncRNA"]$element_length)),rep(3,length(class_transcripts_table[class=="overlapping_lncRNA"]$element_length))))


figure_three<-ggdraw() +
  draw_plot(number_of_lncrna_classes_plot, x = 0.109, y = 0.5, width = .36, height = .38) +
  draw_plot(class_genes_table_sum_width_plot, x = 0.5, y = 0.5, width = .36, height = .38) +
  draw_plot(class_lengths_plot, x = 0.13, y = 0.03, width = 0.7, height = 0.45) +
  draw_plot_label(label = c("A", "B", "C"), size = 15,
                  x = c(0.1, 0.5, 0.15), y = c(0.9, 0.9, 0.5))

figure_three

ggsave(figure_three,filename = "figure_three.jpg",height = 10,width = 10)



#intergenic lncRNA - distance to genes
lncrnas_final_reduced_together_dt<-as.data.table(lncrnas_final_reduced_together)

class_gene_together<-unique(rbind(class_genes_rnaspades,class_genes_trinity))
setnames(lncrnas_final_reduced_together_dt,c("seqnames1","start1","end1","width1","strand1","gene","gene_id"))
lncrnas_final_reduced_together_dt_merged<-merge(lncrnas_final_reduced_together_dt,class_gene_together,by=c("seqnames1","start1","end1","width1"),all.x=T)
intergenic_lncrna_dt<-lncrnas_final_reduced_together_dt_merged[class=="intergenic_lncRNA"]
intergenic_lncrna_granges<-makeGRangesFromDataFrame(intergenic_lncrna_dt,seqnames.field = "seqnames1",start.field = "start1",end.field = "end1",keep.extra.columns = T)
intergenic_nearest_gene<-nearest(intergenic_lncrna_granges,protein_coding_genes_granges)
prot_cod_intergenic_nearest<-protein_coding_genes[intergenic_nearest_gene]


intergenic_prot_nearest_table<-data.table(seqnames1=as.character(seqnames(intergenic_lncrna_granges)),start1=start(intergenic_lncrna_granges),end1=end(intergenic_lncrna_granges),lncrna=intergenic_lncrna_granges$gene_id,seqnames2=prot_cod_intergenic_nearest$seqname,start2=prot_cod_intergenic_nearest$start,end2=prot_cod_intergenic_nearest$end,prot_cod_gene=prot_cod_intergenic_nearest$attribute)

inter_on_contigs_without_genes<-which(is.na(intergenic_prot_nearest_table$seqnames2))
length(inter_on_contigs_without_genes)

intergenic_prot_nearest_table<-intergenic_prot_nearest_table[is.na(seqnames2)==F]
intergenic_prot_nearest_table[,distance:=ifelse(start1>start2,start1-end2-1,end1-start2+1)]

inter_prot_cod_distance_plot<-ggplot(data=intergenic_prot_nearest_table,aes(x=distance)) +
  geom_histogram(color="black",fill="skyblue",binwidth = 250) +
  theme_bw()  +
  xlab("Distance to nearest protein-coding gene / bp") +
  ylab("Number of lncRNA-coding genes") +
  coord_cartesian(xlim=c(-10000,10000)) +
  theme(axis.title.x   = element_text(size=9)) +
  theme(axis.title.y   = element_text(size=9))  

inter_prot_cod_distance_plot
  
ggsave(inter_prot_cod_distance_plot,file="inter_prot_cod_distance_plot.png",width = 5,height = 3.8)


#overlapping proteins vs non-overlapping proteins

#intronic


intronic_prots<-unique(overlaps_lncRNA_prot_within_table$attribute)

#overlapping

overlapping_prots<-unique(overlaps_prot_lncrna_within_table$attribute)


#intergenic - 1 kb

intergenic_prots<-unique(intergenic_prot_nearest_table[distance<=1000 & distance >= -1000]$prot_cod_gene)

category_prots<-c(intronic_prots,overlapping_prots,intergenic_prots)
no_category_prots<-setdiff(protein_coding_genes$attribute,category_prots)

prot_classification_table<-data.table(gene=c(category_prots,no_category_prots),lncrna_class=c(rep("intronic",length(intronic_prots)),rep("overlapping",length(overlapping_prots)),rep("intergenic_1kb",length(intergenic_prots)),rep("no_lncRNA",length(no_category_prots))))

protein_coding_annotation_longest_isoforms_copy<-copy(protein_transcripts)
protein_coding_annotation_longest_isoforms_copy[,gene:=str_extract(attribute,".*(?=\\.t\\d*)")]
prot_lncrna_overlaps_genes<-protein_coding_annotation_longest_isoforms_copy[gene%in%prot_classification_table[lncrna_class!="no_lncRNA"]$gene]$attribute

prot_lncrna_intronic_overlaps_genes<-protein_coding_annotation_longest_isoforms_copy[gene%in%prot_classification_table[lncrna_class=="intronic"]$gene]$attribute
prot_lncrna_overlapping_overlaps_genes<-protein_coding_annotation_longest_isoforms_copy[gene%in%prot_classification_table[lncrna_class=="overlapping"]$gene]$attribute
prot_lncrna_intergenic_1kb_overlaps_genes<-protein_coding_annotation_longest_isoforms_copy[gene%in%prot_classification_table[lncrna_class=="intergenic_1kb"]$gene]$attribute

write.table(prot_lncrna_overlaps_genes,file="prot_lncrna_overlaps_genes.txt",row.names = F,col.names = F,quote = F,sep = "\t")
write.table(prot_lncrna_intronic_overlaps_genes,file="prot_lncrna_intronic_overlaps_genes.txt",row.names = F,col.names = F,quote = F,sep = "\t")
write.table(prot_lncrna_overlapping_overlaps_genes,file="prot_lncrna_overlapping_overlaps_genes.txt",row.names = F,col.names = F,quote = F,sep = "\t")
write.table(prot_lncrna_intergenic_1kb_overlaps_genes,file="prot_lncrna_intergenic_1kb_overlaps_genes.txt",row.names = F,col.names = F,quote = F,sep = "\t")

lncRNA_classification_table_with_gene_id<-merge(lncRNA_classification_table,isoform_map_table_together)
#GC content

genome_gc<-100*(alphabetFrequency(unlist(esu_genome),as.prob = T)["C"]+alphabetFrequency(unlist(esu_genome),as.prob = T)["G"])
prot_gc<-100*(alphabetFrequency(protein_transcript_sequences,as.prob = T)[,"C"]+alphabetFrequency(protein_transcript_sequences,as.prob = T)[,"G"])
lncrna_gc<-100*(alphabetFrequency(lncrna_together_sequences,as.prob = T)[,"C"]+alphabetFrequency(lncrna_together_sequences,as.prob = T)[,"G"])

gc_table<-data.table(transcript_gc=c(lncrna_gc,prot_gc),category=c(rep("lncRNA",length(lncrna_gc)),rep("protein_coding_transcripts",length(prot_gc))))


gc_boxplot<-ggplot(data=gc_table,aes(x=category,y=transcript_gc,fill=category)) +
  geom_boxplot(width=0.5,alpha=0.8) +
  theme_bw() +
  theme(legend.position = "none") +
    scale_fill_manual(values=c("indianred2","steelblue")) +
  theme(axis.title.x = element_blank()) +
  ylab("GC content") +
    scale_x_discrete(labels=c("lncRNA","mRNA"))


gc_boxplot

ggsave(gc_boxplot,file="gc_boxplot.jpg",width = 6,height = 4.6)

t.test(lncrna_gc,prot_gc)  



exons_clean_lncrna_rnaspades<-rnaspades_filtered_exons
exons_clean_lncrna_trinity<-trinity_filtered_exons
exons_clean_lncrna_together<-rbind(exons_clean_lncrna_rnaspades,exons_clean_lncrna_trinity)
rnaspades_paf_filtered_iso<-makeGRangesFromDataFrame(rnaspades_paf_filtered_iso,keep.extra.columns = T,seqnames.field = "target_name",start.field = "target_start",end.field = "target_end")
trinity_paf_filtered_iso<-makeGRangesFromDataFrame(trinity_paf_filtered_iso,keep.extra.columns = T,seqnames.field = "target_name",start.field = "target_start",end.field = "target_end")

rnaspades_paf_filtered_iso_split<-split(rnaspades_paf_filtered_iso,rnaspades_paf_filtered_iso$query_name)
trinity_paf_filtered_iso_split<-split(trinity_paf_filtered_iso,trinity_paf_filtered_iso$query_name)

exons_clean_lncrna_rnaspades_gr<-makeGRangesFromDataFrame(exons_clean_lncrna_rnaspades,keep.extra.columns = T)
exons_clean_lncrna_rnaspades_gr<-exons_clean_lncrna_rnaspades_gr[exons_clean_lncrna_rnaspades_gr$transcript%in%rnaspades_paf_filtered_iso$query_name]
exons_clean_lncrna_rnaspades_gr_split<-split(exons_clean_lncrna_rnaspades_gr,exons_clean_lncrna_rnaspades_gr$transcript)

rnaspades_filtered_introns_clean<-unlist(GenomicRanges::setdiff(rnaspades_paf_filtered_iso_split,exons_clean_lncrna_rnaspades_gr_split))



exons_clean_lncrna_trinity_gr<-makeGRangesFromDataFrame(exons_clean_lncrna_trinity,keep.extra.columns = T)
exons_clean_lncrna_trinity_gr<-exons_clean_lncrna_trinity_gr[exons_clean_lncrna_trinity_gr$transcript%in%trinity_paf_filtered_iso$query_name]
exons_clean_lncrna_trinity_gr_split<-split(exons_clean_lncrna_trinity_gr,exons_clean_lncrna_trinity_gr$transcript)

trinity_filtered_introns_clean<-unlist(GenomicRanges::setdiff(trinity_paf_filtered_iso_split,exons_clean_lncrna_trinity_gr_split))


lncrna_filtered_introns_clean<-c(rnaspades_filtered_introns_clean,trinity_filtered_introns_clean)
lncrna_filtered_introns_clean$transcript<-names(lncrna_filtered_introns_clean)

lncrna_filtered_exons_clean<-c(exons_clean_lncrna_rnaspades_gr,exons_clean_lncrna_trinity_gr)

lncrna_filtered_introns_clean_dt<-as.data.table(lncrna_filtered_introns_clean)
isoform_map_table_together_cp<-copy(isoform_map_table_together)
setnames(isoform_map_table_together_cp,c("transcript","gene_id"))
lncrna_filtered_introns_clean_dt<-merge(lncrna_filtered_introns_clean_dt,isoform_map_table_together_cp)
lncrna_filtered_introns_clean_dt_gr<-makeGRangesFromDataFrame(lncrna_filtered_introns_clean_dt,keep.extra.columns = T)
lncrna_filtered_introns_clean_dt_gr_split<-split(lncrna_filtered_introns_clean_dt_gr,lncrna_filtered_introns_clean_dt_gr$gene_id)
lncrna_filtered_introns_clean_dt_gr_split_red<-reduce(lncrna_filtered_introns_clean_dt_gr_split)
lncrna_filtered_introns_clean_dt_gr_split_red_un<-unlist(lncrna_filtered_introns_clean_dt_gr_split,use.names = T)
lncrna_filtered_introns_clean_dt_gr_split_red_un_dt<-as.data.table(lncrna_filtered_introns_clean_dt_gr_split_red_un)

lncrna_filtered_introns_clean_dt_gr_split_red_un_dt_merge<-merge(lncrna_filtered_introns_clean_dt_gr_split_red_un_dt,lncRNA_classification_table_with_gene_id,by="gene_id")
introns_sum_table<-lncrna_filtered_introns_clean_dt_gr_split_red_un_dt_merge[,.(sum_el=sum(width)),by="class"]

lncrna_filtered_exons_clean_dt<-as.data.table(lncrna_filtered_exons_clean)
lncrna_filtered_exons_clean_dt<-merge(lncrna_filtered_exons_clean_dt,isoform_map_table_together_cp)
lncrna_filtered_exons_clean_dt_gr<-makeGRangesFromDataFrame(lncrna_filtered_exons_clean_dt,keep.extra.columns = T)
lncrna_filtered_exons_clean_dt_gr_split<-split(lncrna_filtered_exons_clean_dt_gr,lncrna_filtered_exons_clean_dt_gr$gene_id)
lncrna_filtered_exons_clean_dt_gr_split_red<-reduce(lncrna_filtered_exons_clean_dt_gr_split)
lncrna_filtered_exons_clean_dt_gr_split_red_un<-unlist(lncrna_filtered_exons_clean_dt_gr_split,use.names = T)
lncrna_filtered_exons_clean_dt_gr_split_red_un_dt<-as.data.table(lncrna_filtered_exons_clean_dt_gr_split_red_un)
lncrna_filtered_exons_clean_dt_gr_split_red_un_dt_merge<-merge(lncrna_filtered_exons_clean_dt_gr_split_red_un_dt,lncRNA_classification_table_with_gene_id,by="gene_id")
exons_sum_table<-lncrna_filtered_exons_clean_dt_gr_split_red_un_dt_merge[,.(sum_el=sum(width)),by="class"]


protein_coding_annotation_exons_clean[,gene:=str_extract(attribute,".*(?=\\.t\\d*)")]
protein_exons_new<-makeGRangesFromDataFrame(protein_coding_annotation_exons_clean,keep.extra.columns = T)
protein_exons_split<-split(protein_exons_new,protein_exons_new$gene)
protein_exons_split_red<-reduce(protein_exons_split)
protein_exons_split_red_un<-unlist(protein_exons_split_red,use.names = T)
protein_exons_split_red_un_dt<-as.data.table(protein_exons_split_red_un)
protein_exons_split_red_un_dt$attribute<-names(protein_exons_split_red_un)
protein_exons_new<-makeGRangesFromDataFrame(protein_exons_split_red_un_dt,keep.extra.columns = T)


protein_introns_new<-copy(protein_introns)

protein_introns_new$gene<-str_extract(protein_introns_new$transcript,".*(?=\\.t\\d*)")
protein_introns_new_split<-split(protein_introns_new,protein_introns_new$gene)
protein_introns_new_split_red<-reduce(protein_introns_new_split)
protein_introns_new_split_red_un<-unlist(protein_introns_new_split_red,use.names = T)
protein_introns_new_split_red_un_dt<-as.data.table(protein_introns_new_split_red_un)
protein_introns_new_split_red_un_dt$attribute<-names(protein_introns_new_split_red_un)
protein_introns_new<-makeGRangesFromDataFrame(protein_introns_new_split_red_un_dt,keep.extra.columns = T)





#repeats

repeatmasker<-readRDS("AllRepeatMaskerFiles_filtered.RDS")
setnames(repeatmasker,c("SW_score","perc.sub","perc.del","perc.ins","query_name","start","end","base_past_end","strand_ignore","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrretID","type","elementtype"))
repeatmasker_esu<-repeatmasker[species=="Eunapius"]
repeatmasker_esu<-repeatmasker_esu[repeat_class!="Low complexity" & repeat_class!="Simple repeat"]
repeatmasker_esu[,ID:=.GRP,by=c("query_name","start","end")]
repeatmasker_esu[,rep_width:=abs(repeat_start-repeat_end)+1]
repeatmasker_esu[,rep_gen_width:=abs(start-end)+1]
repeatmasker_esu[,genperc:=100*rep_width/rep_gen_width]
repeatmasker_esu<-repeatmasker_esu[query_name%in%taxonomy_table_nonbact]
repeatmasker_esu<-repeatmasker_esu[genperc>=90]
repeatmasker_esu_granges<-makeGRangesFromDataFrame(repeatmasker_esu,seqnames.field = "query_name",keep.extra.columns = T,ignore.strand = T)

#within overlaps
lncrnas_final_reduced_together_1kb<-lncrnas_final_reduced_together[start(lncrnas_final_reduced_together)>1000]
lncrnas_final_reduced_together_1kb_short<-lncrnas_final_reduced_together[start(lncrnas_final_reduced_together)<=1000]
regions_lncrna_short<-lncrnas_final_reduced_together_1kb_short
end(regions_lncrna_short)<-(start(regions_lncrna_short)-1)
start(lncrnas_final_reduced_together_1kb_short)<-(rep(1,length(lncrnas_final_reduced_together_1kb_short)))
start(regions_lncrna_short)<-(rep(1,length(lncrnas_final_reduced_together_1kb_short)))
new_starts<-start(lncrnas_final_reduced_together_1kb)-1000
start(lncrnas_final_reduced_together_1kb)<-new_starts
lncrnas_final_reduced_together_1kb_final<-c(lncrnas_final_reduced_together_1kb,lncrnas_final_reduced_together_1kb_short)
norm_width_1bb_lncrna<-1000*length(lncrnas_final_reduced_together_1kb)+sum(width(regions_lncrna_short))


protein_final_reduced_together_1kb<-protein_coding_genes_granges[start(protein_coding_genes_granges)>1000]
protein_final_reduced_together_1kb_short<-protein_coding_genes_granges[start(protein_coding_genes_granges)<=1000]
regions_prot_short<-protein_final_reduced_together_1kb_short
end(regions_prot_short)<-(start(regions_prot_short)-1)
start(protein_final_reduced_together_1kb_short)<-(rep(1,length(protein_final_reduced_together_1kb_short)))
start(regions_prot_short)<-(rep(1,length(protein_final_reduced_together_1kb_short)))
new_starts_prot<-start(protein_final_reduced_together_1kb)-1000
start(protein_final_reduced_together_1kb)<-new_starts_prot
protein_final_reduced_together_1kb_final<-c(protein_final_reduced_together_1kb,protein_final_reduced_together_1kb_short)
norm_width_1bb_prot<-1000*length(protein_final_reduced_together_1kb)+sum(width(regions_prot_short))



contig_length_table<-data.table(seqnames1=names(esu_genome),cont_length=sapply(esu_genome,length))

lncrnas_final_reduced_together_dt_cont_length_merge<-plyr::join(lncrnas_final_reduced_together_dt,contig_length_table,by="seqnames1")


lncrnas_final_reduced_together_20kb_right<-lncrnas_final_reduced_together[end(lncrnas_final_reduced_together)<(lncrnas_final_reduced_together_dt_cont_length_merge$cont_length-1000)]
lncrnas_final_reduced_together_20kb_short_right<-lncrnas_final_reduced_together[end(lncrnas_final_reduced_together)>=(lncrnas_final_reduced_together_dt_cont_length_merge$cont_length-1000)]
lncrnas_final_reduced_together_20kb_short_right_dt<-as.data.table(lncrnas_final_reduced_together_20kb_short_right)
colnames(lncrnas_final_reduced_together_20kb_short_right_dt)[1]<-"seqnames1"
lncrnas_final_reduced_together_20kb_short_right_dt_merge<-plyr::join(lncrnas_final_reduced_together_20kb_short_right_dt,contig_length_table,by="seqnames1")

end(lncrnas_final_reduced_together_20kb_short_right)<-lncrnas_final_reduced_together_20kb_short_right_dt_merge$cont_length

end(lncrnas_final_reduced_together_20kb_right)<-end(lncrnas_final_reduced_together_20kb_right)+999
dt_lncrna_right<-c(lncrnas_final_reduced_together_20kb_short_right,lncrnas_final_reduced_together_20kb_right)


dt_lncrna_right_dt<-as.data.table(dt_lncrna_right)
dt_lncrna_right_dt<-merge(dt_lncrna_right_dt,lncrnas_final_reduced_together,by="gene_id")
dt_lncrna_right_dt[,start.x:=end.y]
dt_lncrna_right_dt<-dt_lncrna_right_dt[,.SD,.SDcols=c("seqnames.x","start.x","end.x","width.x","strand.x","gene_id")]
setnames(dt_lncrna_right_dt,c("seqnames","start","end","width","strand","gene_id"))


dt_lncrna_right<-makeGRangesFromDataFrame(dt_lncrna_right_dt,keep.extra.columns = T)
norm_width_1bb_lncrna<-norm_width_1bb_lncrna+sum(width(dt_lncrna_right))
lncrnas_final_reduced_together_1kb_final_dt<-as.data.table(lncrnas_final_reduced_together_1kb_final)
lncrnas_final_reduced_together_1kb_final_dt<-merge(lncrnas_final_reduced_together_1kb_final_dt,dt_lncrna_right_dt,by="gene_id")
lncrnas_final_reduced_together_1kb_final_dt[,end.x:=end.y]
lncrnas_final_reduced_together_1kb_final_dt<-lncrnas_final_reduced_together_1kb_final_dt[,.SD,.SDcols=c("seqnames.x","start.x","end.x","strand.x","gene","gene_id")]
setnames(lncrnas_final_reduced_together_1kb_final_dt,c("seqnames","start","end","strand","gene","gene_id"))
lncrnas_final_reduced_together_1kb_final<-makeGRangesFromDataFrame(lncrnas_final_reduced_together_1kb_final_dt,keep.extra.columns = T)



setnames(contig_length_table,c("seqname","cont_length"))
protein_cont_length_merge<-plyr::join(protein_coding_genes,contig_length_table,by="seqname")


protein_20kb_right<-protein_coding_genes_granges[end(protein_coding_genes_granges)<(protein_cont_length_merge$cont_length-1000)]
protein_20kb_short_right<-protein_coding_genes_granges[end(protein_coding_genes_granges)>=(protein_cont_length_merge$cont_length-1000)]

protein_20kb_short_right_dt<-as.data.table(protein_20kb_short_right)
colnames(protein_20kb_short_right_dt)[1]<-"seqname"
protein_20kb_short_right_dt_merge<-plyr::join(protein_20kb_short_right_dt,contig_length_table,by="seqname")

end(protein_20kb_short_right)<-protein_20kb_short_right_dt_merge$cont_length

end(protein_20kb_right)<-end(protein_20kb_right)+999

dt_prot_right<-c(protein_20kb_short_right,protein_20kb_right)
dt_prot_right_dt<-as.data.table(dt_prot_right)
dt_prot_right_dt<-merge(dt_prot_right_dt,protein_coding_genes,by="attribute")
dt_prot_right_dt[,start.x:=end.y]
dt_prot_right_dt<-dt_prot_right_dt[,.SD,.SDcols=c("seqnames","start.x","end.x","width.x","strand","source.x","feature.x","score.x","strand_ignore.x","frame.x","attribute")]
setnames(dt_prot_right_dt,c("seqnames","start","end","width","strand","source","feature","score","strand_ignore","frame","attribute"))


dt_prot_right<-makeGRangesFromDataFrame(dt_prot_right_dt,keep.extra.columns = T)
norm_width_1bb_prot<-norm_width_1bb_prot+sum(width(dt_prot_right))






protein_final_reduced_together_1kb_final_dt<-as.data.table(protein_final_reduced_together_1kb_final)
protein_final_reduced_together_1kb_final_dt<-as.data.table(merge(protein_final_reduced_together_1kb_final,dt_prot_right_dt,by="attribute"))


protein_final_reduced_together_1kb_final_dt[,end.x:=end.y]
protein_final_reduced_together_1kb_final_dt<-protein_final_reduced_together_1kb_final_dt[,.SD,.SDcols=c("seqnames.x","start.x","end.x","strand.x","source.x","feature.x","score.x","strand_ignore.x","frame.x","attribute")]
setnames(protein_final_reduced_together_1kb_final_dt,c("seqnames","start","end","strand","source","feature","score","stramd_ignore","frame","attribute"))
protein_final_reduced_together_1kb_final<-makeGRangesFromDataFrame(protein_final_reduced_together_1kb_final_dt,keep.extra.columns = T)





#only_1kb_regions
regions_lncrna_long<-lncrnas_final_reduced_together_1kb
end(regions_lncrna_long)<-(start(regions_lncrna_long)+999)
regions_lncrna_together<-c(regions_lncrna_long,regions_lncrna_short)
regions_lncrna_together2<-dt_lncrna_right



regions_prot_long<-protein_final_reduced_together_1kb
end(regions_prot_long)<-(start(regions_prot_long)+999)
regions_prot_together<-c(regions_prot_long,regions_prot_short)
regions_prot_together2<-dt_prot_right


overlaps_repeats_lncrna<-findOverlaps(repeatmasker_esu_granges,lncrnas_final_reduced_together,ignore.strand=T,type="within")
overlaps_repeats_lncrna_table<-cbind(as.data.table(repeatmasker_esu[queryHits(overlaps_repeats_lncrna)]),as.data.table(lncrnas_final_reduced_together[subjectHits(overlaps_repeats_lncrna)]))
setnames(overlaps_repeats_lncrna_table,c("SW_score","perc.sub","perc.del","perc.ins","query_name","start1","end1","base_past_end","strand_ignore","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","element_type","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","gene","gene_id"))

overlaps_repeats_lncrna_1kb<-findOverlaps(repeatmasker_esu_granges,regions_lncrna_together,ignore.strand=T,type="within")
overlaps_repeats_lncrna_table_1kb<-cbind(as.data.table(repeatmasker_esu[queryHits(overlaps_repeats_lncrna_1kb)]),as.data.table(regions_lncrna_together[subjectHits(overlaps_repeats_lncrna_1kb)]))
setnames(overlaps_repeats_lncrna_table_1kb,c("SW_score","perc.sub","perc.del","perc.ins","query_name","start1","end1","base_past_end","strand_ignore","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","element_type","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","gene","gene_id"))

overlaps_repeats_lncrna_1kb_end<-findOverlaps(repeatmasker_esu_granges,regions_lncrna_together2,ignore.strand=T,type="within")
overlaps_repeats_lncrna_table_1kb_end<-cbind(as.data.table(repeatmasker_esu[queryHits(overlaps_repeats_lncrna_1kb_end)]),as.data.table(regions_lncrna_together2[subjectHits(overlaps_repeats_lncrna_1kb_end)]))
setnames(overlaps_repeats_lncrna_table_1kb,c("SW_score","perc.sub","perc.del","perc.ins","query_name","start1","end1","base_past_end","strand_ignore","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","element_type","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","gene","gene_id"))


id_trans_within<-unique(overlaps_repeats_lncrna_table[,.SD,.SDcols=c("ID","gene_id")])

id_trans_region<-unique(rbind(overlaps_repeats_lncrna_table_1kb[,.SD,.SDcols=c("ID","gene_id")],overlaps_repeats_lncrna_table_1kb_end[,.SD,.SDcols=c("ID","gene_id")]))




lncrna_repeat_overlap_category_class<-merge(id_trans_within,lncRNA_classification_table_with_gene_id,by="gene_id")
lncrna_repeat_overlap_category_class_reg<-merge(id_trans_region,lncRNA_classification_table_with_gene_id,by="gene_id")


num_within_genes<-lncrna_repeat_overlap_category_class[,uniqueN(gene_id),by=class]
num_within_regions<-lncrna_repeat_overlap_category_class_reg[,uniqueN(gene_id),by=class]


#protein transposone overlaps

#within overlaps


overlaps_repeats_proteins<-findOverlaps(repeatmasker_esu_granges,protein_coding_genes_granges,ignore.strand=T,type="within")
overlaps_repeats_proteins_table<-cbind(as.data.table(repeatmasker_esu[queryHits(overlaps_repeats_proteins)]),as.data.table(protein_coding_genes_granges[subjectHits(overlaps_repeats_proteins)]))
setnames(overlaps_repeats_proteins_table,c("SW_score","perc.sub","perc.del","perc.ins","query_name","start1","end1","base_past_end","strand_ignore","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","element_type","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","source","feature","score","strand_ignore2","frame","attribute"))

overlaps_repeats_proteins_1kb<-findOverlaps(repeatmasker_esu_granges,regions_prot_together,ignore.strand=T,type="within")
overlaps_repeats_proteins_table_1kb<-cbind(as.data.table(repeatmasker_esu[queryHits(overlaps_repeats_proteins_1kb)]),as.data.table(regions_prot_together[subjectHits(overlaps_repeats_proteins_1kb)]))
setnames(overlaps_repeats_proteins_table_1kb,c("SW_score","perc.sub","perc.del","perc.ins","query_name","start1","end1","base_past_end","strand_ignore","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","element_type","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","source","feature","score","strand_ignore2","frame","attribute"))

overlaps_repeats_proteins_1kb_end<-findOverlaps(repeatmasker_esu_granges,regions_prot_together2,ignore.strand=T,type="within")
overlaps_repeats_proteins_table_1kb_end<-cbind(as.data.table(repeatmasker_esu[queryHits(overlaps_repeats_proteins_1kb_end)]),as.data.table(regions_prot_together2[subjectHits(overlaps_repeats_proteins_1kb_end)]))
setnames(overlaps_repeats_proteins_table_1kb_end,c("SW_score","perc.sub","perc.del","perc.ins","query_name","start1","end1","base_past_end","strand_ignore","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","element_type","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","source","feature","score","strand_ignore2","frame","attribute"))

id_trans_within_prot<-unique(overlaps_repeats_proteins_table[,.SD,.SDcols=c("ID","attribute")])

id_trans_region_prot<-unique(rbind(overlaps_repeats_proteins_table_1kb[,.SD,.SDcols=c("ID","attribute")],overlaps_repeats_proteins_table_1kb_end[,.SD,.SDcols=c("ID","attribute")]))

num_within_genes_prot<-data.table(class="protein-coding gene",V1=nrow(unique(id_trans_within_prot,by="attribute")))
num_within_regions_prot<-data.table(class="protein-coding gene",V1=nrow(unique(id_trans_region_prot,by="attribute")))


genes_numb_within<-rbind(num_within_genes,num_within_genes_prot)
genes_numb_region<-rbind(num_within_regions,num_within_regions_prot)
genes_numb_region<-genes_numb_region[order(class)]
total_num_genes<-data.table(total_number_of_genes=c(lncRNA_classification_table[class=="intergenic_lncRNA",.N],lncRNA_classification_table[class=="intronic_lncRNA",.N],lncRNA_classification_table[class=="overlapping_lncRNA",.N],protein_coding_genes[,.N]))

overlaps_repeat_together_num_table<-cbind(genes_numb_within,genes_numb_region$V1,total_num_genes)
setnames(overlaps_repeat_together_num_table,c("class","genes_with_transposones_inside","genes_with_transposones_region","total_genes"))
overlaps_repeat_together_num_table[,percentage_genes_within:=100*genes_with_transposones_inside/total_genes]
overlaps_repeat_together_num_table[,percentage_genes_region:=100*genes_with_transposones_region/total_genes]

perc_within_trans<-ggplot(data=overlaps_repeat_together_num_table,aes(x=class,y=percentage_genes_within,fill=class)) +
  geom_bar(stat="identity",color="black",width=0.6) +
  theme_bw()  +
  theme(legend.position = "none") +
  coord_cartesian(ylim=c(0,100)) +
       theme(axis.text.x = element_text(angle=45,hjust=0.9)) +
  scale_x_discrete(labels=c("Intronic lncRNA","Intergenic lncRNA","Overlapping lncRNA","Protein-coding genes")) +
    theme(axis.title.x = element_blank()) +
  ylab("Percentage of genes") +
  theme(axis.title.y = element_text(size=8))


perc_region_trans<-ggplot(data=overlaps_repeat_together_num_table,aes(x=class,y=percentage_genes_region,fill=class)) +
  geom_bar(stat="identity",color="black",width=0.6) +
  theme_bw()  +
  theme(legend.position = "none") +
   coord_cartesian(ylim=c(0,100)) +
  scale_y_continuous(breaks=c(0,25,50,75,100)) +
     theme(axis.text.x = element_text(angle=45,hjust=0.9)) +
  scale_x_discrete(labels=c("Intronic lncRNA","Intergenic lncRNA","Overlapping lncRNA","Protein-coding genes")) +
  theme(axis.title.x = element_blank()) +
  ylab("Percentage of genes") +
   theme(axis.title.y = element_text(size=8))
  










#percentage of sequence

overlaps_repeats_lncrna_table[,repeat_width:=abs(repeat_start-repeat_end)+1]
overlaps_repeats_lncrna_table[,perc_seq:=100*repeat_width/width2]

mean(overlaps_repeats_lncrna_table$perc_seq)


overlaps_repeats_proteins_table[,repeat_width:=abs(repeat_start-repeat_end)+1]
overlaps_repeats_proteins_table[,perc_seq:=100*repeat_width/width2]


mean(overlaps_repeats_proteins_table$perc_seq)


#20 bp overlap

#any overlaps


overlaps_repeats_lncrna_bases<-findOverlaps(repeatmasker_esu_granges,lncrnas_final_reduced_together_1kb_final,ignore.strand=T,type="any")
overlaps_repeats_lncrna_bases_table<-cbind(as.data.table(repeatmasker_esu[queryHits(overlaps_repeats_lncrna_bases)]),as.data.table(lncrnas_final_reduced_together_1kb_final[subjectHits(overlaps_repeats_lncrna_bases)]))
setnames(overlaps_repeats_lncrna_bases_table,c("SW_score","perc.sub","perc.del","perc.ins","query_name","start1","end1","base_past_end","strand_ignore","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","element_type","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","gene","gene_id"))
inter_width_vector<-c()
for (i in 1:nrow(overlaps_repeats_lncrna_bases_table)) {
  inter_width_vector<-c(inter_width_vector,Overlap(c(overlaps_repeats_lncrna_bases_table$start1[i],overlaps_repeats_lncrna_bases_table$end1[i]),c(overlaps_repeats_lncrna_bases_table$start2[i],overlaps_repeats_lncrna_bases_table$end2[i])))
}
overlaps_repeats_lncrna_bases_table[,intersect_width:=inter_width_vector]
#overlaps_repeats_lncrna_bases_table<-overlaps_repeats_lncrna_bases_table[inter_width_vector>=10]
#repeats exons overlap


overlaps_repeats_exons_20<-findOverlaps(repeatmasker_esu_granges,lncrna_filtered_exons_clean_dt_gr_split_red_un,ignore.strand=T)
overlaps_repeats_exons_table_20<-cbind(as.data.table(repeatmasker_esu_granges[queryHits(overlaps_repeats_exons_20)]),as.data.table(lncrna_filtered_exons_clean_dt_gr_split_red_un_dt[subjectHits(overlaps_repeats_exons_20)]))
setnames(overlaps_repeats_exons_table_20,c("seqnames1","start1","end1","width1","strand1","SW_score","perc.sub","perc.del","perc.ins","base_past_end","strand_ignore","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","elementtype","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","transcript","gene_id"))
#overlaps_repeats_exons_table_20<-merge(overlaps_repeats_exons_table_20,isoform_map_table_together_cp)
inter_width_vector<-c()
for (i in 1:nrow(overlaps_repeats_exons_table_20)) {
  inter_width_vector<-c(inter_width_vector,Overlap(c(overlaps_repeats_exons_table_20$start1[i],overlaps_repeats_exons_table_20$end1[i]),c(overlaps_repeats_exons_table_20$start2[i],overlaps_repeats_exons_table_20$end2[i])))
}
overlaps_repeats_exons_table_20[,intersect_width:=inter_width_vector]

overlaps_repeats_exons_table_20[,inter_width_sum_trans:=sum(intersect_width),by=c("ID","gene_id")]


lncrna_exons_total_width<-lncrna_filtered_exons_clean_dt_gr_split_red_un_dt_merge[,.(total_ex_len=sum(width)),by=gene_id]
lncrna_ex_total_length<-overlaps_repeats_exons_table_20[,.SD,.SDcols=c("gene_id","inter_width_sum_trans")]

lncrna_ex_total_length_merge<-merge(lncrna_ex_total_length,lncrna_exons_total_width,by="gene_id")
lncrna_ex_total_length_merge[,perc_ex:=100*inter_width_sum_trans/total_ex_len]

mean(lncrna_ex_total_length_merge$perc_ex)
median(lncrna_ex_total_length_merge$perc_ex)



#repeats introns overlap

overlaps_repeats_introns_20<-findOverlaps(repeatmasker_esu_granges,lncrna_filtered_introns_clean_dt_gr_split_red_un,ignore.strand=T)
overlaps_repeats_introns_table_20<-cbind(as.data.table(repeatmasker_esu_granges[queryHits(overlaps_repeats_introns_20)]),as.data.table(lncrna_filtered_introns_clean_dt_gr_split_red_un[subjectHits(overlaps_repeats_introns_20)]))
setnames(overlaps_repeats_introns_table_20,c("seqnames1","start1","end1","width1","strand1","SW_score","perc.sub","perc.del","perc.ins","base_past_end","strand_ignore","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","elementtype","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","transcript","gene_id"))
#overlaps_repeats_introns_table_20<-merge(overlaps_repeats_introns_table_20,isoform_map_table_together_cp)
inter_width_vector<-c()
for (i in 1:nrow(overlaps_repeats_introns_table_20)) {
  inter_width_vector<-c(inter_width_vector,Overlap(c(overlaps_repeats_introns_table_20$start1[i],overlaps_repeats_introns_table_20$end1[i]),c(overlaps_repeats_introns_table_20$start2[i],overlaps_repeats_introns_table_20$end2[i])))
}
overlaps_repeats_introns_table_20[,intersect_width:=inter_width_vector]

#ex in

overlaps_repeats_exons_table_20[,gene_trans_paste:=paste(gene_id,ID,sep="_")]
overlaps_repeats_introns_table_20[,gene_trans_paste:=paste(gene_id,ID,sep="_")]




100*length(unique(overlaps_repeats_exons_table_20[gene_trans_paste%in%overlaps_repeats_introns_table_20$gene_trans_paste]$gene_trans_paste))/length(unique(overlaps_repeats_exons_table_20$gene_trans_paste))






##regions
overlaps_repeats_lncrna_bases_regions<-findOverlaps(repeatmasker_esu_granges,regions_lncrna_together,ignore.strand=T,type="any")
overlaps_repeats_lncrna_bases_table_regions<-cbind(as.data.table(repeatmasker_esu[queryHits(overlaps_repeats_lncrna_bases_regions)]),as.data.table(regions_lncrna_together[subjectHits(overlaps_repeats_lncrna_bases_regions)]))
setnames(overlaps_repeats_lncrna_bases_table_regions,c("SW_score","perc.sub","perc.del","perc.ins","query_name","start1","end1","base_past_end","strand_ignore","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","element_type","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","gene","gene_id"))
inter_width_vector<-c()
for (i in 1:nrow(overlaps_repeats_lncrna_bases_table_regions)) {
  inter_width_vector<-c(inter_width_vector,Overlap(c(overlaps_repeats_lncrna_bases_table_regions$start1[i],overlaps_repeats_lncrna_bases_table_regions$end1[i]),c(overlaps_repeats_lncrna_bases_table_regions$start2[i],overlaps_repeats_lncrna_bases_table_regions$end2[i])))
}
overlaps_repeats_lncrna_bases_table_regions[,intersect_width:=inter_width_vector]



overlaps_repeats_lncrna_bases_regions_end<-findOverlaps(repeatmasker_esu_granges,regions_lncrna_together2,ignore.strand=T,type="any")
overlaps_repeats_lncrna_bases_table_regions_end<-cbind(as.data.table(repeatmasker_esu[queryHits(overlaps_repeats_lncrna_bases_regions_end)]),as.data.table(regions_lncrna_together2[subjectHits(overlaps_repeats_lncrna_bases_regions_end)]))
setnames(overlaps_repeats_lncrna_bases_table_regions_end,c("SW_score","perc.sub","perc.del","perc.ins","query_name","start1","end1","base_past_end","strand_ignore","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","element_type","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","gene_id"))
inter_width_vector<-c()
for (i in 1:nrow(overlaps_repeats_lncrna_bases_table_regions_end)) {
  inter_width_vector<-c(inter_width_vector,Overlap(c(overlaps_repeats_lncrna_bases_table_regions_end$start1[i],overlaps_repeats_lncrna_bases_table_regions_end$end1[i]),c(overlaps_repeats_lncrna_bases_table_regions_end$start2[i],overlaps_repeats_lncrna_bases_table_regions_end$end2[i])))
}
overlaps_repeats_lncrna_bases_table_regions_end[,intersect_width:=inter_width_vector]


lncrna_overlap_exons_20<-unique(data.table(ID=overlaps_repeats_exons_table_20$ID,gene_id=overlaps_repeats_exons_table_20$gene_id,intersect=overlaps_repeats_exons_table_20$intersect_width))


lncrna_overlap_introns_20<-unique(data.table(ID=overlaps_repeats_introns_table_20$ID,gene_id=overlaps_repeats_introns_table_20$gene_id,intersect=overlaps_repeats_introns_table_20$intersect_width))


lncrna_overlap_regions_20<-unique(data.table(ID=overlaps_repeats_lncrna_bases_table_regions$ID,gene_id=overlaps_repeats_lncrna_bases_table_regions$gene_id,intersect=overlaps_repeats_lncrna_bases_table_regions$intersect_width))


lncrna_overlap_regions_20_end<-unique(data.table(ID=overlaps_repeats_lncrna_bases_table_regions_end$ID,gene_id=overlaps_repeats_lncrna_bases_table_regions_end$gene_id,intersect=overlaps_repeats_lncrna_bases_table_regions_end$intersect_width))



lncrna_overlap_exons_20[,category:="exon"]
lncrna_overlap_introns_20[,category:="intron"]
lncrna_overlap_regions_20[,category:="proximity"]
lncrna_overlap_regions_20_end[,category:="proximity"]

lncrna_repeat_overlap_category_20<-rbind(lncrna_overlap_exons_20,lncrna_overlap_introns_20,lncrna_overlap_regions_20,lncrna_overlap_regions_20_end)




lncrna_repeat_overlap_category_20_class<-merge(lncrna_repeat_overlap_category_20,lncRNA_classification_table_with_gene_id,by="gene_id")

lncrna_repeat_overlap_category_20_class_merged<-merge(lncrna_repeat_overlap_category_20_class,unique(repeatmasker_esu[,.SD,.SDcols=c("ID","repeat_class")]),by="ID")

lncrna_repeat_overlap_category_20_class_merged[,more:=ifelse(nrow(.SD)>1,T,F),by=c("ID","gene_id")]
lncrna_repeat_overlap_category_20_class_merged[,max_width_element:=ifelse(.SD==max(.SD),T,F),.SDcols="intersect",by=c("gene_id","ID")]
lncrna_repeat_overlap_category_20_class_merged[max_width_element==T,more:=F]
lncrna_repeat_overlap_category_20_class_merged<-lncrna_repeat_overlap_category_20_class_merged[more==F]

overlaps_repeat_lncrna_num_table_20<-lncrna_repeat_overlap_category_20_class_merged[,.(number_of_bases=sum(intersect)),by=c("repeat_class","category","class")]


#protein transposone overlaps - 20



#any verlaps


overlaps_repeats_proteins_bases<-findOverlaps(repeatmasker_esu_granges,protein_final_reduced_together_1kb_final,ignore.strand=T,type="any")
overlaps_repeats_proteins_bases_table<-cbind(as.data.table(repeatmasker_esu[queryHits(overlaps_repeats_proteins_bases)]),as.data.table(protein_final_reduced_together_1kb_final[subjectHits(overlaps_repeats_proteins_bases)]))
setnames(overlaps_repeats_proteins_bases_table,c("SW_score","perc.sub","perc.del","perc.ins","query_name","start1","end1","base_past_end","strand_ignore","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","element_type","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","source","feature","score","strand_ignore2","frame","attribute"))
inter_width_vector<-c()
for (i in 1:nrow(overlaps_repeats_proteins_bases_table)) {
  inter_width_vector<-c(inter_width_vector,Overlap(c(overlaps_repeats_proteins_bases_table$start1[i],overlaps_repeats_proteins_bases_table$end1[i]),c(overlaps_repeats_proteins_bases_table$start2[i],overlaps_repeats_proteins_bases_table$end2[i])))
}
overlaps_repeats_proteins_bases_table[,intersect_width:=inter_width_vector]



#ovl lncrna and prot
isoform_map_table_together3<-copy(isoform_map_table_together)
setnames(isoform_map_table_together3,c("query_name","gene_id"))
overlaps_prot_lncrna_within_table_merge<-merge(overlaps_prot_lncrna_within_table,isoform_map_table_together3,by="query_name")


lncrna_repeat_overlap_category_20_class_merged2<-lncrna_repeat_overlap_category_20_class_merged[class=="overlapping_lncRNA"]
length(unique(lncrna_repeat_overlap_category_20_class_merged2$ID))

length(which(unique(lncrna_repeat_overlap_category_20_class_merged2$ID)%in%unique(overlaps_repeats_proteins_bases_table[attribute%in%overlaps_prot_lncrna_within_table_merge$attribute]$ID)))



overlaps_repeats_proteins_bases_table
#repeats exons overlap


overlaps_repeats_protein_exons_20_prot<-findOverlaps(repeatmasker_esu_granges,protein_exons_new,ignore.strand=T)
overlaps_repeats_protein_exons_table_20_prot<-cbind(as.data.table(repeatmasker_esu_granges[queryHits(overlaps_repeats_protein_exons_20_prot)]),as.data.table(protein_exons_new[subjectHits(overlaps_repeats_protein_exons_20_prot)]))
setnames(overlaps_repeats_protein_exons_table_20_prot,c("seqnames1","start1","end1","width1","strand1","SW_score","perc.sub","perc.del","perc.ins","base_past_end","strand_ignore","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","elementtype","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","attribute"))
inter_width_vector<-c()
for (i in 1:nrow(overlaps_repeats_protein_exons_table_20_prot)) {
  inter_width_vector<-c(inter_width_vector,Overlap(c(overlaps_repeats_protein_exons_table_20_prot$start1[i],overlaps_repeats_protein_exons_table_20_prot$end1[i]),c(overlaps_repeats_protein_exons_table_20_prot$start2[i],overlaps_repeats_protein_exons_table_20_prot$end2[i])))
}
overlaps_repeats_protein_exons_table_20_prot[,intersect_width:=inter_width_vector]
overlaps_repeats_protein_exons_table_20_prot[,inter_width_sum_trans:=sum(intersect_width),by=c("ID","attribute")]
protein_exons_new_dt<-as.data.table(protein_exons_new)
protein_exons_new_dt_total_width<-protein_exons_new_dt[,.(total_ex_len=sum(width)),by=attribute]
prot_ex_total_length<-overlaps_repeats_protein_exons_table_20_prot[,.SD,.SDcols=c("attribute","inter_width_sum_trans")]

prot_ex_total_length_merge<-merge(prot_ex_total_length,protein_exons_new_dt_total_width)
prot_ex_total_length_merge[,perc_ex:=100*inter_width_sum_trans/total_ex_len]
mean(prot_ex_total_length_merge$perc_ex)
median(prot_ex_total_length_merge$perc_ex)





#repeats introns overlap

overlaps_repeats_protein_introns_20_prot<-findOverlaps(repeatmasker_esu_granges,protein_introns_new,ignore.strand=T)
overlaps_repeats_protein_introns_table_20_prot<-cbind(as.data.table(repeatmasker_esu_granges[queryHits(overlaps_repeats_protein_introns_20_prot)]),as.data.table(protein_introns_new[subjectHits(overlaps_repeats_protein_introns_20_prot)]))
setnames(overlaps_repeats_protein_introns_table_20_prot,c("seqnames1","start1","end1","width1","strand1","SW_score","perc.sub","perc.del","perc.ins","base_past_end","strand_ignore","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","elementtype","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","attribute"))
inter_width_vector<-c()
for (i in 1:nrow(overlaps_repeats_protein_introns_table_20_prot)) {
  inter_width_vector<-c(inter_width_vector,Overlap(c(overlaps_repeats_protein_introns_table_20_prot$start1[i],overlaps_repeats_protein_introns_table_20_prot$end1[i]),c(overlaps_repeats_protein_introns_table_20_prot$start2[i],overlaps_repeats_protein_introns_table_20_prot$end2[i])))
}
overlaps_repeats_protein_introns_table_20_prot[,intersect_width:=inter_width_vector]



#ex in 

overlaps_repeats_protein_exons_table_20_prot[,gene_trans_paste:=paste(attribute,ID,sep="_")]
overlaps_repeats_protein_introns_table_20_prot[,gene_trans_paste:=paste(attribute,ID,sep="_")]




100*length(unique(overlaps_repeats_protein_exons_table_20_prot[gene_trans_paste%in%overlaps_repeats_protein_introns_table_20_prot$gene_trans_paste]$gene_trans_paste))/length(unique(overlaps_repeats_protein_exons_table_20_prot$gene_trans_paste))

##regions
overlaps_repeats_protein_regions_20_prot_regions<-findOverlaps(repeatmasker_esu_granges,regions_prot_together,ignore.strand=T,type="any")
overlaps_repeats_protein_regions_20_prot_regions_table<-cbind(as.data.table(repeatmasker_esu[queryHits(overlaps_repeats_protein_regions_20_prot_regions)]),as.data.table(regions_prot_together[subjectHits(overlaps_repeats_protein_regions_20_prot_regions)]))
setnames(overlaps_repeats_protein_regions_20_prot_regions_table,c("SW_score","perc.sub","perc.del","perc.ins","query_name","start1","end1","base_past_end","strand_ignore1","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","elementtype","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","source","feature","score","strand_ignore2","frame","attribute"))
inter_width_vector<-c()
for (i in 1:nrow(overlaps_repeats_protein_regions_20_prot_regions_table)) {
  inter_width_vector<-c(inter_width_vector,Overlap(c(overlaps_repeats_protein_regions_20_prot_regions_table$start1[i],overlaps_repeats_protein_regions_20_prot_regions_table$end1[i]),c(overlaps_repeats_protein_regions_20_prot_regions_table$start2[i],overlaps_repeats_protein_regions_20_prot_regions_table$end2[i])))
}
overlaps_repeats_protein_regions_20_prot_regions_table[,intersect_width:=inter_width_vector]

overlaps_repeats_protein_regions_20_prot_regions_end<-findOverlaps(repeatmasker_esu_granges,regions_prot_together2,ignore.strand=T,type="any")
overlaps_repeats_protein_regions_20_prot_regions_table_end<-cbind(as.data.table(repeatmasker_esu[queryHits(overlaps_repeats_protein_regions_20_prot_regions_end)]),as.data.table(regions_prot_together2[subjectHits(overlaps_repeats_protein_regions_20_prot_regions_end)]))
setnames(overlaps_repeats_protein_regions_20_prot_regions_table_end,c("SW_score","perc.sub","perc.del","perc.ins","query_name","start1","end1","base_past_end","strand_ignore1","repeat_name","repeat_start","repeat_end","repeat_left_bases","ID","V16","repeat_class","repeat_family","species","size","library","group","ltrret_ID","type","elementtype","rep_width","rep_gen_width","genperc","seqnames2","start2","end2","width2","strand2","source","feature","score","strand_ignore2","frame","attribute"))
inter_width_vector<-c()
for (i in 1:nrow(overlaps_repeats_protein_regions_20_prot_regions_table_end)) {
  inter_width_vector<-c(inter_width_vector,Overlap(c(overlaps_repeats_protein_regions_20_prot_regions_table_end$start1[i],overlaps_repeats_protein_regions_20_prot_regions_table_end$end1[i]),c(overlaps_repeats_protein_regions_20_prot_regions_table_end$start2[i],overlaps_repeats_protein_regions_20_prot_regions_table_end$end2[i])))
}
overlaps_repeats_protein_regions_20_prot_regions_table_end[,intersect_width:=inter_width_vector]




protein_overlap_exons_20_prot<-unique(data.table(ID=overlaps_repeats_protein_exons_table_20_prot$ID,attribute=overlaps_repeats_protein_exons_table_20_prot$attribute,intersect=overlaps_repeats_protein_exons_table_20_prot$intersect_width))

protein_overlap_introns_20_prot<-unique(data.table(ID=overlaps_repeats_protein_introns_table_20_prot  $ID,attribute=overlaps_repeats_protein_introns_table_20_prot$attribute,intersect=overlaps_repeats_protein_introns_table_20_prot$intersect_width))

protein_overlap_regions_20<-unique(data.table(ID=overlaps_repeats_protein_regions_20_prot_regions_table$ID,attribute=overlaps_repeats_protein_regions_20_prot_regions_table$attribute,intersect=overlaps_repeats_protein_regions_20_prot_regions_table$intersect_width))

protein_overlap_regions_20_end<-unique(data.table(ID=overlaps_repeats_protein_regions_20_prot_regions_table_end$ID,attribute=overlaps_repeats_protein_regions_20_prot_regions_table_end$attribute,intersect=overlaps_repeats_protein_regions_20_prot_regions_table_end$intersect_width))





protein_overlap_exons_20_prot[,category:="exon"]
protein_overlap_introns_20_prot[,category:="intron"]
protein_overlap_regions_20[,category:="proximity"]
protein_overlap_regions_20_end[,category:="proximity"]

protein_repeat_overlap_category_20<-rbind(protein_overlap_exons_20_prot,protein_overlap_introns_20_prot,protein_overlap_regions_20,protein_overlap_regions_20_end)


protein_repeat_overlap_category_20[,class:="protein_coding_genes"]



protein_repeat_overlap_category_20_prot_merged<-merge(protein_repeat_overlap_category_20,unique(overlaps_repeats_proteins_bases_table[,.SD,.SDcols=c("ID","repeat_class")]),by="ID")



protein_repeat_overlap_category_20_prot_merged[,more:=ifelse(nrow(.SD)>1,T,F),by=c("ID","attribute")]
protein_repeat_overlap_category_20_prot_merged[,max_width_element:=ifelse(.SD==max(.SD),T,F),.SDcols="intersect",by=c("attribute","ID")]
protein_repeat_overlap_category_20_prot_merged[max_width_element==T,more:=F]
protein_repeat_overlap_category_20_prot_merged<-protein_repeat_overlap_category_20_prot_merged[more==F]

overlaps_repeat_protein_num_table_20<-protein_repeat_overlap_category_20_prot_merged[,.(number_of_bases=sum(intersect)),by=c("class","category","repeat_class")]


overlaps_repeat_together_num_table_20<-rbind(overlaps_repeat_lncrna_num_table_20,overlaps_repeat_protein_num_table_20)


overlaps_repeat_together_num_table_20[,category:=factor(category,levels = c("intron","exon","proximity"))]

overlaps_repeat_together_num_table_normalized_sum_bases_20<-copy(overlaps_repeat_together_num_table_20)


overlaps_repeat_together_num_table_20[,class:=factor(class,levels = c("intergenic_lncRNA","intronic_lncRNA","overlapping_lncRNA","protein_coding_genes"))]


#length sum normalization (20)

regions_lncrna_together_dt<-as.data.table(regions_lncrna_together)
regions_lncrna_together_dt_class<-merge(regions_lncrna_together_dt,lncRNA_classification_table_with_gene_id,by="gene_id")
regions_sum_table<-regions_lncrna_together_dt_class[,.(sum_el=sum(width)),by=class]

regions_lncrna_together_dt2<-as.data.table(regions_lncrna_together2)
regions_lncrna_together_dt_class2<-merge(regions_lncrna_together_dt2,lncRNA_classification_table_with_gene_id,by="gene_id")
regions_sum_table2<-regions_lncrna_together_dt_class2[,.(sum_el=sum(width)),by=class]

regions_sum_table[,sum_el:=sum_el+regions_sum_table2$sum_el]


overlapping_exons_sum<-exons_sum_table[class=="overlapping_lncRNA"]$sum_el
overlapping_introns_sum<-introns_sum_table[class=="overlapping_lncRNA"]$sum_el
overlapping_regions_sum<-regions_sum_table[class=="overlapping_lncRNA"]$sum_el
intronic_exons_sum<-exons_sum_table[class=="intronic_lncRNA"]$sum_el
intronic_introns_sum<-introns_sum_table[class=="intronic_lncRNA"]$sum_el
intronic_regions_sum<-regions_sum_table[class=="intronic_lncRNA"]$sum_el
intergenic_exons_sum<-exons_sum_table[class=="intergenic_lncRNA"]$sum_el
intergenic_introns_sum<-introns_sum_table[class=="intergenic_lncRNA"]$sum_el
intergenic_regions_sum<-regions_sum_table[class=="intergenic_lncRNA"]$sum_el
protein_exons_sum<-sum(width(unique(protein_exons_new)))
protein_introns_sum<-sum(width(unique(protein_introns_new)))
protein_regions_sum<-norm_width_1bb_prot

genome_length<-113195154

overlapping_exons_sum_exp<-(overlapping_exons_sum/genome_length)
overlapping_introns_sum_exp<-(overlapping_introns_sum/genome_length)
overlapping_regions_sum_exp<-(overlapping_regions_sum/genome_length)
intronic_exons_sum_exp<-(intronic_exons_sum/genome_length)
intronic_introns_sum_exp<-(intronic_introns_sum/genome_length)
intronic_regions_sum_exp<-(intronic_regions_sum/genome_length)
intergenic_exons_sum_exp<-(intergenic_exons_sum/genome_length)
intergenic_introns_sum_exp<-(intergenic_introns_sum/genome_length)
intergenic_regions_sum_exp<-(intergenic_regions_sum/genome_length)
protein_exons_sum_exp<-(protein_exons_sum/genome_length)
protein_introns_sum_exp<-(protein_introns_sum/genome_length)
protein_regions_sum_exp<-(protein_regions_sum/genome_length)



#total_lengths_dt<-data.table(class=c(rep("overlapping_lncRNA",3),rep("intronic_lncRNA",3),rep("intergenic_lncRNA",3),rep("protein_coding_genes",3)),category=rep(c("exon","intron","five_prime"),4),total_sum_inside=c(overlapping_exons_sum,overlapping_introns_sum,overlapping_regions_sum,intronic_exons_sum,intronic_introns_sum,intronic_regions_sum,intergenic_exons_sum,intergenic_introns_sum,intergenic_regions_sum,protein_exons_sum,protein_introns_sum,protein_regions_sum))
#repeat lengths normalization
class_sum<-repeatmasker_esu[,.(total_sum=sum(abs(end-start)+1)),by="repeat_class"]
overlaps_repeat_together_num_table_normalized_sum_bases_20<-merge(overlaps_repeat_together_num_table_normalized_sum_bases_20,class_sum,by="repeat_class")


#overlaps_repeat_together_num_table_normalized_sum_bases_20<-merge(overlaps_repeat_together_num_table_normalized_sum_bases_20,total_lengths_dt,by=c("class","category"))

#overlaps_repeat_together_num_table_normalized_sum_bases_20[,by_chance:=total_sum/total_sum_inside]


overlaps_repeat_together_num_table_normalized_sum_bases_20[,number_of_bases:=as.double(number_of_bases)]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="overlapping_lncRNA" & category=="exon",number_of_bases_rel:=(number_of_bases-overlapping_exons_sum_exp*total_sum)/(overlapping_exons_sum_exp*total_sum)]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="overlapping_lncRNA" & category=="intron",number_of_bases_rel:=(number_of_bases-overlapping_introns_sum_exp*total_sum)/(overlapping_introns_sum_exp*total_sum)]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="overlapping_lncRNA" & category=="proximity",number_of_bases_rel:=(number_of_bases-overlapping_regions_sum_exp*total_sum)/(overlapping_regions_sum_exp*total_sum)]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="intronic_lncRNA" & category=="exon",number_of_bases_rel:=(number_of_bases-intronic_exons_sum_exp*total_sum)/(intronic_exons_sum_exp*total_sum)]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="intronic_lncRNA" & category=="intron",number_of_bases_rel:=(number_of_bases-intronic_introns_sum_exp*total_sum)/(intronic_introns_sum_exp*total_sum)]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="intronic_lncRNA" & category=="proximity",number_of_bases_rel:=(number_of_bases-intronic_regions_sum_exp*total_sum)/(intronic_regions_sum_exp*total_sum)]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="intergenic_lncRNA" & category=="exon",number_of_bases_rel:=(number_of_bases-intergenic_exons_sum_exp*total_sum)/(intergenic_exons_sum_exp*total_sum)]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="intergenic_lncRNA" & category=="intron",number_of_bases_rel:=(number_of_bases-intergenic_introns_sum_exp*total_sum)/(intergenic_introns_sum_exp*total_sum)]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="intergenic_lncRNA" & category=="proximity",number_of_bases_rel:=(number_of_bases-intergenic_regions_sum_exp*total_sum)/(intergenic_regions_sum_exp*total_sum)]

overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="protein_coding_genes" & category=="exon",number_of_bases_rel:=(number_of_bases-protein_exons_sum_exp*total_sum)/(protein_exons_sum_exp*total_sum)]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="protein_coding_genes" & category=="intron",number_of_bases_rel:=(number_of_bases-protein_introns_sum_exp*total_sum)/(protein_introns_sum_exp*total_sum)]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="protein_coding_genes" & category=="proximity",number_of_bases_rel:=(number_of_bases-protein_regions_sum_exp*total_sum)/(protein_regions_sum_exp*total_sum)]






overlaps_repeat_together_num_table_normalized_sum_bases_20_2<-copy(overlaps_repeat_together_num_table_normalized_sum_bases_20)
overlaps_repeat_together_num_table_normalized_sum_bases_20[,number_of_bases:=number_of_bases/total_sum]

overlaps_repeat_together_num_table_normalized_sum_bases_20[,number_of_bases:=as.double(number_of_bases)]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="overlapping_lncRNA" &  category=="exon",number_of_bases:=number_of_bases/overlapping_exons_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="overlapping_lncRNA" & category=="intron",number_of_bases:=number_of_bases/overlapping_introns_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="overlapping_lncRNA" & category=="proximity",number_of_bases:=number_of_bases/overlapping_regions_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="intronic_lncRNA" & category=="exon",number_of_bases:=number_of_bases/intronic_exons_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="intronic_lncRNA" & category=="intron",number_of_bases:=number_of_bases/intronic_introns_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="intronic_lncRNA" & category=="proximity",number_of_bases:=number_of_bases/intronic_regions_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="intergenic_lncRNA" & category=="exon",number_of_bases:=number_of_bases/intergenic_exons_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="intergenic_lncRNA" & category=="intron",number_of_bases:=number_of_bases/intergenic_introns_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="intergenic_lncRNA" & category=="proximity",number_of_bases:=number_of_bases/intergenic_regions_sum]


overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="protein_coding_genes" & category=="exon",number_of_bases:=number_of_bases/protein_exons_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="protein_coding_genes" & category=="intron",number_of_bases:=number_of_bases/protein_introns_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="protein_coding_genes" & category=="proximity",number_of_bases:=number_of_bases/protein_regions_sum]


overlaps_repeat_together_num_table_normalized_sum_bases_20[,class:=factor(class,levels = c("intergenic_lncRNA","intronic_lncRNA","overlapping_lncRNA","protein_coding_genes"))]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="intergenic_lncRNA",class:="Intergenic lncRNA"]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="intronic_lncRNA",class:="Intronic lncRNA"]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="overlapping_lncRNA",class:="Overlapping lncRNA"]
overlaps_repeat_together_num_table_normalized_sum_bases_20[class=="protein_coding_genes",class:="Protein-coding genes"]
overlaps_repeat_together_num_table_normalized_sum_bases_20[,repeat_class:=factor(repeat_class,levels = c("Unknown","LINE","LTR","DNA"))]
overlaps_repeat_together_num_table_normalized_sum_bases_20<-overlaps_repeat_together_num_table_normalized_sum_bases_20[is.na(repeat_class)==F]
overlaps_repeat_together_num_table_normalized_sum_bases_20[,category:=factor(category,levels = c("proximity","exon","intron"))]


repeats_plot_normalized_sum_bases_20_rel<-ggplot(data=overlaps_repeat_together_num_table_normalized_sum_bases_20,aes(x=repeat_class,y=number_of_bases_rel,fill=category)) +
  geom_bar(position=position_dodge(preserve = "single"),stat = "identity", width=0.7,alpha=0.8) +
  theme_bw() +
  facet_grid(row=vars(class)) +
  theme(legend.position = "bottom") + 
  #theme(axis.text.x = element_text(angle = 50, hjust = 1,size=5)) +
  xlab("Transposone class") +
  ylab("Normalised total length") +
  labs(fill = "lncRNA element",size=5) +
  theme(axis.title=element_text(size=10)) +
  theme(axis.text=element_text(size=9))+
  theme(legend.title = element_text(size=10,face="bold")) +
  theme(legend.text = element_text(size=10)) +
  scale_fill_got(discrete = T,option="Martell",name="Element",labels=c("Proximity","Exons","Introns")) +
  theme(strip.text.y = element_text(size = 7)) 
  
repeats_plot_normalized_sum_bases_20_rel

ggsave(repeats_plot_normalized_sum_bases_20_rel,file="repeats_plot_normalized_sum_bases_20_rel.jpg",width=7,height = 6)


repeats_plot_normalized_sum_bases_20<-ggplot(data=overlaps_repeat_together_num_table_normalized_sum_bases_20,aes(x=repeat_class,y=number_of_bases,fill=category)) +
  geom_bar(position=position_dodge(preserve = "single"),stat = "identity", width=0.7,col="black",alpha=0.8) +
  theme_bw() +
  facet_grid(row=vars(class)) +
  theme(legend.position = "bottom") + 
  #theme(axis.text.x = element_text(angle = 50, hjust = 1,size=5)) +
  xlab("Transposone class") +
  ylab("Normalised total length") +
  labs(fill = "lncRNA element",size=5) +
  theme(axis.title=element_text(size=10)) +
  theme(axis.text=element_text(size=9))+
  theme(legend.title = element_text(size=10,face="bold")) +
  theme(legend.text = element_text(size=10)) +
  scale_fill_got(discrete = T,option="Martell",name="Element",labels=c("Proximity","Exons","Introns")) +
  scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
  theme(strip.text.y = element_text(size = 7))
  
repeats_plot_normalized_sum_bases_20

ggsave(repeats_plot_normalized_sum_bases_20,file="repeats_plot_normalized_sum_bases_20.png",width=7,height = 6)

overlaps_repeat_lncrna_num_table_20_new<-lncrna_repeat_overlap_category_20_class_merged[,.(number_of_bases=sum(intersect)),by=c("class","category")]
overlaps_repeat_protein_num_table_20_new<-protein_repeat_overlap_category_20_prot_merged[,.(number_of_bases=sum(intersect)),by=c("class","category")]

overlaps_repeat_together_num_table_normalized_sum_bases_20_2<-rbind(overlaps_repeat_lncrna_num_table_20_new,overlaps_repeat_protein_num_table_20_new)
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[,number_of_bases:=as.double(number_of_bases)]

#relativno

total_trans_length<-sum(repeatmasker_esu$rep_gen_width)
  




overlapping_exons_sum<-exons_sum_table[class=="overlapping_lncRNA"]$sum_el
overlapping_introns_sum<-introns_sum_table[class=="overlapping_lncRNA"]$sum_el
overlapping_regions_sum<-regions_sum_table[class=="overlapping_lncRNA"]$sum_el
intronic_exons_sum<-exons_sum_table[class=="intronic_lncRNA"]$sum_el
intronic_introns_sum<-introns_sum_table[class=="intronic_lncRNA"]$sum_el
intronic_regions_sum<-regions_sum_table[class=="intronic_lncRNA"]$sum_el
intergenic_exons_sum<-exons_sum_table[class=="intergenic_lncRNA"]$sum_el
intergenic_introns_sum<-introns_sum_table[class=="intergenic_lncRNA"]$sum_el
intergenic_regions_sum<-regions_sum_table[class=="intergenic_lncRNA"]$sum_el
protein_exons_sum<-sum(width(unique(protein_exons_new)))
protein_introns_sum<-sum(width(unique(protein_introns_new)))
protein_regions_sum<-norm_width_1bb_prot

overlapping_exons_sum_exp<-(overlapping_exons_sum/genome_length)*total_trans_length
overlapping_introns_sum_exp<-(overlapping_introns_sum/genome_length)*total_trans_length
overlapping_regions_sum_exp<-(overlapping_regions_sum/genome_length)*total_trans_length
intronic_exons_sum_exp<-(intronic_exons_sum/genome_length)*total_trans_length
intronic_introns_sum_exp<-(intronic_introns_sum/genome_length)*total_trans_length
intronic_regions_sum_exp<-(intronic_regions_sum/genome_length)*total_trans_length
intergenic_exons_sum_exp<-(intergenic_exons_sum/genome_length)*total_trans_length
intergenic_introns_sum_exp<-(intergenic_introns_sum/genome_length)*total_trans_length
intergenic_regions_sum_exp<-(intergenic_regions_sum/genome_length)*total_trans_length
protein_exons_sum_exp<-(protein_exons_sum/genome_length)*total_trans_length
protein_introns_sum_exp<-(protein_introns_sum/genome_length)*total_trans_length
protein_regions_sum_exp<-(protein_regions_sum/genome_length)*total_trans_length




#relative

overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="overlapping_lncRNA" & category=="exon",number_of_bases_rel:=(number_of_bases-overlapping_exons_sum_exp)/overlapping_exons_sum_exp]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="overlapping_lncRNA" & category=="intron",number_of_bases_rel:=(number_of_bases-overlapping_introns_sum_exp)/overlapping_introns_sum_exp]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="overlapping_lncRNA" & category=="proximity",number_of_bases_rel:=(number_of_bases-overlapping_regions_sum_exp)/overlapping_regions_sum_exp]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="intronic_lncRNA" & category=="exon",number_of_bases_rel:=(number_of_bases-intronic_exons_sum_exp)/intronic_exons_sum_exp]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="intronic_lncRNA" & category=="intron",number_of_bases_rel:=(number_of_bases-intronic_introns_sum_exp)/intronic_introns_sum_exp]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="intronic_lncRNA" & category=="proximity",number_of_bases_rel:=(number_of_bases-intronic_regions_sum_exp)/intronic_regions_sum_exp]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="intergenic_lncRNA" & category=="exon",number_of_bases_rel:=(number_of_bases-intergenic_exons_sum_exp)/intergenic_exons_sum_exp]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="intergenic_lncRNA" & category=="intron",number_of_bases_rel:=(number_of_bases-intergenic_introns_sum_exp)/intergenic_introns_sum_exp]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="intergenic_lncRNA" & category=="proximity",number_of_bases_rel:=(number_of_bases-intergenic_regions_sum_exp)/intergenic_regions_sum_exp]


overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="protein_coding_genes" & category=="exon",number_of_bases_rel:=(number_of_bases-protein_exons_sum_exp)/protein_exons_sum_exp]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="protein_coding_genes" & category=="intron",number_of_bases_rel:=(number_of_bases-protein_introns_sum_exp)/protein_introns_sum_exp]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="protein_coding_genes" & category=="proximity",number_of_bases_rel:=(number_of_bases-protein_regions_sum_exp)/protein_regions_sum_exp]


overlaps_repeat_together_num_table_normalized_sum_bases_20_2[,class:=factor(class,levels = c("intergenic_lncRNA","intronic_lncRNA","overlapping_lncRNA","protein_coding_genes"))]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="intergenic_lncRNA",class:="Intergeic lncRNA"]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="intronic_lncRNA",class:="Intronic lncRNA"]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="overlapping_lncRNA",class:="Overlapping lncRNA"]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="protein_coding_genes",class:="Protein-coding genes"]


overlaps_repeat_together_num_table_normalized_sum_bases_20_2[,category:=factor(category,levels = c("proximity","exon","intron"))]
repeats_plot_normalized_sum_bases_20_2_rel<-ggplot(data=overlaps_repeat_together_num_table_normalized_sum_bases_20_2,aes(x=class,y=number_of_bases_rel,fill=category)) +
  geom_bar(position="dodge",stat = "identity",col="black",alpha=0.8) +
  theme_bw() +
  theme(legend.position = "top") + 
  #theme(axis.text.x = element_text(angle = 50, hjust = 1,size=5)) +
  xlab("Gene class") +
  ylab("Normalised total length") +
  labs(fill = "lncRNA element",size=5) +
  theme(axis.title=element_text(size=11)) +
  theme(axis.text=element_text(size=10.3))+
  theme(legend.title = element_text(size=10,face="bold")) +
  theme(legend.text = element_text(size=10)) +
  scale_fill_got(discrete = T,option="Martell",name="Element",labels=c("Proximity","Exon","Intron")) +
  scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) 

  


  
repeats_plot_normalized_sum_bases_20_2_rel

ggsave(repeats_plot_normalized_sum_bases_20_2_rel,file="repeats_plot_normalized_sum_bases_20_2_rel.jpg",width=7.6,height = 3.5)


#relativno

#normalization
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="Overlapping lncRNA" & category=="exon",number_of_bases:=number_of_bases/overlapping_exons_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="Overlapping lncRNA" & category=="intron",number_of_bases:=number_of_bases/overlapping_introns_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="Overlapping lncRNA" & category=="proximity",number_of_bases:=number_of_bases/overlapping_regions_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="Intronic lncRNA" & category=="exon",number_of_bases:=number_of_bases/intronic_exons_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="Intronic lncRNA" & category=="intron",number_of_bases:=number_of_bases/intronic_introns_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="Intronic lncRNA" & category=="proximity",number_of_bases:=number_of_bases/intronic_regions_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="Intergeic lncRNA" & category=="exon",number_of_bases:=number_of_bases/intergenic_exons_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="Intergeic lncRNA" & category=="intron",number_of_bases:=number_of_bases/intergenic_introns_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="Intergeic lncRNA" & category=="proximity",number_of_bases:=number_of_bases/intergenic_regions_sum]


overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="Protein-coding genes" & category=="exon",number_of_bases:=number_of_bases/protein_exons_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="Protein-coding genes" & category=="intron",number_of_bases:=number_of_bases/protein_introns_sum]
overlaps_repeat_together_num_table_normalized_sum_bases_20_2[class=="Protein-coding genes" & category=="proximity",number_of_bases:=number_of_bases/protein_regions_sum]



overlaps_repeat_together_num_table_normalized_sum_bases_20_2[,category:=factor(category,levels = c("proximity","exon","intron"))]
repeats_plot_normalized_sum_bases_20_2<-ggplot(data=overlaps_repeat_together_num_table_normalized_sum_bases_20_2,aes(x=class,y=number_of_bases,fill=category)) +
  geom_bar(position="stack",stat = "identity",col="black",alpha=0.8,width=0.8) +
  theme_bw() +
  theme(legend.position = "none") + 
  #theme(axis.text.x = element_text(angle = 50, hjust = 1,size=5)) +
  xlab("Gene class") +
  ylab("Normalised total length") +
  labs(fill = "lncRNA element",size=5) +
  theme(axis.title=element_text(size=11)) +
  theme(axis.text.x=element_text(size=8,angle=45,hjust=0.9))+
  theme(legend.title = element_text(size=10,face="bold")) +
  theme(legend.text = element_text(size=10)) +
  scale_fill_got(discrete = T,option="Martell",name="Element",labels=c("Proximity","Exon","Intron")) +
  scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
    scale_x_discrete(labels=c("Intronic lncRNA","Intergenic lncRNA","Overlapping lncRNA","Protein-coding genes"))

  


  
repeats_plot_normalized_sum_bases_20_2

ggsave(repeats_plot_normalized_sum_bases_20_2,file="repeats_plot_normalized_sum_bases_20_2.png",width=7.6,height = 3.5)



figure_four<-ggdraw() +
  draw_plot(perc_within_trans, x = 0, y = 0.48, width = .2, height = .38) +
  draw_plot(perc_region_trans, x = .25, y = 0.48, width = .2, height = .38) +
  draw_plot(repeats_plot_normalized_sum_bases_20_2, x = 0.05, y = 0.05, width = 0.35, height = 0.4) +
  draw_plot(repeats_plot_normalized_sum_bases_20_rel, x = 0.5, y = 0.2, width = 0.5, height = 0.6) +
  draw_plot_label(label = c("A", "B", "C","D"), size = 15,
                  x = c(0.035, 0.27, 0.05,0.5), y = c(0.9, 0.9, 0.47,0.83))


figure_four

ggsave(figure_four,filename = "figure_four.jpg",height = 10,width = 13)




#Sponge BLASTn analysis

blast_results_sponge_transcriptomes<-fread("together_longest_iso_merged_sponge_resources_custom_format_manual_hm_w64_l18.txt")
setnames(blast_results_sponge_transcriptomes,c("qseqid","sseqid","pident","length","mismatch","gapopen","qstart","qend","qlen","sstart","send","slen","evalue","bitscore"))

blast_results_sponge_transcriptomes[,qperclen:=100*(abs(qend-qstart)/qlen)]
blast_results_sponge_transcriptomes[grepl("genome",sseqid),species:=str_extract(sseqid,".*_genome")]
blast_results_sponge_transcriptomes[grepl("transcriptome",sseqid),species:=str_extract(sseqid,".*_transcriptome")]

blast_results_sponge_transcriptomes_eval_filt<-blast_results_sponge_transcriptomes[evalue<=1e-13]

blast_results_sponge_transcriptomes_eval_filt[,number_of_species_per_gene:=uniqueN(.SD),.SDcols="species",by=qseqid]
trans_more_than_one_species<-unique(blast_results_sponge_transcriptomes_eval_filt[number_of_species_per_gene>1]$qseqid)

write.table(trans_more_than_one_species,file="trans_more_than_one_species.txt",row.names = F,col.names = F,quote = F,sep = "\t")

numb_gene_in_genomes_table<-blast_results_sponge_transcriptomes_eval_filt[,uniqueN(qseqid),by=number_of_species_per_gene]
setnames(numb_gene_in_genomes_table,c("number_of_genomes","number_of_lncrna"))
total_lncrna_num<-final_lncrna_list_together_length
numb_gene_in_genomes_table<-rbind(data.table(number_of_genomes=0,number_of_lncrna=total_lncrna_num-sum(numb_gene_in_genomes_table$number_of_lncrna)),numb_gene_in_genomes_table)
numb_gene_in_genomes_table<-numb_gene_in_genomes_table[order(number_of_genomes)]

numb_gene_in_genomes_plot<-ggplot(data=numb_gene_in_genomes_table,aes(x=number_of_genomes,y=number_of_lncrna)) +
  geom_bar(position = "dodge",stat="identity",width=0.6,col="black", fill="indianred1") +
  theme_bw() +
  xlab("Number of species hit by a lncRNA") +
  ylab("Number of lncRNA")

numb_gene_in_genomes_plot

ggsave(numb_gene_in_genomes_plot,filename = "numb_gene_in_genomes_plot.png")

blast_results_sponge_transcriptomes_eval_filt[sstart>send,':='(sstart=send,send=sstart)]
blast_gr<-makeGRangesFromDataFrame(blast_results_sponge_transcriptomes_eval_filt,start.field = "qstart",end.field="qend",seqnames.field = "qseqid",ignore.strand=T,keep.extra.columns = T)
blast_gr_q_split<-split(blast_gr,blast_gr$sseqid)
blast_gr_q_split_reduce<-reduce(blast_gr_q_split)
blast_gr_q_split_reduce_unlist<-unlist(blast_gr_q_split_reduce,use.names = T)

blast_gr_q_split_reduce_unlist_dt<-as.data.table(blast_gr_q_split_reduce_unlist)
blast_gr_q_split_reduce_unlist_dt[,sseqid:=names(blast_gr_q_split_reduce_unlist)]
setnames(blast_gr_q_split_reduce_unlist_dt,c("qseqid","qstart","qend","width","strand","sseqid"))
blast_gr_q_split_reduce_unlist_dt_lengths<-merge(blast_gr_q_split_reduce_unlist_dt,unique(blast_results_sponge_transcriptomes_eval_filt[,.SD,.SDcols=c("qseqid","qlen")]),by="qseqid")
blast_gr_q_split_reduce_unlist_dt_lengths[,qperclen:=100*(width/qlen)]
blast_gr_q_split_reduce_unlist_dt_lengths[,sum_width:=sum(.SD),.SDcols="width",by=c("qseqid","sseqid")]
blast_gr_q_split_reduce_unlist_dt_lengths[,sumqperclen:=sum(.SD),.SDcols="qperclen",by=c("qseqid","sseqid")]
blast_gr_q_split_reduce_unlist_dt_lengths[grepl("genome",sseqid),species:=str_extract(sseqid,".*_genome")]
blast_gr_q_split_reduce_unlist_dt_lengths[grepl("transcriptome",sseqid),species:=str_extract(sseqid,".*_transcriptome")]

blast_gr_q_split_reduce_unlist_dt_lengths_one_per_species<-blast_gr_q_split_reduce_unlist_dt_lengths[,.(sumqperclen=max(sumqperclen)),by=c("qseqid","species")]

#blast_results_sponge_transcriptomes_eval_filt_one_per_transcriptome[,log_e_value:=-log10(e_value+1e-300)]

species_list<-c("Aphrocallistes_vastus_transcriptome","Chondrilla_nucula_transcriptome","Clathrina_sp_transcriptome","Corticium_candelabrum_transcriptome","Crella_elegans_transcriptome","Ephydatia_mulleri_transcriptome","Ephydatia_mullerig_genone","Haliclona_amboinensis_transcriptome","Haliclona_indistincta_transcriptome","Haliclona_oculata_transcriptome","Haliclona_simulans_transcriptome","Haliclona_tubifera_transcriptome","Halisraca_caerulea_transcriptome","Ircinia_fasciculata_transcriptome","Leucosolenia_complicata_transcriptome","Oscarella_carmela_genome","Oscarella_sp_transcriptome","Petrosia_ficiformis_transcriptome","Pseudospongosorites_suberitoides_transcriptome","Stylissa_carteri_genome","Sycon_ciliatum_genome","Xestospongia_testudinaria_genome","Amphimedon_queenslandica_genome")
nonexistant_species<-setdiff(species_list,unique(blast_gr_q_split_reduce_unlist_dt_lengths_one_per_species$species))




blast_gr_q_split_reduce_unlist_dt_lengths_one_per_species<-rbind(blast_gr_q_split_reduce_unlist_dt_lengths_one_per_species,data.table(qseqid=rep("NODE_6217_length_4525_cov_17.657251_g1739_i1",length(nonexistant_species)),species=nonexistant_species,sumqperclen=rep(NA,length(nonexistant_species))))

species_ordered<-blast_gr_q_split_reduce_unlist_dt_lengths_one_per_species[,.N,by=species][order(-N)]$species
label_cols<-ifelse(grepl("genome",species_ordered)==T,"brown2","gray52")

blast_gr_q_split_reduce_unlist_dt_lengths_one_per_species[,species:=str_remove(species,"_genome")]
blast_gr_q_split_reduce_unlist_dt_lengths_one_per_species[,species:=str_remove(species,"_transcriptome")]
species_ordered2<-str_remove(species_ordered,"_genome")
species_ordered2<-unique(str_remove(species_ordered2,"_transcriptome"))
species_ordered3<-species_ordered2
species_ordered3[species_ordered3=="Ephydatia_mullerig"]<-"Ephydatia_mulleri" 
species_ordered3<-str_replace(species_ordered3,"_"," ")
blast_heatmap_plot<-ggplot(data=blast_gr_q_split_reduce_unlist_dt_lengths_one_per_species,aes(y=species,x=qseqid,fill=sumqperclen)) +
  geom_tile(alpha=1.2) +
  scale_fill_hp(house = "HermioneGranger",name="Percentage of lncRNA aligned\n to subject sequence",na.value="white",direction = -1) +
  xlab("lncRNA") + 
  theme(legend.position="top") +
  theme(legend.key.size = unit(0.3, "cm"),
        legend.key.width = unit(0.5,"cm")) +
  theme(legend.title = element_text( size = 9, face="bold")) +
  theme(axis.title.y = element_blank()) +
  theme(axis.title.x = element_text(size=9.2)) +
  theme(axis.ticks.x=element_blank()) +
  theme(axis.text.x=element_blank())  +
  scale_y_discrete(limits=rev(species_ordered2),labels=rev(species_ordered3)) +
  theme(axis.text.y = element_text(size=9,colour = rev(label_cols))) 



blast_heatmap_plot


ggsave(blast_heatmap_plot,file="blast_heatmap_plot.png",width = 7.6,height = 8)

figure_five<-ggdraw() +
  draw_plot(numb_gene_in_genomes_plot, x = 0, y = 0.42, width = .35, height = .38) +
  draw_plot(blast_heatmap_plot, x = .4, y = 0.35, width = .5, height = .6) +
  draw_plot_label(label = c("A", "B","C"), size = 15,
                  x = c(0.02, 0.46,0.12), y = c(0.8, 0.94,0.32))


figure_five

ggsave(figure_five,filename = "figure_five.jpg",height = 11,width = 13)

#findig candidatess for multiple alignment

blast_gr_q_split_reduce_unlist_dt_lengths<-blast_gr_q_split_reduce_unlist_dt_lengths[grepl("transcriptome",species)]

blast_gr_q_split_reduce_unlist_dt_lengths[,is_max_sumqperclen:=ifelse(.SD==max(.SD),T,F),.SDcols="sumqperclen",by=c("qseqid","species")]

only_max_sumperceln_width_filtered<-blast_gr_q_split_reduce_unlist_dt_lengths[is_max_sumqperclen==T & sum_width>=150]

only_max_sumperceln_width_filtered_all_width<-blast_gr_q_split_reduce_unlist_dt_lengths[is_max_sumqperclen==T]
only_max_sumperceln_width_filtered_all_width<-unique(only_max_sumperceln_width_filtered_all_width,by=c("qseqid","species"))


sumandwidth_filtered<-merge(blast_results_sponge_transcriptomes_eval_filt,unique(only_max_sumperceln_width_filtered[,.SD,.SDcols=c("qseqid","sseqid","is_max_sumqperclen")]),by=c("qseqid","sseqid"),all.x=T)[is_max_sumqperclen==T]
sumandwidth_filtered<-sumandwidth_filtered[order(evalue)]
sumandwidth_filtered<-unique(sumandwidth_filtered,by=c("qseqid","species"))
sumandwidth_filtered[sstart>30,sstart:=sstart-30]
sumandwidth_filtered[send<slen-30,send:=send+30]
sumandwidth_filtered_gr<-makeGRangesFromDataFrame(sumandwidth_filtered,seqnames.field="sseqid",start.field = "sstart",end.field = "send",ignore.strand = T,keep.extra.columns = T)



sumandwidth_filtered_gr_table<-as.data.table(sumandwidth_filtered_gr)[,.SD,.SDcols=c("seqnames","start","end","qseqid")]
colnames(sumandwidth_filtered_gr_table)[4]<-"transcript"
#blast - bed conversion 

sumandwidth_filtered_gr_table[,start:=start-1]

sumandwidth_filtered_gr_table[,species_num:=nrow(.SD),by=transcript]
sumandwidth_filtered_gr_table<-sumandwidth_filtered_gr_table[species_num>1]
sumandwidth_filtered_gr_table$species_num<-NULL


sumandwidth_filtered_gr_table_split<-split(sumandwidth_filtered_gr_table,sumandwidth_filtered_gr_table$transcript)


for (i in sumandwidth_filtered_gr_table_split) {
  write.table(i,file=paste(i$transcript[1],"msa_part_trans.txt",sep="_"),row.names = F, col.names = F, quote = F,sep = "\t")
}

#whole transcripts
sumandwidth_filtered_gr_table_for_merge<-copy(sumandwidth_filtered_gr_table)

setnames(sumandwidth_filtered_gr_table_for_merge,c("sseqid","sstart","send","qseqid"))
sumandwidth_filtered_gr_table_merged<-merge(sumandwidth_filtered_gr_table_for_merge,unique(blast_results_sponge_transcriptomes[,.SD,.SDcols=c("sseqid","qseqid","slen")],by=c("sseqid","qseqid")))

sumandwidth_filtered_gr_table_merged[,':='(sstart=0,send=slen)]
sumandwidth_filtered_gr_table_merged$slen<-NULL
sumandwidth_filtered_gr_table_merged<-sumandwidth_filtered_gr_table_merged[,.SD,.SDcols=c("sseqid","sstart","send","qseqid")]
sumandwidth_filtered_gr_table_whole_trans_split<-split(sumandwidth_filtered_gr_table_merged,sumandwidth_filtered_gr_table_merged$qseqid)

for (i in sumandwidth_filtered_gr_table_whole_trans_split) {
  write.table(i,file=paste(i$qseqid[1],"msa_whole_trans.txt",sep="_"),row.names = F, col.names = F, quote = F,sep = "\t")
}

#SAF format for featurecounts

protein_exons_dt<-as.data.table(protein_exons)

protein_exons_dt_saf<-protein_exons_dt[,.SD,.SDcols=c("attribute","seqnames","start","end","strand_ignore")]
protein_exons_dt_saf[,attribute:=str_extract(attribute,".*(?=\\.t\\d*)")]
setnames(protein_exons_dt_saf,c("GeneID","Chr","Start","End","Strand"))

filtered_together_paf_union_exons_saf<-filtered_together_paf_union_exons[,.SD,.SDcols=c("transcript","chromosome","start","end")]
setnames(filtered_together_paf_union_exons_saf,c("transcript_name","chromosome","start","end"))
filtered_together_paf_union_exons_saf<-merge(filtered_together_paf_union_exons_saf,isoform_map_table_together,by="transcript_name")
filtered_together_paf_union_exons_saf[,strand:="+"]
filtered_together_paf_union_exons_saf$transcript_name<-filtered_together_paf_union_exons_saf$gene_id
filtered_together_paf_union_exons_saf$gene_id<-NULL
setnames(filtered_together_paf_union_exons_saf,c("GeneID","Chr","Start","End","Strand"))


saf_together<-unique(rbind(protein_exons_dt_saf,filtered_together_paf_union_exons_saf))

write.table(saf_together,file="lncrna_and_prot_annotation.saf",row.names = F,quote = F,sep = "\t")
```





```{r}
setnames(intergenic_prot_nearest_table,c("seqnames1","start1","end1","gene_id","seqnames2","start2","end2","attribute","distance"))

#overlapping proteins vs non-overlapping proteins

#intronic


intronic_prots_dt<-overlaps_lncRNA_prot_within_table
isoform_map_table_together_cp2<-copy(isoform_map_table_together_cp)
setnames(isoform_map_table_together_cp2,c("query_name","gene_id"))
intronic_prots_dt<-merge(intronic_prots_dt,isoform_map_table_together_cp2,by="query_name")
intronic_prots_dt[,distance:=0]

#overlapping

overlapping_prots_dt<-overlaps_prot_lncrna_within_table
overlapping_prots_dt<-merge(overlapping_prots_dt,isoform_map_table_together_cp2,by="query_name")
overlapping_prots_dt[,distance:=0]



prot_exons_dt_first_last<-protein_coding_annotation[feature=="CDS"]
prot_first_exons<-prot_exons_dt_first_last[,.SD[1],by=gene,.SDcols=c("seqname","start","end")]
prot_last_exons<-prot_exons_dt_first_last[,.SD[.N],by=gene,.SDcols=c("seqname","start","end")]
setnames(prot_first_exons,c("attribute","seqnames_exon","start_exon","end_exon"))
setnames(prot_last_exons,c("attribute","seqnames_exon","start_exon","end_exon"))
prot_first_exons[,exon_num:="first"]
prot_last_exons[,exon_num:="last"]
table_adj_lncrna_prot<-rbind(intergenic_prot_nearest_table[,.SD,.SDcols=c("gene_id","attribute","distance")],intronic_prots_dt[,.SD,.SDcols=c("gene_id","attribute","distance")],overlapping_prots_dt[,.SD,.SDcols=c("gene_id","attribute","distance")])
table_adj_lncrna_prot<-merge(table_adj_lncrna_prot,lncRNA_classification_table_with_gene_id,by="gene_id")
table_adj_lncrna_prot_first_merge<-merge(table_adj_lncrna_prot,prot_first_exons,by="attribute")
table_adj_lncrna_prot_last_merge<-merge(table_adj_lncrna_prot,prot_last_exons,by="attribute")

table_adj_lncrna_prot_together_merge<-rbind(table_adj_lncrna_prot_first_merge,table_adj_lncrna_prot_last_merge)

table_adj_lncrna_prot_together_merge[class=="intergenic_lncRNA",keep_col:=ifelse((distance<0 & exon_num=="first") | (distance > 0 & exon_num=="last"),T,F)]
table_adj_lncrna_prot_together_merge[class!="intergenic_lncRNA",keep_col:=T]
table_adj_lncrna_prot_together_merge<-table_adj_lncrna_prot_together_merge[keep_col==T]

write.table(table_adj_lncrna_prot_together_merge,file = "table_adj_lncrna_prot_together_merge.txt",quote = F,row.names = F,col.names = T)
```

```{r}
#counts

read_counts_table<-fread("together_counts.txt")
read_counts_table<-read_counts_table[,.SD,.SDcols=c("Geneid","Length","sorted_RNAseq_together_to_Esu_geonme_10k.bam")]
setnames(read_counts_table,c("gene_id","gene_length","counts"))
total_counts_table<-fread("total_mapped_counts_2.txt")
setnames(total_counts_table,"total_mapped_counts")
#read_counts_table<-read_counts_table[counts!=0]
read_counts_table[,counts_scaled_milion:=1e+06*counts]
read_counts_table[,FPKM:=counts_scaled_milion/(total_counts_table$total_mapped_counts*1e-3*gene_length)]
read_counts_table[grepl("\\D",gene_id)==T,category:="protein_coding_gene"]
read_counts_table[is.na(category)==T,category:="lncRNA"]



read_counts_table_lncrna<-read_counts_table[category=="lncRNA"]
read_counts_table_lncrna[,gene_id:=as.double(gene_id)]
read_counts_table_lncrna_class<-merge(read_counts_table_lncrna,lncRNA_classification_table_with_gene_id,by="gene_id")


read_counts_table_lncrna_class[,class:=factor(class,levels = c("intronic_lncRNA","intergenic_lncRNA","overlapping_lncRNA"))]
read_counts_table_lncrna_class[class=="intronic_lncRNA",class:="Intronic lncRNA"]
read_counts_table_lncrna_class[class=="intergenic_lncRNA",class:="Intergenic lncRNA"]
read_counts_table_lncrna_class[class=="overlapping_lncRNA",class:="Overlapping lncRNA"]

#expression boxplots

fpkm_class_comp<-list(c("Intronic lncRNA","Intergenic lncRNA"),c("Intronic lncRNA","Overlapping lncRNA"),c("Intergenic lncRNA","Overlapping lncRNA"))
fpkm_lncrna_class_plot<-ggplot(data=read_counts_table_lncrna_class,aes(x=class,y=FPKM,fill=class)) +
  geom_boxplot(width=0.45) +
  theme_bw() +
  scale_y_log10() +
  scale_fill_hp(discrete = T,option = "NewtScamander") +
  theme(legend.position = "none") +
  theme(axis.title.x=element_blank()) +
  theme(axis.text.x = element_text(angle = 50, hjust = 1)) +
  scale_x_discrete(labels=c("Intronic lncRNA","Intergenic lncRNA","Overlapping lncRNA")) +
  stat_compare_means(comparisons = fpkm_class_comp,label="p.signif")

fpkm_lncrna_class_plot
ggsave(fpkm_lncrna_class_plot,file="fpkm_lncrna_class_plot.png",width = 6,height = 5)


kruskal.test(FPKM~class,data = read_counts_table_lncrna_class)
DunnTest(FPKM~class,data = read_counts_table_lncrna_class)
#expression and transposons

lncrna_repeat_overlap_category_class_merged_copy<-copy(lncrna_repeat_overlap_category_20_class_merged)
lncrna_repeat_overlap_category_class_merged_copy[,sum_intersect:=sum(intersect),by=c("gene_id","ID")]
lncrna_repeat_overlap_category_class_merged_copy[,max_sum:=ifelse(.SD==max(.SD),T,F),.SDcols="sum_intersect",by=gene_id]
lncrna_repeat_overlap_category_class_merged_copy<-lncrna_repeat_overlap_category_class_merged_copy[max_sum==T]

setnames(lncrna_repeat_overlap_category_class_merged_copy,c("ID","gene_id","intersect","element","transcript","class","repeat_class","more","max_width_element","sum_intersect","max_sum"))
lncrna_repeat_overlap_category_class_merged_copy_merged<-merge(lncrna_repeat_overlap_category_class_merged_copy,read_counts_table_lncrna[,.SD,.SDcols=c("gene_id","FPKM")],by="gene_id")
isoform_map_table_together_cp2<-copy(isoform_map_table_together)
setnames(isoform_map_table_together_cp2,c("query_name","gene_id"))
filtered_together_paf_union_copy<-copy(filtered_together_paf_union)
filtered_together_paf_union_copy<-merge(filtered_together_paf_union_copy,isoform_map_table_together_cp2)
lncrna_no_repeats<-filtered_together_paf_union_copy[gene_id%in%lncrna_repeat_overlap_category_class_merged_copy_merged$gene_id==F]

lncrna_no_repeats_dt<-data.table(gene_id=lncrna_no_repeats$gene_id)
lncrna_no_repeats_dt_merged<-merge(lncrna_no_repeats_dt,read_counts_table_lncrna[,.SD,.SDcols=c("gene_id","FPKM")],by="gene_id")
lncrna_no_repeats_dt_merged[,repeat_class:="No_transposons"]
lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep<-rbind(lncrna_repeat_overlap_category_class_merged_copy_merged,lncrna_no_repeats_dt_merged,fill=T)
lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[,repeat_class:=factor(repeat_class,levels = c("Unknown","DNA","LINE","LTR","RC","No_transposons"))]


lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[,log_FPKM:=log(FPKM+1e-10)]
#lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[,element:=factor(element,levels = levels_for_plot)]

options(scipen=1000)
fpkm_lncrna_repeats_plot<-ggplot(data=lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep,aes(x=repeat_class,y=FPKM,fill=element)) +
  geom_boxplot(alpha=0.8) +
  theme_bw() +
  scale_y_log10() +
  theme(axis.text.x = element_text(angle = 50, hjust = 1)) +
  scale_fill_got(discrete = T,option = "martell",na.value="grey",labels=c("egzon","intron","neodređeno"),direction = -1) +
  xlab("Razred transpozona") +
  #theme(axis.text.x=element_blank()) +
  theme(legend.position = "none") +
  scale_x_discrete(labels=c("Unknown","DNA","LINE","LTR","Bez transpozona"))
 


fpkm_lncrna_repeats_plot
ggsave(fpkm_lncrna_repeats_plot,file="fpkm_lncrna_repeats_plot.png",width = 6,height = 5)




lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[,cate:=paste(element,repeat_class,sep=" ")]
anova_log_fpkm_trans<-aov(log_FPKM~cate,data=lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[is.infinite(log_FPKM)==F])

TukeyHSD(anova_log_fpkm_trans)


t.test(lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="proximity Unknown"]$log_FPKM,lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="NA No_transposons"]$log_FPKM)

t.test(lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="exon Unknown"]$log_FPKM,lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="NA No_transposons"]$log_FPKM)

t.test(lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="intron Unknown"]$log_FPKM,lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="NA No_transposons"]$log_FPKM)

wilcox.test(lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="intron Unknown"]$FPKM,lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="NA No_transposons"]$FPKM)


t.test(lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="proximity DNA"]$log_FPKM,lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="NA No_transposons"]$log_FPKM)

t.test(lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="exon DNA"]$log_FPKM,lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="NA No_transposons"]$log_FPKM)

t.test(lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="intron DNA"]$log_FPKM,lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="NA No_transposons"]$log_FPKM)

t.test(lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="proximity LINE"]$log_FPKM,lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="NA No_transposons"]$log_FPKM)

t.test(lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="exon LINE"]$log_FPKM,lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="NA No_transposons"]$log_FPKM)

t.test(lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="intron LINE"]$log_FPKM,lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="NA No_transposons"]$log_FPKM)

t.test(lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="proximity LTR"]$log_FPKM,lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="NA No_transposons"]$log_FPKM)

#t.test(lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="exon LTR"]$log_FPKM,lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="NA No_transposons"]$log_FPKM)

t.test(lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="intron LTR"]$log_FPKM,lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[cate=="NA No_transposons"]$log_FPKM)








#numbersplot


levels_for_plot<-c("proximity","exon","intron",NA)



lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep[,element:=factor(element,levels = levels_for_plot)]

fpkm_lncrna_repeats_barplot<-ggplot(data=lncrna_repeat_overlap_category_class_merged_copy_merged_with_no_rep,aes(x=repeat_class,fill=element)) +
  geom_bar(position="dodge",col="black",alpha=0.8) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 50, hjust = 1)) +
  theme(legend.position = "top") +
  scale_fill_got(discrete = T,option = "martell",na.value="grey",labels=c("5' regija","egzon","intron","Bez transpozona"), name="Element",direction = -1) +
  xlab("Transposone class") +
  ylab("Number of lncRNA-a genes") +
  theme(legend.title = element_text(size=9,face="bold")) +
  scale_x_discrete(labels=c("Unknown","DNA","LINE","LTR","No transposones")) +
  theme(axis.title.x = element_blank()) +
  theme(axis.text.x  = element_blank())


fpkm_lncrna_repeats_barplot
ggsave(fpkm_lncrna_repeats_barplot,file="fpkm_lncrna_repeats_barplot.png",width=5,height = 4)







#overlapping proteisn vs non-overlapping proteins


read_counts_table_longer_prot<-read_counts_table[category=="protein_coding_gene"]
setnames(prot_classification_table,c("gene_id","lncrna_class"))
read_counts_table_longer_prot_merged<-merge(read_counts_table_longer_prot,prot_classification_table,by="gene_id")
read_counts_table_longer_prot_merged[,lncrna_class:=factor(lncrna_class,levels = c("intronic","intergenic_1kb","overlapping","no_lncRNA"))]

options(scipen=1000)
fpkm_prot_comp<-list(c("no_lncRNA","intergenic_1kb"),c("no_lncRNA","intronic"),c("no_lncRNA","overlapping"))
fpkm_prot_plot<-ggplot(data=read_counts_table_longer_prot_merged,aes(x=lncrna_class,y=FPKM,fill=lncrna_class)) +
  geom_boxplot() +
  theme_bw() +
  scale_y_log10() +
  theme(legend.title = element_text(size=9,face="bold")) +
  theme(axis.title.x=element_blank()) +
  theme(axis.text.x =element_blank()) +
  scale_fill_hp(discrete = TRUE, option = "NewtScamander", name = "Category",labels=c("Proteins with\n an intronic lncRNA","Protein with an intergenic\nlncRNA closer than 1 kb","Proteins in\nan overlapping lncRNA","Other\nproteins")) +
  theme(legend.text = element_text(size=9)) +
  stat_compare_means(comparisons = fpkm_prot_comp,label="p.signif",symnum.args = list(cutpoints = c(0, 1E-15, 0.001, 0.01, 0.05, 1), symbols = c("****", "***", "**", "*", "ns")))

fpkm_prot_plot
ggsave(fpkm_prot_plot,file="fpkm_prot_plot.png",width = 7,height = 4)

anova_class_prot<-aov(log(FPKM+1)~lncrna_class,data=read_counts_table_longer_prot_merged)
TukeyHSD(anova_class_prot)


kruskal.test(FPKM~lncrna_class,data=read_counts_table_longer_prot_merged)
DunnettTest(FPKM~lncrna_class,data=read_counts_table_longer_prot_merged)

wilcox.test(read_counts_table_longer_prot_merged[lncrna_class=="intronic"]$FPKM,read_counts_table_longer_prot_merged[lncrna_class=="no_lncRNA"]$FPKM)
wilcox.test(read_counts_table_longer_prot_merged[lncrna_class=="intergenic_1kb"]$FPKM,read_counts_table_longer_prot_merged[lncrna_class=="no_lncRNA"]$FPKM)
wilcox.test(read_counts_table_longer_prot_merged[lncrna_class=="overlapping"]$FPKM,read_counts_table_longer_prot_merged[lncrna_class=="no_lncRNA"]$FPKM)

read_counts_table_longer_prot_merged

figure_six<-ggdraw() +
  draw_plot(fpkm_lncrna_class_plot, x = 0, y = 0.5, width = .4, height = .4) +
  draw_plot(fpkm_lncrna_repeats_plot, x = 0, y = 0.05, width = .4, height = .4) +
  draw_plot(fpkm_prot_plot, x = .45, y = 0.3, width = .45, height = .45) +
  draw_plot_label(label = c("A", "B","C"), size = 15,
                  x = c(0.02, 0.02,0.49), y = c(0.915, 0.48,0.78))

figure_six

ggsave(figure_six,filename = "figure_six.jpg",height = 10,width = 12)

```



```{r}
#esu emu homology
blast_gr_q_split_reduce_unlist_dt_lengths_ephy<-blast_gr_q_split_reduce_unlist_dt_lengths[species=="Ephydatia_mulleri_transcriptome"]



blast_gr_q_split_reduce_unlist_dt_lengths_ephy[,lncrna_emu:=str_remove(sseqid,"Ephydatia_mulleri_transcriptome_")]


lncrnas_final_reduced_together_dt_emu_rnaspades<-fread("rnaspades_lncrna_paf_more_than_one_exon_percmatches_90_no_prot_cod_emu.txt",header = F)
lncrnas_final_reduced_together_dt_emu_trinity<-fread("trinity_lncrna_paf_more_than_one_exon_percmatches_90_no_prot_cod_emu.txt",header=F)


lncrnas_final_reduced_together_dt_emu<-rbind(lncrnas_final_reduced_together_dt_emu_rnaspades,lncrnas_final_reduced_together_dt_emu_trinity)
lncrnas_final_reduced_together_dt_emu<-lncrnas_final_reduced_together_dt_emu[,1]
setnames(lncrnas_final_reduced_together_dt_emu,"emu_lncrna")

blast_gr_q_split_reduce_unlist_dt_lengths_ephy_uniq<-unique(blast_gr_q_split_reduce_unlist_dt_lengths_ephy,by="lncrna_emu")

blast_gr_q_split_reduce_unlist_dt_lengths_ephy_uniq[lncrna_emu%in%lncrnas_final_reduced_together_dt_emu$emu_lncrna]

```





```{r}
#1 MB

  
lncrnas_final_reduced_together_20kb<-lncrnas_final_reduced_together[start(lncrnas_final_reduced_together)>1000]
lncrnas_final_reduced_together_20kb_short<-lncrnas_final_reduced_together[start(lncrnas_final_reduced_together)<=1000]
end(lncrnas_final_reduced_together_20kb_short)<-(start(lncrnas_final_reduced_together_20kb_short)-1)
start(lncrnas_final_reduced_together_20kb_short)<-(rep(1,length(lncrnas_final_reduced_together_20kb_short)))
new_starts<-start(lncrnas_final_reduced_together_20kb)-1000
start(lncrnas_final_reduced_together_20kb)<-new_starts
lncrnas_final_reduced_together_20kb_final<-c(lncrnas_final_reduced_together_20kb,lncrnas_final_reduced_together_20kb_short)




contig_length_table<-data.table(seqnames1=names(esu_genome),cont_length=sapply(esu_genome,length))

lncrnas_final_reduced_together_dt_cont_length_merge<-plyr::join(lncrnas_final_reduced_together_dt,contig_length_table,by="seqnames1")


lncrnas_final_reduced_together_20kb_right<-lncrnas_final_reduced_together[end(lncrnas_final_reduced_together)<(lncrnas_final_reduced_together_dt_cont_length_merge$cont_length-1000)]
lncrnas_final_reduced_together_20kb_short_right<-lncrnas_final_reduced_together[end(lncrnas_final_reduced_together)>(lncrnas_final_reduced_together_dt_cont_length_merge$cont_length-1000)]

lncrnas_final_reduced_together_20kb_short_right_dt<-as.data.table(lncrnas_final_reduced_together_20kb_short_right)
colnames(lncrnas_final_reduced_together_20kb_short_right_dt)[1]<-"seqnames1"
lncrnas_final_reduced_together_20kb_short_right_dt_merge<-plyr::join(lncrnas_final_reduced_together_20kb_short_right_dt,contig_length_table,by="seqnames1")

end(lncrnas_final_reduced_together_20kb_short_right)<-lncrnas_final_reduced_together_20kb_short_right_dt_merge$cont_length

end(lncrnas_final_reduced_together_20kb_right)<-end(lncrnas_final_reduced_together_20kb_right)+1000


lncrnas_final_reduced_together_20kb_dt<-as.data.table(lncrnas_final_reduced_together_20kb_final)

lncrnas_final_reduced_together_20kb_short_right_dt<-as.data.table(lncrnas_final_reduced_together_20kb_short_right)
colnames(lncrnas_final_reduced_together_20kb_short_right_dt)[3]<-"end_short"

lncrnas_final_reduced_together_20kb_right_dt<-as.data.table(lncrnas_final_reduced_together_20kb_right)
colnames(lncrnas_final_reduced_together_20kb_right_dt)[3]<-"end_norm"

lncrnas_final_reduced_together_20kb_dt_merged<-merge(lncrnas_final_reduced_together_20kb_dt,lncrnas_final_reduced_together_20kb_short_right_dt[,c(3,7)],by="gene_id",all.x=T)

lncrnas_final_reduced_together_20kb_dt_merged2<-merge(lncrnas_final_reduced_together_20kb_dt_merged,lncrnas_final_reduced_together_20kb_right_dt[,c(3,7)],by="gene_id",all.x=T)

lncrnas_final_reduced_together_20kb_dt_merged2[is.na(end_short),end:=end_norm]
lncrnas_final_reduced_together_20kb_dt_merged2[is.na(end_norm),end:=end_short]

lncrnas_final_reduced_together_20kb_dt_merged2_gr<-makeGRangesFromDataFrame(lncrnas_final_reduced_together_20kb_dt_merged2,keep.extra.columns = T)

prot_overlaps_20kb<-findOverlaps(protein_coding_genes_granges,lncrnas_final_reduced_together_20kb_dt_merged2_gr)
prot_overlaps_20kb_dt<-cbind(as.data.table(protein_coding_genes_granges[queryHits(prot_overlaps_20kb)]),as.data.table(lncrnas_final_reduced_together_20kb_dt_merged2_gr[subjectHits(prot_overlaps_20kb)]))
setnames(prot_overlaps_20kb_dt,c("seqnames1","start1","end1","width1","strand1","source","feature","score","strand_ignore","frame","attribute","seqnames2","start2","end2","width2","strand2","gene_id","gene","end_short","end_norm"))


```



```{r}
go_all_esu<-fread("trin_res.csv",header=T)
go_all_esu[gene_ontology_BLASTP==".",gene_ontology_BLASTP:=gene_ontology_BLASTX]
go_all_esu[gene_ontology_Pfam==".",gene_ontology_Pfam:=gene_ontology_BLASTP]

go_all_esu<-go_all_esu[,.SD,.SDcols=c("#gene_id","gene_ontology_Pfam")]
colnames(go_all_esu)[1]<-"attribute"
colnames(go_all_esu)[2]<-"go_names"
go_all_esu_nonempty<-go_all_esu[go_names!="."]


prot_overlaps_20kb_dt_GO<-merge(prot_overlaps_20kb_dt,go_all_esu_nonempty,by="attribute")

max_split<-max(str_count(prot_overlaps_20kb_dt_GO$go_names,"`"))+1
prot_overlaps_20kb_dt_GO[,paste("GO_split",1:max_split,sep="_"):=tstrsplit(go_names, "`")]

prot_overlaps_20kb_dt_GO_longer<-as.data.table(pivot_longer(prot_overlaps_20kb_dt_GO,cols = grep("GO_split",colnames(prot_overlaps_20kb_dt_GO),value = T),values_to = "GO_term"))
prot_overlaps_20kb_dt_GO_longer[,GO_term:=str_remove_all(GO_term,"\\^")]
prot_overlaps_20kb_dt_GO_longer<-prot_overlaps_20kb_dt_GO_longer[is.na(GO_term)==F]
prot_overlaps_20kb_dt_GO_longer<-prot_overlaps_20kb_dt_GO_longer[grepl("cellular_component",GO_term)==F]

prot_overlaps_20kb_dt_GO_longer[grepl("biological_process",GO_term),GO_category:="P"]
prot_overlaps_20kb_dt_GO_longer[grepl("molecular_function",GO_term),GO_category:="F"]
prot_overlaps_20kb_dt_GO_longer[,GO_term:=str_remove(GO_term,"GO:\\d+")]



prot_overlaps_20kb_dt_GO_longer_f<-prot_overlaps_20kb_dt_GO_longer[GO_category=="F"]
prot_overlaps_20kb_dt_GO_longer_f[,GO_term:=str_remove(GO_term,"molecular_function")]
prot_overlaps_20kb_dt_GO_longer_f_N<-prot_overlaps_20kb_dt_GO_longer_f[,.N,by=GO_term][order(-N)]

prot_overlaps_20kb_dt_GO_longer_p<-prot_overlaps_20kb_dt_GO_longer[GO_category=="P"]
prot_overlaps_20kb_dt_GO_longer_p[,GO_term:=str_remove(GO_term,"biological_process")]
prot_overlaps_20kb_dt_GO_longer_p_N<-prot_overlaps_20kb_dt_GO_longer_p[,.N,by=GO_term][order(-N)]

top_go_f<-prot_overlaps_20kb_dt_GO_longer_f_N[1:10]$GO_term
top_go_p<-prot_overlaps_20kb_dt_GO_longer_p_N[1:10]$GO_term

prot_overlaps_20kb_dt_GO_longer_f_N<-prot_overlaps_20kb_dt_GO_longer_f_N[GO_term%in%top_go_f]
prot_overlaps_20kb_dt_GO_longer_f_N[,GO_term:=factor(GO_term,levels = prot_overlaps_20kb_dt_GO_longer_f_N$GO_term)]


num_f<-ggplot(data=prot_overlaps_20kb_dt_GO_longer_f_N,aes(x=reorder(GO_term,N),y=N,fill=GO_term)) +
  geom_bar(stat="identity",color="black",alpha=0.9) +
   theme_bw() +
  theme(legend.position = "none") +
  scale_fill_brewer(palette="Spectral") +
  coord_flip() +
  theme(axis.text.y = element_text(size=7)) +
  ylab("Number of proteins") +
  xlab("GO term")

num_f

ggsave(num_f,filename = "num_f.png",height = 4,width = 6)


prot_overlaps_20kb_dt_GO_longer_p_N<-prot_overlaps_20kb_dt_GO_longer_p_N[GO_term%in%top_go_p]
prot_overlaps_20kb_dt_GO_longer_p_N[,GO_term:=factor(GO_term,levels = prot_overlaps_20kb_dt_GO_longer_p_N$GO_term)]


num_p<-ggplot(data=prot_overlaps_20kb_dt_GO_longer_p_N,aes(x=reorder(GO_term,N),y=N,fill=GO_term)) +
  geom_bar(stat="identity",color="black",alpha=0.9) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_fill_brewer(palette="Spectral") +
    coord_flip() +
  theme(axis.text.y = element_text(size=7)) +
  ylab("Number of proteins") +
  xlab("GO term")

num_p

ggsave(num_p,filename = "num_p.png",height = 4,width = 6)





#GO distances - whole gene - KB







prot_overlaps_20kb_dt_lncrna_merge_f<-merge(prot_overlaps_20kb_dt_GO_longer_f,lncrnas_final_reduced_together_dt,by="gene_id")


prot_overlaps_20kb_dt_lncrna_merge_f[,distance:=ifelse(start1.y>start1.x,start1.y-end1.x-1,end1.y-start1.x+1)]


prot_overlaps_20kb_dt_lncrna_merge_p<-merge(prot_overlaps_20kb_dt_GO_longer_p,lncrnas_final_reduced_together_dt,by="gene_id")


prot_overlaps_20kb_dt_lncrna_merge_p[,distance:=ifelse(start1.y>start1.x,start1.y-end1.x-1,end1.y-start1.x+1)]



dist_table_all_go_f<-prot_overlaps_20kb_dt_lncrna_merge_f[GO_term%in%top_go_f]

dist_table_all_go_p<-prot_overlaps_20kb_dt_lncrna_merge_p[GO_term%in%top_go_p]


dist_table_all_go_f[,GO_term:=factor(GO_term,levels = prot_overlaps_20kb_dt_GO_longer_f_N$GO_term)]
dist_table_all_go_p[,GO_term:=factor(GO_term,levels = prot_overlaps_20kb_dt_GO_longer_p_N$GO_term)]








dist_table_all_go_f[,distance:=abs(distance)]
dist_table_all_go_f[,log_distance:=log(distance+1)]


dist_table_all_go_p[,distance:=abs(distance)]
dist_table_all_go_p[,log_distance:=log(distance+1)]

distance_f_prot<-ggplot(data=dist_table_all_go_f,aes(x=GO_term,y=distance,fill=GO_term)) +
  geom_boxplot()+
  theme_bw() +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45,hjust = 0.9,size = 7))  +
  xlab("GO term") +
  ylab("Distance to PC gene") +
  scale_y_log10() +
  scale_fill_brewer(palette="Spectral")


distance_f_prot

ggsave(distance_f_prot,filename = "distance_f_prot.png",height = 6,width = 9)


distance_p_prot<-ggplot(data=dist_table_all_go_p,aes(x=GO_term,y=distance,fill=GO_term)) +
  geom_boxplot()+
  theme_bw() +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45,hjust = 0.9,size = 7))  +
  xlab("GO term") +
  ylab("Distance to PC gene") +
  scale_y_log10() + 
  scale_fill_brewer(palette="Spectral")

distance_p_prot

ggsave(distance_p_prot,filename = "distance_p_prot.png",height = 6,width = 9)

#fisher F

# 4744 65020
#  627 3186
enr_f_set<-prot_overlaps_20kb_dt_GO_longer_f_N$GO_term
p_values_f<-c()
for(i in enr_f_set) {
  a<-go_all_esu[grepl(i,go_names),.N]
  b<-nrow(go_all_esu)-a
  d<-prot_overlaps_20kb_dt_GO_longer_f_N[GO_term==i]$N
  e<-nrow(prot_overlaps_20kb_dt)-d
  
  fisher_matrix<-matrix(c(a,b,d,e),nrow=2,byrow=T)
  p_values_f<-c(p_values_f,fisher.test(fisher_matrix)$p)
}
p_values_f<-p.adjust(p_values_f,method = "hochberg")

enr_p_set<-prot_overlaps_20kb_dt_GO_longer_p_N $GO_term
p_values_p<-c()
for(i in enr_p_set) {
  a<-go_all_esu[grepl(i,go_names),.N]
  b<-nrow(go_all_esu)-a
  d<-prot_overlaps_20kb_dt_GO_longer_p_N[GO_term==i]$N
  e<-nrow(prot_overlaps_20kb_dt)-d
  
  fisher_matrix<-matrix(c(a,b,d,e),nrow=2,byrow=T)
  p_values_p<-c(p_values_p,fisher.test(fisher_matrix)$p)
}
p_values_p<-p.adjust(p_values_p,method = "hochberg")
```

```{r}
#TSS
protein_coding_genes[,trans_start:=ifelse(strand_ignore=="+",start,end)]


dist_table_whole_trans_f<-merge(dist_table_all_go_f,protein_coding_genes[,.SD,.SDcols=c("attribute","trans_start")],by="attribute")

dist_table_whole_trans_p<-merge(dist_table_all_go_p,protein_coding_genes[,.SD,.SDcols=c("attribute","trans_start")],by="attribute")



dist_table_whole_trans_f[,start_dis:=abs(start2-trans_start)]
dist_table_whole_trans_f[,end_dis:=abs(end2-trans_start)]
dist_table_whole_trans_f[,distance:=ifelse(start_dis<end_dis,start_dis,end_dis)]

dist_table_whole_trans_p[,start_dis:=abs(start2-trans_start)]
dist_table_whole_trans_p[,end_dis:=abs(end2-trans_start)]
dist_table_whole_trans_p[,distance:=ifelse(start_dis<end_dis,start_dis,end_dis)]





dist_table_whole_trans_f<-dist_table_whole_trans_f[GO_term%in%top_go_f]

dist_table_whole_trans_p<-dist_table_whole_trans_p[GO_term%in%top_go_p]

dist_table_whole_trans_f[,GO_term:=factor(GO_term,levels = prot_overlaps_20kb_dt_GO_longer_f_N$GO_term)]
dist_table_whole_trans_p[,GO_term:=factor(GO_term,levels = prot_overlaps_20kb_dt_GO_longer_p_N$GO_term)]


dist_f<-ggplot(data=dist_table_whole_trans_f,aes(x=GO_term,y=distance,fill=GO_term)) +
  geom_boxplot(alpha=0.9)+
  theme_bw() +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45,hjust = 0.9,size = 9)) +
  scale_fill_brewer(palette = "Spectral")

dist_f
ggsave(dist_f,filename = "dist_f.png",height = 6,width = 9)




dist_p<-ggplot(data=dist_table_whole_trans_p,aes(x=GO_term,y=distance,fill=GO_term)) +
  geom_boxplot(alpha=0.9)+
  theme_bw() +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45,hjust = 0.9,size = 9)) +
  scale_fill_brewer(palette = "Spectral") 


dist_p
ggsave(dist_p,filename = "dist_p.png",height = 6,width = 9)


```



```{r}
#synteny analysis prot 

blocks_table<-fread("Repo_spec.txt.dagchainer.aligncoords_prot.spans",header = F)
setnames(blocks_table,c("Emu_scaffold","Emu_span_range","Emu_span_length","Esu_scaffold","Esu_span_range","Esu_span_length","span_strand","V8","V9"))
blocks_table[,Emu_scaffold:=str_remove(Emu_scaffold,"Emu;Emu")]
blocks_table[,Emu_scaffold:=str_replace(Emu_scaffold,"scaffold(\\d*)","scaffold_\\1")]
blocks_table[,Esu_scaffold:=str_remove(Esu_scaffold,"Esu;Esu")]
blocks_table[,Esu_scaffold:=str_remove(Esu_scaffold,"pilon")]
blocks_table[,Esu_scaffold:=str_replace(Esu_scaffold,"contig(\\d*)","contig_\\1")]
blocks_table[,Esu_scaffold:=paste(Esu_scaffold,"pilon",sep = "_")]

blocks_table[,Emu_start:=str_extract(Emu_span_range,".*(?=\\-)")]
blocks_table[,Emu_end:=str_extract(Emu_span_range,"(?<=\\-).*")]
blocks_table[,Esu_start:=str_extract(Esu_span_range,".*(?=\\-)")]
blocks_table[,Esu_end:=str_extract(Esu_span_range,"(?<=\\-).*")]
blocks_table[,Emu_ID:=paste(Emu_scaffold,Emu_span_range,sep="_")]
blocks_table[,Esu_ID:=paste(Esu_scaffold,Esu_span_range,sep="_")]

blocks_table_emu<-blocks_table[,.SD,.SDcols=c("Emu_scaffold","Emu_start","Emu_end","Esu_ID")]
blocks_table_esu<-blocks_table[,.SD,.SDcols=c("Esu_scaffold","Esu_start","Esu_end","Emu_ID")]

blocks_table_emu_gr<-makeGRangesFromDataFrame(blocks_table_emu,seqnames.field = "Emu_scaffold",start.field = "Emu_start",end.field = "Emu_end",ignore.strand = T,keep.extra.columns = T)

blocks_table_esu_gr<-makeGRangesFromDataFrame(blocks_table_esu,seqnames.field = "Esu_scaffold",start.field = "Esu_start",end.field = "Esu_end",ignore.strand = T,keep.extra.columns = T)

blocks_table_emu_gr_split<-split(blocks_table_emu_gr,blocks_table_emu_gr$Esu_ID)
blocks_table_esu_gr_split<-split(blocks_table_esu_gr,blocks_table_esu_gr$Emu_ID)

blocks_table_emu_gr_split_red<-reduce(blocks_table_emu_gr_split,ignore.strand=T)
blocks_table_esu_gr_split_red<-reduce(blocks_table_esu_gr_split,ignore.strand=T)

blocks_table_emu_gr_split_red_un<-unlist(blocks_table_emu_gr_split_red,use.names = T)
blocks_table_esu_gr_split_red_un<-unlist(blocks_table_esu_gr_split_red,use.names = T)

blocks_table_emu_gr_split_red_un$esu_ID<-names(blocks_table_emu_gr_split_red_un)
blocks_table_esu_gr_split_red_un$emu_ID<-names(blocks_table_esu_gr_split_red_un)

blocks_table_emu_gr_split_red_un_dt<-as.data.table(blocks_table_emu_gr_split_red_un)
blocks_table_esu_gr_split_red_un_dt<-as.data.table(blocks_table_esu_gr_split_red_un)

setnames(blocks_table_emu_gr_split_red_un_dt,c("Emu_scaffold_2","Emu_start_2","Emu_end_2","Emu_width_2","Emu_strand_2","Esu_ID"))

setnames(blocks_table_esu_gr_split_red_un_dt,c("Esu_scaffold","Esu_start","Esu_end","width_esu","strand_esu","Emu_ID"))

blocks_table_emu_merged<-merge(blocks_table_emu,blocks_table_emu_gr_split_red_un_dt,by="Esu_ID",allow.cartesian=TRUE)
blocks_table_emu_merged[,Emu_span_range_1:=paste(Emu_start,Emu_end,sep="-")]
blocks_table_emu_merged[,Emu_ID:=paste(Emu_scaffold,Emu_start,sep="_")]
blocks_table_emu_merged[,Emu_ID:=paste(Emu_ID,Emu_end,sep="-")]
blocks_table_emu_merged[,Emu_ID_two:=paste(Emu_scaffold,Emu_span_range_1,sep="_")]

blocks_table_emu_esu_merged<-merge(blocks_table_emu_merged,blocks_table_esu_gr_split_red_un_dt[,c(1:3,6)],by="Emu_ID",allow.cartesian=TRUE)
blocks_table_emu_merged[,Emu_ID:=Emu_ID_two]
blocks_table_emu_esu_final<-unique(blocks_table_emu_esu_merged,by=c("Emu_scaffold_2","Emu_start_2","Emu_end_2","Esu_scaffold","Esu_start","Esu_end"))[,.SD,.SDcols=c("Emu_scaffold_2","Emu_start_2","Emu_end_2","Esu_scaffold","Esu_start","Esu_end")]

blocks_table_emu_esu_final[,Emu_ID:=paste(Emu_scaffold_2,Emu_start_2,Emu_end_2,sep="_")]
blocks_table_emu_esu_final[,Esu_ID:=paste(Esu_scaffold,Esu_start,Esu_end,sep="_")]


Emu_block_final_dt<-blocks_table_emu_esu_final[,.SD,.SDcols=c("Emu_scaffold_2","Emu_start_2","Emu_end_2","Esu_ID")]


Esu_block_final_dt<-blocks_table_emu_esu_final[,.SD,.SDcols=c("Esu_scaffold","Esu_start","Esu_end","Emu_ID")]



Emu_block_final_gr<-makeGRangesFromDataFrame(df=Emu_block_final_dt,seqnames.field = "Emu_scaffold_2",start.field = "Emu_start_2",end.field = "Emu_end_2",ignore.strand = T,keep.extra.columns = T)

Esu_block_final_gr<-makeGRangesFromDataFrame(df=Esu_block_final_dt,seqnames.field = "Esu_scaffold",start.field = "Esu_start",end.field = "Esu_end",ignore.strand = T,keep.extra.columns = T)

emu_prot_genes<-fread("emu_prot_genes.txt",header = F) 
setnames(emu_prot_genes,c("seqname","source","feature","start","end","score","strand","V8","attribute","width","gene"))
emu_prot_genes<-emu_prot_genes[,.SD,.SDcols=c("seqname","start","end","attribute")]
emu_prot_genes<-makeGRangesFromDataFrame(emu_prot_genes,ignore.strand = T,keep.extra.columns = T)


#emu_class_table<-fread("emu_class_table.txt",header = F)
#setnames(emu_class_table,c("transcript","class","gene_id"))





emu_span_prot_ovl<-findOverlaps(emu_prot_genes,Emu_block_final_gr,ignore.strand=T,type="within")
emu_span_prot_ovl_dt<-cbind(as.data.table(emu_prot_genes[queryHits(emu_span_prot_ovl)]),as.data.table(Emu_block_final_gr[subjectHits(emu_span_prot_ovl)]))
setnames(emu_span_prot_ovl_dt,c("seqnames_prot","start_prot","end_prot","width_prot","strand_prot","attribute","seqnames_span","start_span","end_span","width_span","strand_span","Esu_ID"))
emu_span_prot_ovl_dt_paste<-emu_span_prot_ovl_dt[,.(prots=paste0(attribute,collapse = " ")),by=c("seqnames_span","start_span","end_span","Esu_ID")]




esu_span_prot_ovl<-findOverlaps(protein_coding_genes_granges,Esu_block_final_gr,ignore.strand=T,type="within")
esu_span_prot_ovl_dt<-cbind(as.data.table(protein_coding_genes_granges[queryHits(esu_span_prot_ovl)]),as.data.table(Esu_block_final_gr[subjectHits(esu_span_prot_ovl)]))
setnames(esu_span_prot_ovl_dt,c("seqnames_prot","start_prot","end_prot","width_prot","strand_prot","source","feature","score","strand_ignore","frame","attribute","seqnames_span","start_span","end_span","width_span","strand_span","Emu_ID"))
esu_span_prot_ovl_dt_paste<-esu_span_prot_ovl_dt[,.(prots=paste0(attribute,collapse = " ")),by=c("seqnames_span","start_span","end_span","Emu_ID")]



esu_span_prot_ovl_dt_paste[,Esu_ID:=paste(seqnames_span,start_span,end_span,sep="_")]

synteny_table_merged_prot<-merge(emu_span_prot_ovl_dt_paste,esu_span_prot_ovl_dt_paste,by="Esu_ID",allow.cartesian=T)
setnames(synteny_table_merged_prot,c("Esu_ID","Emu_seqname","Emu_start","Emu_end","Emu_prots","Esu_seqname","Esu_start","Esu_end","Emu_ID","Esu_prots"))
synteny_table_merged_prot<-unique(synteny_table_merged_prot,by=c("Emu_seqname","Emu_start","Emu_end","Esu_seqname","Esu_start","Esu_end"))
keycol<-c("Emu_seqname","Esu_ID","Emu_start")
setorderv(synteny_table_merged_prot, keycol)

write.table(synteny_table_merged_prot,file="synteny_table_prot.txt",row.names = F,col.names = F,quote = F,sep = "\t")
```

```{r}
protein_coding_annotation_emu<-fread("Emu_genome_v1.gff",fill=T)[is.na(V4)==F]
setnames(protein_coding_annotation_emu,c("seqname","source","feature","start","end","score","strand_ignore","frame","attribute"))

protein_coding_annotation_emu[grepl("Name=",attribute),attribute:=str_extract(attribute,"(?=Name=).*")]
protein_coding_annotation_emu[grepl("Parent=",attribute),attribute:=str_extract(attribute,"(?=Parent=).*")]
protein_coding_annotation_emu[,attribute:=str_remove(attribute,"Name=")]
protein_coding_annotation_emu[,attribute:=str_remove(attribute,"Parent=")]

protein_coding_annotation_emu[feature!="gene",gene:=str_extract(attribute,".*(?=\\.t\\d*)")]
protein_coding_annotation_emu[feature=="gene",gene:=attribute]

lncrnas_final_reduced_together_dt_emu<-fread("lncrnas_final_reduced_together_dt_emu.txt")
colnames(lncrnas_final_reduced_together_dt_emu)[c(4,5)]<-c("width1","strand1")
lncrnas_final_reduced_together_gr_emu<-makeGRangesFromDataFrame(lncrnas_final_reduced_together_dt_emu,seqnames.field = "seqnames",start.field = "start",end.field = "end",ignore.strand = T,keep.extra.columns = T)


synteny_table_merged_prot[,Emu_start:=Emu_start+100]
synteny_table_merged_prot[,Emu_end:=Emu_end+100]
synteny_table_merged_prot[,Esu_start:=Esu_start+100]
synteny_table_merged_prot[,Esu_end:=Esu_end+100]

synteny_table_merged_prot_emu_gr<-makeGRangesFromDataFrame(synteny_table_merged_prot[,.SD,.SDcols=c("Emu_seqname","Emu_start","Emu_end","Emu_prots")],seqnames.field = "Emu_seqname",start.field = "Emu_start",end.field = "Emu_end",ignore.strand = T,keep.extra.columns = T)

synteny_table_merged_prot_emu_gr$block<-1:length(synteny_table_merged_prot_emu_gr)

synteny_table_merged_prot_esu_gr<-makeGRangesFromDataFrame(synteny_table_merged_prot[,.SD,.SDcols=c("Esu_seqname","Esu_start","Esu_end","Esu_prots")],seqnames.field = "Esu_seqname",start.field = "Esu_start",end.field = "Esu_end",ignore.strand = T,keep.extra.columns = T)

synteny_table_merged_prot_esu_gr$block<-1:length(synteny_table_merged_prot_esu_gr)
  
synteny_table_merged_prot_esu_gr_lncrna_ovl<-findOverlaps(lncrnas_final_reduced_together,synteny_table_merged_prot_esu_gr)
synteny_table_merged_prot_esu_gr_lncrna_ovl_dt<-cbind(as.data.table(lncrnas_final_reduced_together[queryHits(synteny_table_merged_prot_esu_gr_lncrna_ovl)]),as.data.table(synteny_table_merged_prot_esu_gr[subjectHits(synteny_table_merged_prot_esu_gr_lncrna_ovl)]))
setnames(synteny_table_merged_prot_esu_gr_lncrna_ovl_dt,c("seqnames1","start1","end1","width1","strand1","gene","gene_id","seqnames2","start2","end2","width2","strand2","Esu_prots","block"))

synteny_table_merged_prot_emu_gr_lncrna_ovl<-findOverlaps(lncrnas_final_reduced_together_gr_emu,synteny_table_merged_prot_emu_gr)
synteny_table_merged_prot_emu_gr_lncrna_ovl_dt<-cbind(as.data.table(lncrnas_final_reduced_together_gr_emu[queryHits(synteny_table_merged_prot_emu_gr_lncrna_ovl)]),as.data.table(synteny_table_merged_prot_emu_gr[subjectHits(synteny_table_merged_prot_emu_gr_lncrna_ovl)]))
setnames(synteny_table_merged_prot_emu_gr_lncrna_ovl_dt,c("seqnames1","start1","end1","width1","strand1","width_ignore","strand_ignore","gene","gene_id","seqnames2","start2","end2","width2","strand2","Emu_prots","block"))
block_lncrna_merge<-merge(synteny_table_merged_prot_esu_gr_lncrna_ovl_dt,synteny_table_merged_prot_emu_gr_lncrna_ovl_dt,by="block",allow.cartesian=T)

```

```{r}

emu_genome<-readDNAStringSet("Emu_genome_v1.fa")

setwd("F:/1_script/rad/synt")
 
block_lncrna_merge[,lncrna_esu_all:=paste0(gene_id.x,collapse = " "),by="block"]
block_lncrna_merge[,lncrna_emu_all:=paste0(gene_id.y,collapse = " "),by="block"]
block_lncrna_merge<-unique(block_lncrna_merge,by="block")

for (i in 1:length(unique(block_lncrna_merge$block))) {
numb<-block_lncrna_merge$block[i]
Esu_genome_ir<-IRanges(start = rep(1,length(unique(contig_length_table$seqname))),end=contig_length_table$cont_length)
Esu_genome_gr<-GRanges(seqnames=unique(contig_length_table$seqname),ranges = Esu_genome_ir)

esu_genome_track<-GenomeAxisTrack(range=Esu_genome_gr[contig_length_table$seqname==block_lncrna_merge[block==numb]$seqnames2.x][1])

protein_coding_annotation_clean2[feature!="gene",gene:=str_extract(attribute,".*(?=\\.t\\d*)")]
protein_coding_annotation_clean2[feature=="gene",gene:=attribute]



lncrna_sub<-lncrnas_final_reduced_together_dt[gene_id%in%str_split(block_lncrna_merge[block==numb]$lncrna_esu_all," " )[[1]]]


lncrna_sub_dt<-data.table(seqname=lncrna_sub$seqnames1,source="AUGUSTUS",feature="gene",start=lncrna_sub$start1,end=lncrna_sub$end1,score=".",strand_ignore="+",frame=".",attribute=lncrna_sub$gene_id,gene=lncrna_sub$gene_id)

prot_sub<-protein_coding_annotation_clean2[gene%in%str_split(block_lncrna_merge[block==numb]$Esu_prots," ")[[1]] & (feature =="gene")]


genes_sub<-rbind(prot_sub,lncrna_sub_dt)

colnames(genes_sub)[10]<-"symbol"

esu_genes_track<-GeneRegionTrack(genes_sub,genome=Esu_genome_gr[contig_length_table$seqname==block_lncrna_merge[block==numb]$seqnames2.x],transcriptAnnotation = "symbol",background.panel = "grey95",background.title = NA)


png(file=paste("Esu",numb,".png",sep=""),
width=1024, height=600)

plotTracks(list(esu_genome_track,esu_genes_track),shape="arrow",fill=c(rep("steelblue",nrow(prot_sub)),rep("indianred1",nrow(lncrna_sub_dt))),sizes = c(0.5,0.3),from=block_lncrna_merge[block==numb]$start2.x-100,to=block_lncrna_merge[block==numb]$end2.x+100,main = paste("Esu",numb,sep=" "),cex.main = 1,cex=2,fontsize.group=30)

dev.off()

contig_length_table_emu<-data.table(seqname=names(emu_genome),cont_length=sapply(emu_genome,length))

Emu_genome_ir<-IRanges(start = rep(1,length(unique(contig_length_table_emu$seqname))),end=contig_length_table_emu$cont_length)
Emu_genome_gr<-GRanges(seqnames=unique(contig_length_table_emu$seqname),ranges = Emu_genome_ir)

emu_genome_track<-GenomeAxisTrack(range=Emu_genome_gr[contig_length_table_emu$seqname==block_lncrna_merge[block==numb]$seqnames2.y][1])



lncrna_sub_emu<-lncrnas_final_reduced_together_dt_emu[gene_id%in%str_split(block_lncrna_merge[block==numb]$lncrna_emu_all," " )[[1]]]


lncrna_sub_dt_emu<-data.table(seqname=lncrna_sub_emu$seqnames,source="AUGUSTUS",feature="gene",start=lncrna_sub_emu$start,end=lncrna_sub_emu$end,score=".",strand_ignore="+",frame=".",attribute=lncrna_sub_emu$gene_id,gene=lncrna_sub_emu$gene_id)

prot_sub_emu<-protein_coding_annotation_emu[gene%in%str_split(block_lncrna_merge[block==numb]$Emu_prots," ")[[1]] & (feature =="gene")]


genes_sub_emu<-rbind(prot_sub_emu,lncrna_sub_dt_emu)

colnames(genes_sub_emu)[10]<-"symbol"

emu_genes_track<-GeneRegionTrack(genes_sub_emu,genome=Emu_genome_gr[contig_length_table_emu$seqname==block_lncrna_merge[block==numb]$seqnames2.y],transcriptAnnotation = "symbol",background.panel = "grey95",background.title = NA)


png(file=paste("Emu",numb,".png",sep=""),
width=1024, height=600)

plotTracks(list(emu_genome_track,emu_genes_track),shape="arrow",fill=c(rep("steelblue",nrow(prot_sub_emu)),rep("indianred1",nrow(lncrna_sub_dt_emu))),sizes = c(0.5,0.3),from=block_lncrna_merge[block==numb]$start2.y-100,to=block_lncrna_merge[block==numb]$end2.y+100,main = paste("Emu",numb,sep=" "),cex.main = 1,cex=2,fontsize.group=30)

dev.off()

}
```

```{r}
#synteny analysis - results

synt_pairs_table<-fread("synt_pairs2.csv")


synt_pairs_table_uq<-unique(synt_pairs_table,by="Esu_lncrna")

synt_pairs_table_uq_class<-synt_pairs_table_uq[,.N,by=class]

synt_pairs_table_uq_class



```

```{r}

synt_lncrna<-unique(synt_pairs_table$Esu_lncrna)
cons_lncrna<-unique(isoform_map_table_together[transcript_name%in%blast_gr_q_split_reduce_unlist_dt_lengths_one_per_species$qseqid]$gene_id)
synt_cons_lncrna<-unique(intersect(synt_lncrna,cons_lncrna))
synt_lncrna2<-setdiff(synt_lncrna,synt_cons_lncrna)
cons_lncrna2<-setdiff(cons_lncrna,synt_cons_lncrna)
other_lncrna<-unique(isoform_map_table_together[gene_id%in%c(synt_lncrna,cons_lncrna)==F]$gene_id)
synt_cons_lncrna_table<-data.table(gene_id=c(synt_lncrna2,cons_lncrna2,synt_cons_lncrna,other_lncrna),cons_cat=c(rep("syntenic",length(synt_lncrna2)),rep("conserved",length(cons_lncrna2)),rep("syntenic and conserved",length(synt_cons_lncrna)),rep("other",length(other_lncrna))))


prot_overlaps_20kb_dt_GO_longer_f_merged<-merge(prot_overlaps_20kb_dt_GO_longer_f,synt_cons_lncrna_table)
prot_overlaps_20kb_dt_GO_longer_p_merged<-merge(prot_overlaps_20kb_dt_GO_longer_p,synt_cons_lncrna_table)

prot_overlaps_20kb_dt_GO_longer_f_merged[GO_term=="guanyl-nucleotide exchange factor activity",GO_term:="GTP exchange activity"]


prot_overlaps_20kb_dt_GO_longer_f_merged_N<-prot_overlaps_20kb_dt_GO_longer_f_merged[,.N,by=c("GO_term","cons_cat")][order(-N)]
prot_overlaps_20kb_dt_GO_longer_p_merged_N<-prot_overlaps_20kb_dt_GO_longer_p_merged[,.N,by=c("GO_term","cons_cat")][order(-N)]

prot_overlaps_20kb_dt_GO_longer_f_merged_N<-prot_overlaps_20kb_dt_GO_longer_f_merged_N[GO_term%in%top_go_f]
prot_overlaps_20kb_dt_GO_longer_f_merged_N[,GO_term:=factor(GO_term,levels = rev(top_go_f))]

prot_overlaps_20kb_dt_GO_longer_p_merged_N<-prot_overlaps_20kb_dt_GO_longer_p_merged_N[GO_term%in%top_go_p]
prot_overlaps_20kb_dt_GO_longer_p_merged_N[,GO_term:=factor(GO_term,levels = rev(top_go_p))]

prot_overlaps_20kb_dt_GO_longer_f_merged_N[,cons_cat:=factor(cons_cat,levels=c("conserved","syntenic","syntenic and conserved","other"))]
prot_overlaps_20kb_dt_GO_longer_p_merged_N[,cons_cat:=factor(cons_cat,levels=c("conserved","syntenic","syntenic and conserved","other"))]



num_f_cons<-ggplot(data=prot_overlaps_20kb_dt_GO_longer_f_merged_N,aes(x=GO_term,N,y=N,fill=cons_cat)) +
  geom_bar(stat="identity",color="black",alpha=0.7,position="stack") +
  theme_bw() +
  theme(legend.position = "none") +
  scale_fill_brewer(palette="Spectral",name="lncRNA category") +
  xlab("GO term") +
  ylab("Number of protein-coding genes") +
  coord_flip() +
  theme(axis.text.y = element_text(size=8)) +
  theme(axis.title=element_text(size=8))


num_f_cons

ggsave(num_f_cons,filename = "num_f_cons.png",height = 6,width = 9)


ratio_f_cons<-ggplot(data=prot_overlaps_20kb_dt_GO_longer_f_merged_N,aes(x=GO_term,y=N,fill=cons_cat)) +
  geom_bar(stat="identity",color="black",alpha=0.7,position="fill") +
  theme_bw() +
  theme(legend.position = "top") +
  theme(axis.text.x = element_text(angle = 45,hjust = 0.9,size = 9)) +
  scale_fill_brewer(palette="Set2",direction = -1) +
    scale_fill_brewer(palette="Spectral",name="lncRNA category") +
  xlab("GO term") +
  ylab("lncRNA category ratio")


ratio_f_cons

ggsave(ratio_f_cons,filename = "ratio_f_cons.png",height = 6,width = 9)

num_p_cons<-ggplot(data=prot_overlaps_20kb_dt_GO_longer_p_merged_N,aes(x=GO_term,y=N,fill=cons_cat)) +
  geom_bar(stat="identity",color="black",alpha=0.7,position="stack") +
  theme_bw() +
  theme(legend.position = "none") +
  scale_fill_brewer(palette="Spectral",name="lncRNA category") +
  xlab("GO term") +
  ylab("Number of protein-coding genes") +
    coord_flip() +
  theme(axis.text.y = element_text(size=8)) +
  theme(axis.title=element_text(size=8))

num_p_cons
ggsave(num_p_cons,filename = "num_p_cons.png",height = 6,width = 9)

ratio_p_cons<-ggplot(data=prot_overlaps_20kb_dt_GO_longer_p_merged_N,aes(x=GO_term,y=N,fill=cons_cat)) +
  geom_bar(stat="identity",color="black",alpha=0.7,position="fill",width=0.7) +
  theme_bw() +
  theme(legend.position = "top") +
  theme(axis.text.x = element_text(angle = 45,hjust = 0.9,size = 7)) +
  scale_fill_brewer(palette="Spectral",name="lncRNA category") +
  xlab("GO term") +
  ylab("lncRNA category ratio")

ratio_p_cons

ggsave(ratio_p_cons,filename = "ratio_p_cons.png",height = 6,width = 9)



dist_table_whole_go_trans_f_merge<-merge(dist_table_whole_trans_f,synt_cons_lncrna_table,by="gene_id")

dist_table_whole_go_trans_p_merge<-merge(dist_table_whole_trans_p,synt_cons_lncrna_table,by="gene_id")

dist_table_whole_go_trans_f_merge[,GO_term:=factor(GO_term,levels = top_go_f)]

dist_table_whole_go_trans_p_merge[,GO_term:=factor(GO_term,levels = top_go_p)]


dist_table_whole_go_trans_f_merge[,cons_cat:=factor(cons_cat,levels=c("conserved","syntenic","syntenic and conserved","other"))]
dist_table_whole_go_trans_p_merge[,cons_cat:=factor(cons_cat,levels=c("conserved","syntenic","syntenic and conserved","other"))]



distance_f_cons<-ggplot(data=dist_table_whole_go_trans_f_merge,aes(x=GO_term,y=distance,fill=cons_cat)) +
  geom_boxplot(alpha=0.7)+
  theme_bw() +
  theme(legend.position = "top") +
  theme(axis.text.x = element_text(angle = 45,hjust = 0.9,size = 7)) +
  scale_fill_brewer(palette = "Set2") +
    scale_fill_brewer(palette = "Spectral",name="lncRNA category") +
  scale_y_log10() +
  xlab("GO term") +  
  ylab("Distance to TSS")

distance_f_cons


ggsave(distance_f_cons,filename = "distance_f_cons.png",height = 6,width = 9)



distance_p_cons<-ggplot(data=dist_table_whole_go_trans_p_merge,aes(x=GO_term,y=distance,fill=cons_cat)) +
  geom_boxplot(alpha=0.7)+
  theme_bw() +
  theme(legend.position = "top") +
  theme(axis.text.x = element_text(angle = 45,hjust = 0.9,size = 7)) +
  scale_fill_brewer(palette = "Spectral",name="lncRNA category") +
  scale_y_log10() +
  xlab("GO term") +  
  ylab("Distance to TSS")

distance_p_cons

ggsave(distance_p_cons,filename = "distance_p_cons.png",height = 6,width = 9)


distance_cons<-ggplot(data=dist_table_whole_go_trans_p_merge,aes(x=cons_cat,y=distance,fill=cons_cat)) +
  geom_boxplot(alpha=0.7,width=0.5)+
  theme_bw() +
  theme(legend.position = "top") +
  theme(axis.text.x = element_text(angle = 45,hjust = 0.9,size = 7)) +
  scale_fill_brewer(palette = "Set2") +
  scale_fill_brewer(palette = "Spectral",name="lncRNA category") +
    scale_y_log10() +
  xlab("GO term") +  
  ylab("Distance to TSS")

distance_cons

ggsave(distance_cons,filename = "distance_cons.png",height = 6,width = 9)


#class


prot_overlaps_20kb_dt_GO_longer_f_merged_class<-merge(prot_overlaps_20kb_dt_GO_longer_f_merged,lncRNA_classification_table_with_gene_id,by="gene_id")
prot_overlaps_20kb_dt_GO_longer_p_merged_class<-merge(prot_overlaps_20kb_dt_GO_longer_p_merged,lncRNA_classification_table_with_gene_id,by="gene_id")


prot_overlaps_20kb_dt_GO_longer_f_merged_class_N<-prot_overlaps_20kb_dt_GO_longer_f_merged_class[,.N,by=c("GO_term","cons_cat","class")][order(-N)]
prot_overlaps_20kb_dt_GO_longer_p_merged_class_N<-prot_overlaps_20kb_dt_GO_longer_p_merged_class[,.N,by=c("GO_term","cons_cat","class")][order(-N)]

prot_overlaps_20kb_dt_GO_longer_f_merged_class_N<-prot_overlaps_20kb_dt_GO_longer_f_merged_class_N[GO_term%in%top_go_f]
prot_overlaps_20kb_dt_GO_longer_f_merged_class_N[,GO_term:=factor(GO_term,levels = rev(top_go_f))]

prot_overlaps_20kb_dt_GO_longer_p_merged_class_N<-prot_overlaps_20kb_dt_GO_longer_p_merged_class_N[GO_term%in%top_go_p]
prot_overlaps_20kb_dt_GO_longer_p_merged_class_N[,GO_term:=factor(GO_term,levels = rev(top_go_p))]

prot_overlaps_20kb_dt_GO_longer_f_merged_class_N[,cons_cat:=factor(cons_cat,levels=c("conserved","syntenic","syntenic and conserved","other"))]
prot_overlaps_20kb_dt_GO_longer_p_merged_class_N[,cons_cat:=factor(cons_cat,levels=c("conserved","syntenic","syntenic and conserved","other"))]


num_f_cons_class<-ggplot(data=prot_overlaps_20kb_dt_GO_longer_f_merged_class_N,aes(x=GO_term,y=N,fill=cons_cat)) +
  geom_bar(stat="identity",color="black",alpha=0.7,position="stack") +
  theme_bw() +
  theme(legend.position = "top") +
  facet_grid(rows=vars(class)) +
  theme(axis.text.x = element_text(angle = 45,hjust = 0.9,size = 9)) +
  scale_fill_brewer(palette="Spectral",name="lncRNA category") +
  xlab("GO term") +
  ylab("lncRNA category number") +
  coord_flip()


num_f_cons_class

ggsave(num_f_cons_class,filename = "num_f_cons_class.png",height = 6,width = 9)


num_p_cons_class<-ggplot(data=prot_overlaps_20kb_dt_GO_longer_p_merged_class_N,aes(x=GO_term,y=N,fill=cons_cat)) +
  geom_bar(stat="identity",color="black",alpha=0.7,position="stack") +
  theme_bw() +
  theme(legend.position = "top") +
  facet_grid(rows=vars(class)) +
  scale_fill_brewer(palette="Spectral",name="lncRNA category") +
  xlab("GO term") +
  ylab("lncRNA category number") +
    coord_flip() +
    xlab("GO term") +
  ylab("Number of protein-coding genes") +
    coord_flip() +
  theme(axis.text.y = element_text(size=8)) +
  theme(axis.title=element_text(size=8))


num_p_cons_class

ggsave(num_p_cons_class,filename = "num_p_cons_class.png",height = 7,width = 9)
```
lncRNA_classification_table_with_gene_id

```{r}
#go and expression
prot_overlaps_20kb_dt_GO_longer_p_merged2<-prot_overlaps_20kb_dt_GO_longer_p_merged[GO_term%in%top_go_p]


read_counts_table_longer_prot2<-copy(read_counts_table_longer_prot)
colnames(read_counts_table_longer_prot2)[1]<-"attribute"
prot_overlaps_20kb_dt_GO_longer_p_merged2_exp<-merge(prot_overlaps_20kb_dt_GO_longer_p_merged2,read_counts_table_longer_prot2,by="attribute")

go_expression_plot<-ggplot(data=prot_overlaps_20kb_dt_GO_longer_p_merged2_exp,aes(x=GO_term,y=FPKM,fill=GO_term)) +
  geom_boxplot() +
  theme_bw() +
  theme(legend.position = "none") +
  scale_y_log10() +
  coord_flip() +
    theme(axis.text.y = element_text(size=8)) +
  theme(axis.title=element_text(size=8)) +
  xlab("GO term") +
  scale_fill_npg()

go_expression_plot

ggsave(go_expression_plot,filename = "go_expression_plot.jpg",height = 4,width = 6)



figure_seven

ggsave(figure_seven,filename = "figure_seven.jpg",height = 10,width = 10)

figure_seven<-ggdraw() +
  draw_plot(num_f_cons, x = 0, y = 0.48, width = .2, height = .38) +
  draw_plot(num_p_cons, x = .21, y = 0.48, width = .255, height = .38) +
  draw_plot(go_expression_plot, x = 0.05, y = 0.05, width = 0.35, height = 0.4) +
  draw_plot(num_p_cons_class, x = 0.5, y = 0.15, width = 0.5, height = 0.7) +
  draw_plot_label(label = c("A", "B", "D","C"), size = 15,
                  x = c(0.035, 0.27, 0.05,0.5), y = c(0.9, 0.9, 0.47,0.83))


figure_seven

ggsave(figure_seven,filename = "figure_seven.jpg",height = 10,width = 15)
```

```{r}

```

